// @hostaway/design-system/dist/components/TextEditor/index.esm.js
import { H as HtmlEmbedDialog, N as Node3, m as mergeAttributes } from '../_shared/TextEditor-JTjEQ5To.esm.js';
export { T as TextEditor } from '../_shared/TextEditor-JTjEQ5To.esm.js';
import { NodeViewWrapper, ReactNodeViewRenderer } from '@tiptap/react';
import { jsxs, jsx } from 'react/jsx-runtime';
import { useState } from 'react';
import { Code02 } from '@untitledui/icons';
import '../_shared/MyCustomIcon-BRHTusqN.esm.js';
import { c as cx } from '../_shared/cx-DnIR8RkH.esm.js';
import '@tiptap/extension-image';
import '@tiptap/extension-placeholder';
import '@tiptap/extension-text-align';
import '@tiptap/extension-text-style';
import '@tiptap/starter-kit';
import '../_shared/hint-text-Ck8Vj_nN.esm.js';
import 'react-aria-components';
import '../_shared/label-pgzo6IFb.esm.js';
import '../_shared/tooltip-l8auK4sw.esm.js';
import '@tiptap/pm/transform';
import '@tiptap/pm/commands';
import '@tiptap/pm/state';
import '@tiptap/pm/model';
import '@tiptap/pm/schema-list';
import '@tiptap/pm/view';
import '@tiptap/pm/keymap';
import '../_shared/TextArea-DAe_yhoc.esm.js';
import '../_shared/input-BOpcbRQD.esm.js';
import '../_shared/FieldWrapper-UtX7P_J-.esm.js';
import '../_shared/ConditionalTooltip-DT2TnX8y.esm.js';
import '../_shared/Modal-Kz0XPBwI.esm.js';
import '../_shared/close-button-CGPYesyF.esm.js';
import '../_shared/featured-icon-CNeK_nGg.esm.js';
import '../_shared/is-react-component-CYKUd65Q.esm.js';
import '../_shared/index-UrSxsuvd.esm.js';
import '../_shared/Button-C3YUMQZU.esm.js';
import '../_shared/BaseButton-CTtHjEZn.esm.js';
import '../_shared/BaseSelect-CHht82wi.esm.js';
import '../_shared/PopoverSearchInput-D81SVeJp.esm.js';
import '../_shared/InputField-8Udd41T0.esm.js';
import '../_shared/avatar-DeJPrYaB.esm.js';
import '@tiptap/react/menus';
import '../_shared/Icon-BYFFIpjv.esm.js';
import 'tailwind-merge';

function HtmlEmbedNodeView({ node, updateAttributes, deleteNode, selected, editor, }) {
    var _a, _b;
    const [isOpen, setIsOpen] = useState(false);
    const html = ((_a = node.attrs.html) !== null && _a !== void 0 ? _a : '');
    const isEditable = (_b = editor === null || editor === void 0 ? void 0 : editor.isEditable) !== null && _b !== void 0 ? _b : true;
    return (jsxs(NodeViewWrapper, { as: "div", "data-html-embed-wrapper": "", className: cx('group relative my-3 rounded-lg border border-dashed border-secondary bg-primary_subtle px-3 py-2', selected && 'ring-2 ring-brand'), children: [isEditable && (jsxs("div", { className: "absolute right-2 top-2 z-1 flex gap-1 opacity-0 transition group-hover:opacity-100 focus-within:opacity-100", contentEditable: false, children: [jsxs("button", { type: "button", onClick: () => setIsOpen(true), className: "flex items-center gap-1 rounded-md bg-primary px-2 py-1 text-xs font-medium text-secondary shadow-xs ring-1 ring-secondary_alt hover:bg-primary_hover", children: [jsx(Code02, { className: "size-3.5" }), "Edit HTML"] }), jsx("button", { type: "button", onClick: () => deleteNode(), className: "rounded-md bg-primary px-2 py-1 text-xs font-medium text-error-primary shadow-xs ring-1 ring-secondary_alt hover:bg-primary_hover", children: "Remove" })] })), html ? (jsx("pre", { contentEditable: false, className: "my-1 max-h-80 overflow-auto rounded-md bg-secondary px-3 py-2 pr-24 font-mono text-xs whitespace-pre-wrap break-all text-tertiary", children: html })) : (jsxs("div", { contentEditable: false, className: "flex items-center gap-2 py-2 text-sm text-tertiary", children: [jsx(Code02, { className: "size-4" }), "Empty HTML embed \u2014 click \"Edit HTML\" to add markup."] })), jsx(HtmlEmbedDialog, { isOpen: isOpen, initialHtml: html, onClose: () => setIsOpen(false), onSave: (next) => updateAttributes({ html: next }) })] }));
}

/**
 * `htmlEmbed` — TipTap node for inline raw-HTML blocks (iframes, scripts,
 * styles, custom widgets, marketing snippets). Designed as a drop-in
 * replacement for CKEditor's `htmlEmbed` plugin.
 *
 * The node is an atom (no inner editing) and serializes to
 * `<div data-html-embed>{{ html }}</div>` so existing CKEditor content
 * round-trips cleanly.
 *
 * **Security**: embedded HTML is rendered as-is — no sanitization. This
 * extension is intentionally a privileged feature; only enable it on
 * screens where authors are trusted (admins/owners). Consumers are
 * responsible for any sanitization they need at their own boundaries.
 */
const HtmlEmbedExtension = Node3.create({
    name: 'htmlEmbed',
    group: 'block',
    atom: true,
    selectable: true,
    draggable: true,
    addOptions() {
        return {
            HTMLAttributes: {},
        };
    },
    addAttributes() {
        return {
            html: {
                default: '',
                parseHTML: (element) => {
                    // Recover raw inner HTML so existing CKEditor output round-trips.
                    return element.innerHTML.trim();
                },
                renderHTML: () => ({}),
            },
        };
    },
    parseHTML() {
        return [{ tag: 'div[data-html-embed]' }];
    },
    renderHTML({ HTMLAttributes, node }) {
        var _a, _b;
        const html = ((_a = node.attrs.html) !== null && _a !== void 0 ? _a : '');
        const attrs = mergeAttributes((_b = this.options.HTMLAttributes) !== null && _b !== void 0 ? _b : {}, HTMLAttributes, { 'data-html-embed': '' });
        // TipTap's renderHTML cannot emit raw HTML strings as children, so we
        // serialize the embed via DOM nodes that carry the markup verbatim.
        const dom = document.createElement('div');
        Object.entries(attrs).forEach(([key, value]) => {
            if (value === undefined || value === null)
                return;
            dom.setAttribute(key, String(value));
        });
        dom.innerHTML = html;
        return dom;
    },
    addNodeView() {
        return ReactNodeViewRenderer(HtmlEmbedNodeView);
    },
    addCommands() {
        return {
            insertHtmlEmbed: (html) => ({ chain }) => chain()
                .insertContent({
                type: this.name,
                attrs: { html },
            })
                .run(),
            updateHtmlEmbed: (html) => ({ commands, state }) => {
                if (state.selection.empty)
                    return false;
                const node = state.doc.nodeAt(state.selection.from);
                if (!node || node.type.name !== this.name)
                    return false;
                return commands.updateAttributes(this.name, { html });
            },
        };
    },
});

export { HtmlEmbedExtension };
