// @hostaway/design-system/dist/components/TextEditor/index.cjs.js
'use strict';

var TextEditor = require('../_shared/TextEditor-Dwg_wxYI.cjs.js');
var react = require('@tiptap/react');
var jsxRuntime = require('react/jsx-runtime');
var h = require('react');
var icons = require('@untitledui/icons');
require('../_shared/MyCustomIcon-DYQF4lJT.cjs.js');
var cx = require('../_shared/cx-CmkbyMOt.cjs.js');
require('@tiptap/extension-image');
require('@tiptap/extension-placeholder');
require('@tiptap/extension-text-align');
require('@tiptap/extension-text-style');
require('@tiptap/starter-kit');
require('../_shared/hint-text-LRIVuQnk.cjs.js');
require('react-aria-components');
require('../_shared/label-D0WtNYRO.cjs.js');
require('../_shared/tooltip-DDB8niG0.cjs.js');
require('@tiptap/pm/transform');
require('@tiptap/pm/commands');
require('@tiptap/pm/state');
require('@tiptap/pm/model');
require('@tiptap/pm/schema-list');
require('@tiptap/pm/view');
require('@tiptap/pm/keymap');
require('../_shared/TextArea-D_Bt79ko.cjs.js');
require('../_shared/input-CVSUknGT.cjs.js');
require('../_shared/FieldWrapper-D6dXzMyx.cjs.js');
require('../_shared/ConditionalTooltip-FLw9yEoi.cjs.js');
require('../_shared/Modal-BVz6xjJq.cjs.js');
require('../_shared/close-button-C3L86jhr.cjs.js');
require('../_shared/featured-icon-CeUxkT8F.cjs.js');
require('../_shared/is-react-component-DkG20B86.cjs.js');
require('../_shared/index-BNlrFbvQ.cjs.js');
require('../_shared/Button-BE28ft73.cjs.js');
require('../_shared/BaseButton-CchEmlms.cjs.js');
require('../_shared/BaseSelect-0MCwMQdF.cjs.js');
require('../_shared/PopoverSearchInput-BFtG7lQx.cjs.js');
require('../_shared/InputField-DaibKVKA.cjs.js');
require('../_shared/avatar-C5INEgx9.cjs.js');
require('@tiptap/react/menus');
require('../_shared/Icon-BF_rHwZP.cjs.js');
require('tailwind-merge');

function HtmlEmbedNodeView({ node, updateAttributes, deleteNode, selected, editor, }) {
    var _a, _b;
    const [isOpen, setIsOpen] = h.useState(false);
    const html = ((_a = node.attrs.html) !== null && _a !== void 0 ? _a : '');
    const isEditable = (_b = editor === null || editor === void 0 ? void 0 : editor.isEditable) !== null && _b !== void 0 ? _b : true;
    return (jsxRuntime.jsxs(react.NodeViewWrapper, { as: "div", "data-html-embed-wrapper": "", className: cx.cx('group relative my-3 rounded-lg border border-dashed border-secondary bg-primary_subtle px-3 py-2', selected && 'ring-2 ring-brand'), children: [isEditable && (jsxRuntime.jsxs("div", { className: "absolute right-2 top-2 z-1 flex gap-1 opacity-0 transition group-hover:opacity-100 focus-within:opacity-100", contentEditable: false, children: [jsxRuntime.jsxs("button", { type: "button", onClick: () => setIsOpen(true), className: "flex items-center gap-1 rounded-md bg-primary px-2 py-1 text-xs font-medium text-secondary shadow-xs ring-1 ring-secondary_alt hover:bg-primary_hover", children: [jsxRuntime.jsx(icons.Code02, { className: "size-3.5" }), "Edit HTML"] }), jsxRuntime.jsx("button", { type: "button", onClick: () => deleteNode(), className: "rounded-md bg-primary px-2 py-1 text-xs font-medium text-error-primary shadow-xs ring-1 ring-secondary_alt hover:bg-primary_hover", children: "Remove" })] })), html ? (jsxRuntime.jsx("pre", { contentEditable: false, className: "my-1 max-h-80 overflow-auto rounded-md bg-secondary px-3 py-2 pr-24 font-mono text-xs whitespace-pre-wrap break-all text-tertiary", children: html })) : (jsxRuntime.jsxs("div", { contentEditable: false, className: "flex items-center gap-2 py-2 text-sm text-tertiary", children: [jsxRuntime.jsx(icons.Code02, { className: "size-4" }), "Empty HTML embed \u2014 click \"Edit HTML\" to add markup."] })), jsxRuntime.jsx(TextEditor.HtmlEmbedDialog, { isOpen: isOpen, initialHtml: html, onClose: () => setIsOpen(false), onSave: (next) => updateAttributes({ html: next }) })] }));
}

/**
 * `htmlEmbed` — TipTap node for inline raw-HTML blocks (iframes, scripts,
 * styles, custom widgets, marketing snippets). Designed as a drop-in
 * replacement for CKEditor's `htmlEmbed` plugin.
 *
 * The node is an atom (no inner editing) and serializes to
 * `<div data-html-embed>{{ html }}</div>` so existing CKEditor content
 * round-trips cleanly.
 *
 * **Security**: embedded HTML is rendered as-is — no sanitization. This
 * extension is intentionally a privileged feature; only enable it on
 * screens where authors are trusted (admins/owners). Consumers are
 * responsible for any sanitization they need at their own boundaries.
 */
const HtmlEmbedExtension = TextEditor.Node3.create({
    name: 'htmlEmbed',
    group: 'block',
    atom: true,
    selectable: true,
    draggable: true,
    addOptions() {
        return {
            HTMLAttributes: {},
        };
    },
    addAttributes() {
        return {
            html: {
                default: '',
                parseHTML: (element) => {
                    // Recover raw inner HTML so existing CKEditor output round-trips.
                    return element.innerHTML.trim();
                },
                renderHTML: () => ({}),
            },
        };
    },
    parseHTML() {
        return [{ tag: 'div[data-html-embed]' }];
    },
    renderHTML({ HTMLAttributes, node }) {
        var _a, _b;
        const html = ((_a = node.attrs.html) !== null && _a !== void 0 ? _a : '');
        const attrs = TextEditor.mergeAttributes((_b = this.options.HTMLAttributes) !== null && _b !== void 0 ? _b : {}, HTMLAttributes, { 'data-html-embed': '' });
        // TipTap's renderHTML cannot emit raw HTML strings as children, so we
        // serialize the embed via DOM nodes that carry the markup verbatim.
        const dom = document.createElement('div');
        Object.entries(attrs).forEach(([key, value]) => {
            if (value === undefined || value === null)
                return;
            dom.setAttribute(key, String(value));
        });
        dom.innerHTML = html;
        return dom;
    },
    addNodeView() {
        return react.ReactNodeViewRenderer(HtmlEmbedNodeView);
    },
    addCommands() {
        return {
            insertHtmlEmbed: (html) => ({ chain }) => chain()
                .insertContent({
                type: this.name,
                attrs: { html },
            })
                .run(),
            updateHtmlEmbed: (html) => ({ commands, state }) => {
                if (state.selection.empty)
                    return false;
                const node = state.doc.nodeAt(state.selection.from);
                if (!node || node.type.name !== this.name)
                    return false;
                return commands.updateAttributes(this.name, { html });
            },
        };
    },
});

exports.TextEditor = TextEditor.TextEditor;
exports.HtmlEmbedExtension = HtmlEmbedExtension;
