// @hostaway/design-system/dist/components/toast/index.esm.js
import { jsxs, jsx } from 'react/jsx-runtime';
import { useRef, useState, useCallback, useEffect } from 'react';
import { c as cx } from '../_shared/cx-DnIR8RkH.esm.js';
import { CheckCircle, AlertCircle, InfoCircle } from '@untitledui/icons';
import { A as Avatar } from '../_shared/avatar-DeJPrYaB.esm.js';
import { B as Button } from '../_shared/BaseButton-CTtHjEZn.esm.js';
import { C as CloseButton } from '../_shared/close-button-CGPYesyF.esm.js';
import { P as ProgressBar } from '../_shared/progress-indicators-OJSyQ_qE.esm.js';
import { F as FeaturedIcon } from '../_shared/featured-icon-CNeK_nGg.esm.js';
import { B as Button$1 } from '../_shared/Button-C3YUMQZU.esm.js';
import 'tailwind-merge';
import 'react-aria-components';
import '../_shared/is-react-component-CYKUd65Q.esm.js';
import '../_shared/tooltip-l8auK4sw.esm.js';

const DEFAULT_NOTIFICATION_WIDTH = "400px";
const DEFAULT_DISMISS_LABEL = "Dismiss";
const iconMap = {
    default: InfoCircle,
    brand: InfoCircle,
    gray: InfoCircle,
    error: AlertCircle,
    warning: AlertCircle,
    success: CheckCircle,
};
const IconNotification = ({ title, description, className, width = DEFAULT_NOTIFICATION_WIDTH, confirmLabel, dismissLabel = DEFAULT_DISMISS_LABEL, hideDismissLabel, children, icon, progress, progressFormatter = (value) => `${value}%`, onClose, onConfirm, color = "default", 'data-id': dataId, }) => {
    const showProgress = typeof progress === "number";
    return (jsxs("div", { style: { "--width": width }, className: cx("relative z-[var(--z-index)] flex max-w-full flex-col gap-4 rounded-xl bg-primary_alt p-4 shadow-lg ring ring-secondary_alt xs:w-[var(--width)] xs:flex-row", className), children: [jsx(FeaturedIcon, { icon: icon || iconMap[color], color: color === "default" ? "gray" : color, theme: color === "default" ? "modern" : "outline", size: "md" }), jsxs("div", { className: cx("flex flex-1 flex-col gap-3 md:pr-8", color !== "default" && "md:pt-0.5", showProgress && "gap-4"), children: [jsxs("div", { className: "flex flex-col gap-1", children: [jsx("p", { className: "text-sm font-semibold text-fg-primary", children: title }), jsx("p", { className: "text-sm text-fg-secondary", children: description })] }), showProgress && jsx(ProgressBar, { labelPosition: "bottom", value: progress, valueFormatter: progressFormatter }), jsxs("div", { className: "flex gap-3", children: [!hideDismissLabel && (jsx(Button, { onClick: onClose, size: "sm", color: "link-gray", children: dismissLabel })), confirmLabel && (jsx(Button, { onClick: onConfirm, size: "sm", color: "link-color", children: confirmLabel })), children] })] }), jsx("div", { className: "absolute top-2 right-2 flex items-center justify-center", children: jsx(CloseButton, { onClick: onClose, size: "sm", label: dismissLabel, "data-id": `${dataId}-close-button` }) })] }));
};
const AvatarNotification = ({ name, content, avatar, className, width = DEFAULT_NOTIFICATION_WIDTH, confirmLabel, dismissLabel = DEFAULT_DISMISS_LABEL, hideDismissLabel, children, date, onClose, onConfirm, 'data-id': dataId, }) => {
    return (jsxs("div", { style: { "--width": width }, className: cx("relative z-[var(--z-index)] flex max-w-full flex-col items-start gap-4 rounded-xl bg-primary_alt p-4 shadow-lg ring ring-secondary_alt xs:w-[var(--width)] xs:flex-row", className), children: [jsx(Avatar, { size: "md", src: avatar, alt: name, status: "online" }), jsxs("div", { className: "flex flex-col gap-3 pr-8", children: [jsxs("div", { className: "flex flex-col gap-1", children: [jsxs("div", { className: "flex items-center gap-2", children: [jsx("p", { className: "text-sm font-semibold text-fg-primary", children: name }), jsx("span", { className: "text-sm text-fg-quaternary", children: date })] }), jsx("p", { className: "text-sm text-fg-secondary", children: content })] }), jsxs("div", { className: "flex gap-3", children: [!hideDismissLabel && (jsx(Button, { onClick: onClose, size: "sm", color: "link-gray", children: dismissLabel })), confirmLabel && (jsx(Button, { onClick: onConfirm, size: "sm", color: "link-color", children: confirmLabel })), children] })] }), jsx("div", { className: "absolute top-2 right-2 flex items-center justify-center", children: jsx(CloseButton, { onClick: onClose, size: "sm", label: dismissLabel, "data-id": `${dataId}-close-button` }) })] }));
};
const ImageNotification = ({ title, description, className, confirmLabel, dismissLabel = DEFAULT_DISMISS_LABEL, hideDismissLabel, children, imageMobile, imageDesktop, onClose, onConfirm, 'data-id': dataId, }) => {
    return (jsxs("div", { style: {
            "--width": "496px",
        }, className: cx("relative z-[var(--z-index)] flex max-w-full flex-col gap-3 rounded-xl bg-primary_alt p-4 shadow-lg max-md:ring-1 max-md:ring-secondary_alt xs:w-[var(--width)] xs:flex-row xs:gap-0 md:p-0", className), children: [jsx("div", { className: "-my-px hidden w-40 shrink-0 overflow-hidden rounded-l-xl outline-1 -outline-offset-1 outline-black/10 md:block", children: jsx("img", { "aria-hidden": "true", src: imageMobile, alt: "", className: "t size-full object-cover" }) }), jsxs("div", { className: "flex flex-col gap-4 rounded-r-xl bg-primary_alt md:gap-3 md:p-4 md:pl-5 md:ring-1 md:ring-secondary_alt", children: [jsxs("div", { className: "flex flex-col gap-1 pr-8", children: [jsx("p", { className: "text-sm font-semibold text-fg-primary", children: title }), jsx("p", { className: "text-sm text-fg-secondary", children: description })] }), jsx("div", { className: "h-40 w-full overflow-hidden rounded-md bg-secondary md:hidden", children: jsx("img", { src: imageDesktop, alt: title, className: "size-full object-cover" }) }), jsxs("div", { className: "flex gap-3", children: [!hideDismissLabel && (jsx(Button, { onClick: onClose, size: "sm", color: "link-gray", children: dismissLabel })), confirmLabel && (jsx(Button, { onClick: onConfirm, size: "sm", color: "link-color", children: confirmLabel })), children] })] }), jsx("div", { className: "absolute top-2 right-2 flex items-center justify-center", children: jsx(CloseButton, { onClick: onClose, size: "sm", label: dismissLabel, "data-id": `${dataId}-close-button` }) })] }));
};

const DEFAULT_AUTO_CLOSE_DELAY = 5000;
const EXIT_ANIMATION_DURATION = 200;
const Toast = ({ autoClose = true, autoCloseDelay = DEFAULT_AUTO_CLOSE_DELAY, onClose, onConfirm, 'data-id': dataId, ...props }) => {
    var _a;
    const ref = useRef(null);
    const timer = useRef(null);
    const [exiting, setExiting] = useState(false);
    const exitTimer = useRef(null);
    const close = useCallback(() => {
        setExiting(true);
        exitTimer.current = setTimeout(() => {
            onClose === null || onClose === void 0 ? void 0 : onClose();
        }, EXIT_ANIMATION_DURATION);
    }, [onClose]);
    const startAutoClose = useCallback(() => {
        if (!autoClose)
            return;
        if (timer.current)
            clearTimeout(timer.current);
        timer.current = setTimeout(() => {
            close();
        }, autoCloseDelay);
    }, [autoClose, autoCloseDelay, close]);
    const resetAutoClose = useCallback(() => {
        if (timer.current)
            clearTimeout(timer.current);
    }, []);
    useEffect(() => {
        startAutoClose();
        return () => {
            resetAutoClose();
            if (exitTimer.current)
                clearTimeout(exitTimer.current);
        };
    }, [startAutoClose, resetAutoClose]);
    const shared = {
        onClose: close,
        onConfirm: () => {
            onConfirm === null || onConfirm === void 0 ? void 0 : onConfirm();
            close();
        },
    };
    const layoutClassName = cx(props.className, props.layout === 'stacked' && 'xs:flex-col');
    const customButtons = (_a = props.buttons) === null || _a === void 0 ? void 0 : _a.map((button, index) => 'url' in button ? (jsx(Button$1, { "data-id": `${dataId}-toast-button-${index}`, size: "sm", color: "link-color", href: button.url, children: button.text }, `${button.text}-${index}`)) : (jsx(Button$1, { "data-id": `${dataId}-toast-button-${index}`, size: "sm", color: "link-color", onClick: () => {
            button.onClick();
            close();
        }, children: button.text }, `${button.text}-${index}`)));
    const renderToast = () => {
        var _a, _b, _c, _d, _e, _f, _g;
        switch (props.render) {
            case 'icon':
                return (jsx(IconNotification, { title: props.title, description: (_a = props.description) !== null && _a !== void 0 ? _a : '', className: layoutClassName, width: props.width, confirmLabel: props.confirmLabel, dismissLabel: props.dismissLabel, hideDismissLabel: props.hideDismissLabel, icon: props.icon, color: props.color, progress: props.progress, progressFormatter: props.progressFormatter, "data-id": dataId, ...shared, children: customButtons }));
            case 'image':
                return (jsx(ImageNotification, { title: props.title, description: (_b = props.description) !== null && _b !== void 0 ? _b : '', className: layoutClassName, confirmLabel: (_c = props.confirmLabel) !== null && _c !== void 0 ? _c : '', dismissLabel: props.dismissLabel, hideDismissLabel: props.hideDismissLabel, imageMobile: props.imageMobile, imageDesktop: (_d = props.imageDesktop) !== null && _d !== void 0 ? _d : props.imageMobile, "data-id": dataId, ...shared, children: customButtons }));
            case 'avatar':
                return (jsx(AvatarNotification, { name: props.name, content: (_e = props.description) !== null && _e !== void 0 ? _e : '', className: layoutClassName, width: props.width, avatar: props.avatar, date: (_f = props.date) !== null && _f !== void 0 ? _f : '', confirmLabel: (_g = props.confirmLabel) !== null && _g !== void 0 ? _g : '', dismissLabel: props.dismissLabel, hideDismissLabel: props.hideDismissLabel, "data-id": dataId, ...shared, children: customButtons }));
            default:
                return null;
        }
    };
    return (jsx("div", { ref: ref, className: cx(exiting
            ? 'duration-200 ease-in animate-out fade-out slide-out-to-top-2'
            : 'duration-200 ease-out animate-in fade-in slide-in-from-bottom-2'), onMouseEnter: resetAutoClose, onMouseLeave: startAutoClose, children: renderToast() }));
};

export { Toast };
