// @hostaway/design-system/dist/components/Templater/index.cjs.js
'use strict';

var jsxRuntime = require('react/jsx-runtime');
var h = require('react');
var cx = require('../_shared/cx-CmkbyMOt.cjs.js');
var BaseSelect = require('../_shared/BaseSelect-0MCwMQdF.cjs.js');
var icons = require('@untitledui/icons');
var reactAriaComponents = require('react-aria-components');
var MenuPrimitives = require('../_shared/MenuPrimitives-BeA0-7lW.cjs.js');
var InputSearchField = require('../_shared/InputSearchField-jhe85A4A.cjs.js');
var Button = require('../_shared/Button-BE28ft73.cjs.js');
var InputField = require('../_shared/InputField-DaibKVKA.cjs.js');
var input = require('../_shared/input-CVSUknGT.cjs.js');
var FieldWrapper = require('../_shared/FieldWrapper-D6dXzMyx.cjs.js');
var TextArea = require('../_shared/TextArea-D_Bt79ko.cjs.js');
var TextEditor = require('../_shared/TextEditor-Dwg_wxYI.cjs.js');
var state = require('@tiptap/pm/state');
var view = require('@tiptap/pm/view');
require('tailwind-merge');
require('../_shared/MyCustomIcon-DYQF4lJT.cjs.js');
require('../_shared/Icon-BF_rHwZP.cjs.js');
require('../_shared/PopoverSearchInput-BFtG7lQx.cjs.js');
require('../_shared/avatar-C5INEgx9.cjs.js');
require('../_shared/is-react-component-DkG20B86.cjs.js');
require('../_shared/hint-text-LRIVuQnk.cjs.js');
require('../_shared/label-D0WtNYRO.cjs.js');
require('../_shared/tooltip-DDB8niG0.cjs.js');
require('../_shared/checkbox-D6ve3vaM.cjs.js');
require('../_shared/BaseButton-CchEmlms.cjs.js');
require('../_shared/ConditionalTooltip-FLw9yEoi.cjs.js');
require('@tiptap/extension-image');
require('@tiptap/extension-placeholder');
require('@tiptap/extension-text-align');
require('@tiptap/extension-text-style');
require('@tiptap/react');
require('@tiptap/starter-kit');
require('@tiptap/pm/transform');
require('@tiptap/pm/commands');
require('@tiptap/pm/model');
require('@tiptap/pm/schema-list');
require('@tiptap/pm/keymap');
require('../_shared/Modal-BVz6xjJq.cjs.js');
require('../_shared/close-button-C3L86jhr.cjs.js');
require('../_shared/featured-icon-CeUxkT8F.cjs.js');
require('../_shared/index-BNlrFbvQ.cjs.js');
require('@tiptap/react/menus');

function useTemplaterBackdrop(containerRef) {
    const backdropRef = h.useRef(null);
    const [field, setField] = h.useState(null);
    // Locate the underlying input/textarea once mounted so the resize observer
    // and listeners below can bind to the actual DOM node.
    h.useEffect(() => {
        var _a;
        const found = (_a = containerRef.current) === null || _a === void 0 ? void 0 : _a.querySelector('textarea, input');
        setField(found !== null && found !== void 0 ? found : null);
    }, [containerRef]);
    // Copy font styles once — they don't change at runtime
    h.useEffect(() => {
        if (!field || !backdropRef.current)
            return;
        const backdrop = backdropRef.current;
        const computed = window.getComputedStyle(field);
        backdrop.style.fontSize = computed.fontSize;
        backdrop.style.fontFamily = computed.fontFamily;
        backdrop.style.fontWeight = computed.fontWeight;
        backdrop.style.lineHeight = computed.lineHeight;
        backdrop.style.letterSpacing = computed.letterSpacing;
        backdrop.style.wordSpacing = computed.wordSpacing;
        const isTextarea = field.tagName === 'TEXTAREA';
        backdrop.style.whiteSpace = isTextarea ? 'pre-wrap' : 'pre';
        backdrop.style.wordBreak = isTextarea ? 'break-word' : 'normal';
    }, [field]);
    const positionRaf = h.useRef(0);
    const syncPosition = h.useCallback(() => {
        const container = containerRef.current;
        const backdrop = backdropRef.current;
        if (!container || !field || !backdrop)
            return;
        cancelAnimationFrame(positionRaf.current);
        positionRaf.current = requestAnimationFrame(() => {
            const containerRect = container.getBoundingClientRect();
            const fieldRect = field.getBoundingClientRect();
            backdrop.style.top = `${fieldRect.top - containerRect.top}px`;
            backdrop.style.left = `${fieldRect.left - containerRect.left}px`;
            backdrop.style.width = `${fieldRect.width}px`;
            backdrop.style.height = `${fieldRect.height}px`;
            // Re-read padding — it can change when icons / states toggle
            backdrop.style.padding = window.getComputedStyle(field).padding;
        });
    }, [containerRef, field]);
    // Initial sync once the field is located; cancel any pending frame on unmount.
    h.useEffect(() => {
        syncPosition();
        return () => cancelAnimationFrame(positionRaf.current);
    }, [syncPosition]);
    const fieldRef = h.useMemo(() => ({ current: field }), [field]);
    BaseSelect.useResizeObserver({ ref: fieldRef, onResize: syncPosition });
    // Keep the backdrop scrolled in lockstep with the field
    h.useEffect(() => {
        if (!field || !backdropRef.current) {
            return;
        }
        const backdrop = backdropRef.current;
        let scrollRaf = 0;
        const syncScroll = () => {
            cancelAnimationFrame(scrollRaf);
            scrollRaf = requestAnimationFrame(() => {
                backdrop.scrollTop = field.scrollTop;
                backdrop.scrollLeft = field.scrollLeft;
            });
        };
        field.addEventListener('scroll', syncScroll);
        return () => {
            cancelAnimationFrame(scrollRaf);
            field.removeEventListener('scroll', syncScroll);
        };
    }, [field]);
    return backdropRef;
}

const formatTemplate = (item) => `{{${item}}}`;

const TEMPLATE_VAR_REGEX$1 = /\{\{([^{}]*)\}\}/g;
const ESCAPE_MAP = {
    '&': '&amp;',
    '<': '&lt;',
    '>': '&gt;',
    '"': '&quot;',
    "'": '&#39;',
};
const ESCAPE_REGEX = /[&<>"']/g;
const ZERO_WIDTH_SPACE = String.fromCharCode(0x200b);
const TEMPLATE_VAR_HIGHLIGHT_CLASS = 'rounded-sm outline outline-1 outline-primary font-medium text-primary';
const TEMPLATE_VAR_BRACE_CLASS = 'text-placeholder_subtle';
function highlightTemplateVars(text, validKeywords) {
    const escaped = text.replace(ESCAPE_REGEX, (ch) => ESCAPE_MAP[ch]);
    const highlighted = escaped.replace(TEMPLATE_VAR_REGEX$1, (match, inner) => {
        if (!validKeywords.has(match)) {
            return match;
        }
        return `<span class="${TEMPLATE_VAR_HIGHLIGHT_CLASS}"><span class="${TEMPLATE_VAR_BRACE_CLASS}">{{</span>${inner}<span class="${TEMPLATE_VAR_BRACE_CLASS}">}}</span></span>`;
    });
    // A trailing "\n" in white-space:pre-wrap doesn't produce a visible empty
    // line-box, but a <textarea> does — so the caret sits on a line the backdrop
    // hasn't rendered. A zero-width space gives the final line something to
    // render around.
    return text.endsWith('\n')
        ? `${highlighted}${ZERO_WIDTH_SPACE}`
        : highlighted;
}

function TemplaterMenu({ templateItems, onSelect, isFormatTemplateItems = true, className = '', 'data-id': dataId, isDisabled, }) {
    const [isOpen, setIsOpen] = h.useState(false);
    const [searchTerm, setSearchTerm] = h.useState('');
    const filteredItems = h.useMemo(() => {
        const query = searchTerm.trim().toLowerCase();
        if (!query)
            return templateItems;
        return templateItems.filter((item) => item.toLowerCase().includes(query));
    }, [templateItems, searchTerm]);
    const handleOpenChange = (open) => {
        setIsOpen(open);
        if (!open)
            setSearchTerm('');
    };
    return (jsxRuntime.jsxs(reactAriaComponents.MenuTrigger, { isOpen: isOpen, onOpenChange: handleOpenChange, children: [jsxRuntime.jsx(Button.Button, { color: "tertiary", className: cx.cx('outline-hidden size-8', className), iconLeading: icons.BracketsPlus, isDisabled: isDisabled, "data-id": `${dataId}-trigger` }), jsxRuntime.jsx(MenuPrimitives.MenuPopover, { placement: "bottom end", children: jsxRuntime.jsxs("div", { className: "flex w-64 flex-col gap-1 px-1.5 py-1", children: [jsxRuntime.jsx(InputSearchField.InputSearchField, { placeholder: "Search", value: searchTerm, onChange: setSearchTerm, "data-id": `${dataId}-search`, className: "mt-[5px] text-sm", iconClassName: "size-4" }), jsxRuntime.jsx(MenuPrimitives.MenuList, { className: "max-h-64 overflow-x-hidden", children: filteredItems.map((item) => (jsxRuntime.jsx(MenuPrimitives.MenuItem, { id: item, textValue: item, onAction: () => onSelect(item), "data-id": `${dataId}-item-${item}`, children: jsxRuntime.jsx("span", { className: "block truncate text-sm", children: isFormatTemplateItems ? formatTemplate(item) : item }) }, item))) })] }) })] }));
}

function Templater({ component, templateItems, value = '', onChange, onBlur, className, isReadOnly, isDisabled, isInvalid, triggerClassName, menuTopPos = 'top-1', menuRightPos = 'right-1', isFormatTemplateItems = true, 'data-id': dataId, ...rest }) {
    const InputComponent = (component !== null && component !== void 0 ? component : InputField.Input);
    const showTrigger = !isReadOnly && !isDisabled;
    const containerRef = h.useRef(null);
    const backdropRef = useTemplaterBackdrop(containerRef);
    // Highlight whatever string actually gets inserted: when isFormatTemplateItems
    // is true items are wrapped as {{item}}, otherwise they're already pre-wrapped
    // (e.g. callers pass `{{name}}` directly), so we use them as-is for matching.
    const validKeywords = h.useMemo(() => new Set(isFormatTemplateItems
        ? templateItems.map((item) => formatTemplate(item))
        : templateItems), [templateItems, isFormatTemplateItems]);
    // Update backdrop HTML directly — bypasses React diffing for snappy paint
    h.useEffect(() => {
        if (!backdropRef.current)
            return;
        backdropRef.current.innerHTML = highlightTemplateVars(value !== null && value !== void 0 ? value : '', validKeywords);
    }, [value, validKeywords, backdropRef]);
    // Normalize: react-aria's TextField onChange yields a string, while a raw
    // textarea/input onChange yields a ChangeEvent.
    // Support both so the component works with Input and TextArea (and similar) without adapters.
    const handleInputChange = (next) => {
        if (typeof next === 'string') {
            onChange === null || onChange === void 0 ? void 0 : onChange(next);
            return;
        }
        if (next && typeof next === 'object' && 'target' in next) {
            onChange === null || onChange === void 0 ? void 0 : onChange(next.target.value);
        }
    };
    const handleSelect = (item) => {
        const inserted = isFormatTemplateItems ? formatTemplate(item) : item;
        onChange === null || onChange === void 0 ? void 0 : onChange(`${value !== null && value !== void 0 ? value : ''}${inserted}`);
    };
    return (jsxRuntime.jsxs("div", { ref: containerRef, className: cx.cx('relative w-full', 
        // Hide the underlying field's text so the backdrop's styled HTML shows
        // through. The caret stays visible.
        '[&_input]:!text-transparent [&_textarea]:!text-transparent', '[&_input]:caret-fg-brand-primary [&_textarea]:caret-fg-brand-primary', className), children: [jsxRuntime.jsx(InputComponent, { ...rest, isReadOnly: isReadOnly, isDisabled: isDisabled, isInvalid: isInvalid, value: value, onChange: handleInputChange, onBlur: onBlur, "data-id": `${dataId}-input` }), jsxRuntime.jsx("div", { ref: backdropRef, "aria-hidden": true, className: "pointer-events-none absolute z-0 overflow-hidden text-md text-primary" }), showTrigger && (jsxRuntime.jsx("div", { className: cx.cx('absolute z-10', menuTopPos, menuRightPos), children: jsxRuntime.jsx(TemplaterMenu, { "data-id": dataId, templateItems: templateItems, onSelect: handleSelect, className: triggerClassName, isFormatTemplateItems: isFormatTemplateItems }) }))] }));
}

function FieldInputTemplater({ label, hint, isInvalid, isRequired, tooltip, description, id = '', fieldSize = 'md', labelPosition = 'left', templateItems, value = '', onChange, onBlur, className, isReadOnly, isDisabled, isFormatTemplateItems = true, placeholder, shortcut, 'data-id': dataId, ...rest }) {
    const containerRef = h.useRef(null);
    const backdropRef = useTemplaterBackdrop(containerRef);
    const showTrigger = !isReadOnly && !isDisabled;
    const validKeywords = h.useMemo(() => new Set(isFormatTemplateItems
        ? templateItems.map((item) => formatTemplate(item))
        : templateItems), [templateItems, isFormatTemplateItems]);
    h.useEffect(() => {
        if (!backdropRef.current)
            return;
        backdropRef.current.innerHTML = highlightTemplateVars(value !== null && value !== void 0 ? value : '', validKeywords);
    }, [value, validKeywords, backdropRef]);
    const handleSelect = (item) => {
        const inserted = isFormatTemplateItems ? formatTemplate(item) : item;
        onChange === null || onChange === void 0 ? void 0 : onChange(`${value !== null && value !== void 0 ? value : ''}${inserted}`);
    };
    return (jsxRuntime.jsx(FieldWrapper.FieldWrapper, { id: id, label: label, hint: hint, isInvalid: isInvalid, isRequired: isRequired, tooltip: tooltip, description: description, fieldSize: fieldSize, labelPosition: labelPosition, children: jsxRuntime.jsxs("div", { ref: containerRef, className: cx.cx('relative w-full', 
            // Hide the underlying field's text so the backdrop's styled HTML
            // shows through. The caret stays visible.
            '[&_input]:!text-transparent [&_input]:caret-fg-brand-primary', className), children: [jsxRuntime.jsx(input.TextField, { ...rest, value: value, onChange: onChange, onBlur: onBlur, isInvalid: isInvalid, isReadOnly: isReadOnly, isDisabled: isDisabled, children: jsxRuntime.jsxs("div", { className: cx.cx(
                        // Unified rounded ring around InputBase + trigger, matching
                        // InputGroup's select-mode look (input-group.tsx).
                        'group relative flex h-max w-full flex-row items-stretch rounded-lg bg-primary shadow-xs ring-1 ring-primary ring-inset transition-shadow duration-100 ease-linear', 'focus-within:ring-2 focus-within:ring-brand', isInvalid && 'ring-error_subtle focus-within:ring-error', isDisabled &&
                            'cursor-not-allowed bg-disabled_subtle ring-disabled'), children: [jsxRuntime.jsx(input.InputBase, { placeholder: placeholder, shortcut: shortcut, isDisabled: isDisabled, wrapperClassName: "!shadow-none !ring-0 !rounded-none bg-transparent group-focus-within:!ring-0", "data-id": `${dataId}-input` }), jsxRuntime.jsx("div", { className: "flex shrink-0 items-center", children: jsxRuntime.jsx(TemplaterMenu, { "data-id": dataId, templateItems: templateItems, onSelect: handleSelect, isFormatTemplateItems: isFormatTemplateItems, isDisabled: !showTrigger, className: "mr-1" }) })] }) }), jsxRuntime.jsx("div", { ref: backdropRef, "aria-hidden": true, className: "pointer-events-none absolute z-20 overflow-hidden text-md text-primary" })] }) }));
}

function FieldTextAreaTemplater({ label, hint, isInvalid, isRequired, tooltip, description, id = '', fieldSize = 'md', labelPosition = 'left', ...props }) {
    return (jsxRuntime.jsx(FieldWrapper.FieldWrapper, { id: id, label: label, hint: hint, isInvalid: isInvalid, isRequired: isRequired, tooltip: tooltip, description: description, fieldSize: fieldSize, labelPosition: labelPosition, children: jsxRuntime.jsx(Templater, { ...props, isInvalid: isInvalid, component: TextArea.TextAreaBase, menuTopPos: "top-[5px]", menuRightPos: "right-[15px]" }) }));
}

const TEMPLATE_VAR_REGEX = /\{\{([^{}]*)\}\}/g;
function buildDecorations(doc, validKeywords) {
    if (validKeywords.size === 0) {
        return view.DecorationSet.empty;
    }
    const decorations = [];
    doc.descendants((node, pos) => {
        if (!node.isText || !node.text)
            return;
        const text = node.text;
        let match;
        TEMPLATE_VAR_REGEX.lastIndex = 0;
        while ((match = TEMPLATE_VAR_REGEX.exec(text)) !== null) {
            const full = match[0];
            if (!validKeywords.has(full))
                continue;
            const start = pos + match.index;
            const end = start + full.length;
            // ProseMirror inline decorations split the range at every overlapping
            // decoration boundary, so nesting an outer "highlight" with inner
            // "brace" decorations produces three separate spans — each carrying
            // the rounded outline class. Emit a single decoration so the whole
            // `{{var}}` renders as one continuous border.
            decorations.push(view.Decoration.inline(start, end, { class: TEMPLATE_VAR_HIGHLIGHT_CLASS }));
        }
    });
    return view.DecorationSet.create(doc, decorations);
}
const TemplateVarsHighlight = TextEditor.Extension.create({
    name: 'templateVarsHighlight',
    addOptions() {
        return {
            validKeywords: new Set(),
        };
    },
    addProseMirrorPlugins() {
        const extension = this;
        const pluginKey = new state.PluginKey('templateVarsHighlight');
        return [
            new state.Plugin({
                key: pluginKey,
                state: {
                    init: (_, state) => buildDecorations(state.doc, extension.options.validKeywords),
                    apply: (tr, oldSet) => {
                        if (!tr.docChanged)
                            return oldSet;
                        return buildDecorations(tr.doc, extension.options.validKeywords);
                    },
                },
                props: {
                    decorations(state) {
                        return pluginKey.getState(state);
                    },
                },
            }),
        ];
    },
});

function FieldTextEditorTemplaterMenu({ templateItems, isFormatTemplateItems, 'data-id': dataId, }) {
    const { editor, isDisabled } = TextEditor.useEditorContext();
    if (isDisabled)
        return null;
    const handleSelect = (item) => {
        const inserted = isFormatTemplateItems ? formatTemplate(item) : item;
        editor.chain().focus().insertContent(inserted).run();
    };
    return (jsxRuntime.jsx(TemplaterMenu, { "data-id": dataId, templateItems: templateItems, onSelect: handleSelect, isFormatTemplateItems: isFormatTemplateItems }));
}
function FieldTextEditorTemplater({ label, hint, isInvalid, isRequired, tooltip, description, id = '', fieldSize = 'md', labelPosition = 'left', templateItems, isFormatTemplateItems = true, 'data-id': dataId, extraExtensions, ...textEditorProps }) {
    // Highlight whatever string actually gets inserted: when isFormatTemplateItems
    // is true items are wrapped as {{item}}, otherwise they're already pre-wrapped
    // (e.g. callers pass `{{name}}` directly), so we use them as-is for matching.
    const validKeywords = h.useMemo(() => new Set(isFormatTemplateItems
        ? templateItems.map((item) => formatTemplate(item))
        : templateItems), [templateItems, isFormatTemplateItems]);
    const highlightExtensions = h.useMemo(() => [
        TemplateVarsHighlight.configure({ validKeywords }),
        ...(extraExtensions !== null && extraExtensions !== void 0 ? extraExtensions : []),
    ], [validKeywords, extraExtensions]);
    return (jsxRuntime.jsx(FieldWrapper.FieldWrapper, { id: id, label: label, hint: hint, isInvalid: isInvalid, isRequired: isRequired, tooltip: tooltip, description: description, fieldSize: fieldSize, labelPosition: labelPosition, children: jsxRuntime.jsx(TextEditor.TextEditor, { ...textEditorProps, isInvalid: isInvalid, "data-id": dataId, extraExtensions: highlightExtensions, toolbarExtras: [
                jsxRuntime.jsx(FieldTextEditorTemplaterMenu, { templateItems: templateItems, isFormatTemplateItems: isFormatTemplateItems, "data-id": `${dataId}-templater` }, "templates-menu"),
            ] }) }));
}

exports.FieldInputTemplater = FieldInputTemplater;
exports.FieldTextAreaTemplater = FieldTextAreaTemplater;
exports.FieldTextEditorTemplater = FieldTextEditorTemplater;
exports.Templater = Templater;
exports.TemplaterMenu = TemplaterMenu;
exports.highlightTemplateVars = highlightTemplateVars;
exports.useTemplaterBackdrop = useTemplaterBackdrop;
