// @hostaway/design-system/dist/components/Templater/index.esm.js
import { jsxs, jsx } from 'react/jsx-runtime';
import { useRef, useState, useEffect, useCallback, useMemo } from 'react';
import { c as cx } from '../_shared/cx-DnIR8RkH.esm.js';
import { u as useResizeObserver } from '../_shared/BaseSelect-CHht82wi.esm.js';
import { BracketsPlus } from '@untitledui/icons';
import { MenuTrigger } from 'react-aria-components';
import { a as MenuPopover, b as MenuList, c as MenuItem } from '../_shared/MenuPrimitives-CGzbU4z4.esm.js';
import { I as InputSearchField } from '../_shared/InputSearchField-CTy3JMq7.esm.js';
import { B as Button } from '../_shared/Button-C3YUMQZU.esm.js';
import { a as Input } from '../_shared/InputField-8Udd41T0.esm.js';
import { T as TextField, a as InputBase } from '../_shared/input-BOpcbRQD.esm.js';
import { F as FieldWrapper } from '../_shared/FieldWrapper-UtX7P_J-.esm.js';
import { a as TextAreaBase } from '../_shared/TextArea-DAe_yhoc.esm.js';
import { E as Extension, T as TextEditor, u as useEditorContext } from '../_shared/TextEditor-JTjEQ5To.esm.js';
import { PluginKey, Plugin } from '@tiptap/pm/state';
import { DecorationSet, Decoration } from '@tiptap/pm/view';
import 'tailwind-merge';
import '../_shared/MyCustomIcon-BRHTusqN.esm.js';
import '../_shared/Icon-BYFFIpjv.esm.js';
import '../_shared/PopoverSearchInput-D81SVeJp.esm.js';
import '../_shared/avatar-DeJPrYaB.esm.js';
import '../_shared/is-react-component-CYKUd65Q.esm.js';
import '../_shared/hint-text-Ck8Vj_nN.esm.js';
import '../_shared/label-pgzo6IFb.esm.js';
import '../_shared/tooltip-l8auK4sw.esm.js';
import '../_shared/checkbox-oR9JZpTn.esm.js';
import '../_shared/BaseButton-CTtHjEZn.esm.js';
import '../_shared/ConditionalTooltip-DT2TnX8y.esm.js';
import '@tiptap/extension-image';
import '@tiptap/extension-placeholder';
import '@tiptap/extension-text-align';
import '@tiptap/extension-text-style';
import '@tiptap/react';
import '@tiptap/starter-kit';
import '@tiptap/pm/transform';
import '@tiptap/pm/commands';
import '@tiptap/pm/model';
import '@tiptap/pm/schema-list';
import '@tiptap/pm/keymap';
import '../_shared/Modal-Kz0XPBwI.esm.js';
import '../_shared/close-button-CGPYesyF.esm.js';
import '../_shared/featured-icon-CNeK_nGg.esm.js';
import '../_shared/index-UrSxsuvd.esm.js';
import '@tiptap/react/menus';

function useTemplaterBackdrop(containerRef) {
    const backdropRef = useRef(null);
    const [field, setField] = useState(null);
    // Locate the underlying input/textarea once mounted so the resize observer
    // and listeners below can bind to the actual DOM node.
    useEffect(() => {
        var _a;
        const found = (_a = containerRef.current) === null || _a === void 0 ? void 0 : _a.querySelector('textarea, input');
        setField(found !== null && found !== void 0 ? found : null);
    }, [containerRef]);
    // Copy font styles once — they don't change at runtime
    useEffect(() => {
        if (!field || !backdropRef.current)
            return;
        const backdrop = backdropRef.current;
        const computed = window.getComputedStyle(field);
        backdrop.style.fontSize = computed.fontSize;
        backdrop.style.fontFamily = computed.fontFamily;
        backdrop.style.fontWeight = computed.fontWeight;
        backdrop.style.lineHeight = computed.lineHeight;
        backdrop.style.letterSpacing = computed.letterSpacing;
        backdrop.style.wordSpacing = computed.wordSpacing;
        const isTextarea = field.tagName === 'TEXTAREA';
        backdrop.style.whiteSpace = isTextarea ? 'pre-wrap' : 'pre';
        backdrop.style.wordBreak = isTextarea ? 'break-word' : 'normal';
    }, [field]);
    const positionRaf = useRef(0);
    const syncPosition = useCallback(() => {
        const container = containerRef.current;
        const backdrop = backdropRef.current;
        if (!container || !field || !backdrop)
            return;
        cancelAnimationFrame(positionRaf.current);
        positionRaf.current = requestAnimationFrame(() => {
            const containerRect = container.getBoundingClientRect();
            const fieldRect = field.getBoundingClientRect();
            backdrop.style.top = `${fieldRect.top - containerRect.top}px`;
            backdrop.style.left = `${fieldRect.left - containerRect.left}px`;
            backdrop.style.width = `${fieldRect.width}px`;
            backdrop.style.height = `${fieldRect.height}px`;
            // Re-read padding — it can change when icons / states toggle
            backdrop.style.padding = window.getComputedStyle(field).padding;
        });
    }, [containerRef, field]);
    // Initial sync once the field is located; cancel any pending frame on unmount.
    useEffect(() => {
        syncPosition();
        return () => cancelAnimationFrame(positionRaf.current);
    }, [syncPosition]);
    const fieldRef = useMemo(() => ({ current: field }), [field]);
    useResizeObserver({ ref: fieldRef, onResize: syncPosition });
    // Keep the backdrop scrolled in lockstep with the field
    useEffect(() => {
        if (!field || !backdropRef.current) {
            return;
        }
        const backdrop = backdropRef.current;
        let scrollRaf = 0;
        const syncScroll = () => {
            cancelAnimationFrame(scrollRaf);
            scrollRaf = requestAnimationFrame(() => {
                backdrop.scrollTop = field.scrollTop;
                backdrop.scrollLeft = field.scrollLeft;
            });
        };
        field.addEventListener('scroll', syncScroll);
        return () => {
            cancelAnimationFrame(scrollRaf);
            field.removeEventListener('scroll', syncScroll);
        };
    }, [field]);
    return backdropRef;
}

const formatTemplate = (item) => `{{${item}}}`;

const TEMPLATE_VAR_REGEX$1 = /\{\{([^{}]*)\}\}/g;
const ESCAPE_MAP = {
    '&': '&amp;',
    '<': '&lt;',
    '>': '&gt;',
    '"': '&quot;',
    "'": '&#39;',
};
const ESCAPE_REGEX = /[&<>"']/g;
const ZERO_WIDTH_SPACE = String.fromCharCode(0x200b);
const TEMPLATE_VAR_HIGHLIGHT_CLASS = 'rounded-sm outline outline-1 outline-primary font-medium text-primary';
const TEMPLATE_VAR_BRACE_CLASS = 'text-placeholder_subtle';
function highlightTemplateVars(text, validKeywords) {
    const escaped = text.replace(ESCAPE_REGEX, (ch) => ESCAPE_MAP[ch]);
    const highlighted = escaped.replace(TEMPLATE_VAR_REGEX$1, (match, inner) => {
        if (!validKeywords.has(match)) {
            return match;
        }
        return `<span class="${TEMPLATE_VAR_HIGHLIGHT_CLASS}"><span class="${TEMPLATE_VAR_BRACE_CLASS}">{{</span>${inner}<span class="${TEMPLATE_VAR_BRACE_CLASS}">}}</span></span>`;
    });
    // A trailing "\n" in white-space:pre-wrap doesn't produce a visible empty
    // line-box, but a <textarea> does — so the caret sits on a line the backdrop
    // hasn't rendered. A zero-width space gives the final line something to
    // render around.
    return text.endsWith('\n')
        ? `${highlighted}${ZERO_WIDTH_SPACE}`
        : highlighted;
}

function TemplaterMenu({ templateItems, onSelect, isFormatTemplateItems = true, className = '', 'data-id': dataId, isDisabled, }) {
    const [isOpen, setIsOpen] = useState(false);
    const [searchTerm, setSearchTerm] = useState('');
    const filteredItems = useMemo(() => {
        const query = searchTerm.trim().toLowerCase();
        if (!query)
            return templateItems;
        return templateItems.filter((item) => item.toLowerCase().includes(query));
    }, [templateItems, searchTerm]);
    const handleOpenChange = (open) => {
        setIsOpen(open);
        if (!open)
            setSearchTerm('');
    };
    return (jsxs(MenuTrigger, { isOpen: isOpen, onOpenChange: handleOpenChange, children: [jsx(Button, { color: "tertiary", className: cx('outline-hidden size-8', className), iconLeading: BracketsPlus, isDisabled: isDisabled, "data-id": `${dataId}-trigger` }), jsx(MenuPopover, { placement: "bottom end", children: jsxs("div", { className: "flex w-64 flex-col gap-1 px-1.5 py-1", children: [jsx(InputSearchField, { placeholder: "Search", value: searchTerm, onChange: setSearchTerm, "data-id": `${dataId}-search`, className: "mt-[5px] text-sm", iconClassName: "size-4" }), jsx(MenuList, { className: "max-h-64 overflow-x-hidden", children: filteredItems.map((item) => (jsx(MenuItem, { id: item, textValue: item, onAction: () => onSelect(item), "data-id": `${dataId}-item-${item}`, children: jsx("span", { className: "block truncate text-sm", children: isFormatTemplateItems ? formatTemplate(item) : item }) }, item))) })] }) })] }));
}

function Templater({ component, templateItems, value = '', onChange, onBlur, className, isReadOnly, isDisabled, isInvalid, triggerClassName, menuTopPos = 'top-1', menuRightPos = 'right-1', isFormatTemplateItems = true, 'data-id': dataId, ...rest }) {
    const InputComponent = (component !== null && component !== void 0 ? component : Input);
    const showTrigger = !isReadOnly && !isDisabled;
    const containerRef = useRef(null);
    const backdropRef = useTemplaterBackdrop(containerRef);
    // Highlight whatever string actually gets inserted: when isFormatTemplateItems
    // is true items are wrapped as {{item}}, otherwise they're already pre-wrapped
    // (e.g. callers pass `{{name}}` directly), so we use them as-is for matching.
    const validKeywords = useMemo(() => new Set(isFormatTemplateItems
        ? templateItems.map((item) => formatTemplate(item))
        : templateItems), [templateItems, isFormatTemplateItems]);
    // Update backdrop HTML directly — bypasses React diffing for snappy paint
    useEffect(() => {
        if (!backdropRef.current)
            return;
        backdropRef.current.innerHTML = highlightTemplateVars(value !== null && value !== void 0 ? value : '', validKeywords);
    }, [value, validKeywords, backdropRef]);
    // Normalize: react-aria's TextField onChange yields a string, while a raw
    // textarea/input onChange yields a ChangeEvent.
    // Support both so the component works with Input and TextArea (and similar) without adapters.
    const handleInputChange = (next) => {
        if (typeof next === 'string') {
            onChange === null || onChange === void 0 ? void 0 : onChange(next);
            return;
        }
        if (next && typeof next === 'object' && 'target' in next) {
            onChange === null || onChange === void 0 ? void 0 : onChange(next.target.value);
        }
    };
    const handleSelect = (item) => {
        const inserted = isFormatTemplateItems ? formatTemplate(item) : item;
        onChange === null || onChange === void 0 ? void 0 : onChange(`${value !== null && value !== void 0 ? value : ''}${inserted}`);
    };
    return (jsxs("div", { ref: containerRef, className: cx('relative w-full', 
        // Hide the underlying field's text so the backdrop's styled HTML shows
        // through. The caret stays visible.
        '[&_input]:!text-transparent [&_textarea]:!text-transparent', '[&_input]:caret-fg-brand-primary [&_textarea]:caret-fg-brand-primary', className), children: [jsx(InputComponent, { ...rest, isReadOnly: isReadOnly, isDisabled: isDisabled, isInvalid: isInvalid, value: value, onChange: handleInputChange, onBlur: onBlur, "data-id": `${dataId}-input` }), jsx("div", { ref: backdropRef, "aria-hidden": true, className: "pointer-events-none absolute z-0 overflow-hidden text-md text-primary" }), showTrigger && (jsx("div", { className: cx('absolute z-10', menuTopPos, menuRightPos), children: jsx(TemplaterMenu, { "data-id": dataId, templateItems: templateItems, onSelect: handleSelect, className: triggerClassName, isFormatTemplateItems: isFormatTemplateItems }) }))] }));
}

function FieldInputTemplater({ label, hint, isInvalid, isRequired, tooltip, description, id = '', fieldSize = 'md', labelPosition = 'left', templateItems, value = '', onChange, onBlur, className, isReadOnly, isDisabled, isFormatTemplateItems = true, placeholder, shortcut, 'data-id': dataId, ...rest }) {
    const containerRef = useRef(null);
    const backdropRef = useTemplaterBackdrop(containerRef);
    const showTrigger = !isReadOnly && !isDisabled;
    const validKeywords = useMemo(() => new Set(isFormatTemplateItems
        ? templateItems.map((item) => formatTemplate(item))
        : templateItems), [templateItems, isFormatTemplateItems]);
    useEffect(() => {
        if (!backdropRef.current)
            return;
        backdropRef.current.innerHTML = highlightTemplateVars(value !== null && value !== void 0 ? value : '', validKeywords);
    }, [value, validKeywords, backdropRef]);
    const handleSelect = (item) => {
        const inserted = isFormatTemplateItems ? formatTemplate(item) : item;
        onChange === null || onChange === void 0 ? void 0 : onChange(`${value !== null && value !== void 0 ? value : ''}${inserted}`);
    };
    return (jsx(FieldWrapper, { id: id, label: label, hint: hint, isInvalid: isInvalid, isRequired: isRequired, tooltip: tooltip, description: description, fieldSize: fieldSize, labelPosition: labelPosition, children: jsxs("div", { ref: containerRef, className: cx('relative w-full', 
            // Hide the underlying field's text so the backdrop's styled HTML
            // shows through. The caret stays visible.
            '[&_input]:!text-transparent [&_input]:caret-fg-brand-primary', className), children: [jsx(TextField, { ...rest, value: value, onChange: onChange, onBlur: onBlur, isInvalid: isInvalid, isReadOnly: isReadOnly, isDisabled: isDisabled, children: jsxs("div", { className: cx(
                        // Unified rounded ring around InputBase + trigger, matching
                        // InputGroup's select-mode look (input-group.tsx).
                        'group relative flex h-max w-full flex-row items-stretch rounded-lg bg-primary shadow-xs ring-1 ring-primary ring-inset transition-shadow duration-100 ease-linear', 'focus-within:ring-2 focus-within:ring-brand', isInvalid && 'ring-error_subtle focus-within:ring-error', isDisabled &&
                            'cursor-not-allowed bg-disabled_subtle ring-disabled'), children: [jsx(InputBase, { placeholder: placeholder, shortcut: shortcut, isDisabled: isDisabled, wrapperClassName: "!shadow-none !ring-0 !rounded-none bg-transparent group-focus-within:!ring-0", "data-id": `${dataId}-input` }), jsx("div", { className: "flex shrink-0 items-center", children: jsx(TemplaterMenu, { "data-id": dataId, templateItems: templateItems, onSelect: handleSelect, isFormatTemplateItems: isFormatTemplateItems, isDisabled: !showTrigger, className: "mr-1" }) })] }) }), jsx("div", { ref: backdropRef, "aria-hidden": true, className: "pointer-events-none absolute z-20 overflow-hidden text-md text-primary" })] }) }));
}

function FieldTextAreaTemplater({ label, hint, isInvalid, isRequired, tooltip, description, id = '', fieldSize = 'md', labelPosition = 'left', ...props }) {
    return (jsx(FieldWrapper, { id: id, label: label, hint: hint, isInvalid: isInvalid, isRequired: isRequired, tooltip: tooltip, description: description, fieldSize: fieldSize, labelPosition: labelPosition, children: jsx(Templater, { ...props, isInvalid: isInvalid, component: TextAreaBase, menuTopPos: "top-[5px]", menuRightPos: "right-[15px]" }) }));
}

const TEMPLATE_VAR_REGEX = /\{\{([^{}]*)\}\}/g;
function buildDecorations(doc, validKeywords) {
    if (validKeywords.size === 0) {
        return DecorationSet.empty;
    }
    const decorations = [];
    doc.descendants((node, pos) => {
        if (!node.isText || !node.text)
            return;
        const text = node.text;
        let match;
        TEMPLATE_VAR_REGEX.lastIndex = 0;
        while ((match = TEMPLATE_VAR_REGEX.exec(text)) !== null) {
            const full = match[0];
            if (!validKeywords.has(full))
                continue;
            const start = pos + match.index;
            const end = start + full.length;
            // ProseMirror inline decorations split the range at every overlapping
            // decoration boundary, so nesting an outer "highlight" with inner
            // "brace" decorations produces three separate spans — each carrying
            // the rounded outline class. Emit a single decoration so the whole
            // `{{var}}` renders as one continuous border.
            decorations.push(Decoration.inline(start, end, { class: TEMPLATE_VAR_HIGHLIGHT_CLASS }));
        }
    });
    return DecorationSet.create(doc, decorations);
}
const TemplateVarsHighlight = Extension.create({
    name: 'templateVarsHighlight',
    addOptions() {
        return {
            validKeywords: new Set(),
        };
    },
    addProseMirrorPlugins() {
        const extension = this;
        const pluginKey = new PluginKey('templateVarsHighlight');
        return [
            new Plugin({
                key: pluginKey,
                state: {
                    init: (_, state) => buildDecorations(state.doc, extension.options.validKeywords),
                    apply: (tr, oldSet) => {
                        if (!tr.docChanged)
                            return oldSet;
                        return buildDecorations(tr.doc, extension.options.validKeywords);
                    },
                },
                props: {
                    decorations(state) {
                        return pluginKey.getState(state);
                    },
                },
            }),
        ];
    },
});

function FieldTextEditorTemplaterMenu({ templateItems, isFormatTemplateItems, 'data-id': dataId, }) {
    const { editor, isDisabled } = useEditorContext();
    if (isDisabled)
        return null;
    const handleSelect = (item) => {
        const inserted = isFormatTemplateItems ? formatTemplate(item) : item;
        editor.chain().focus().insertContent(inserted).run();
    };
    return (jsx(TemplaterMenu, { "data-id": dataId, templateItems: templateItems, onSelect: handleSelect, isFormatTemplateItems: isFormatTemplateItems }));
}
function FieldTextEditorTemplater({ label, hint, isInvalid, isRequired, tooltip, description, id = '', fieldSize = 'md', labelPosition = 'left', templateItems, isFormatTemplateItems = true, 'data-id': dataId, extraExtensions, ...textEditorProps }) {
    // Highlight whatever string actually gets inserted: when isFormatTemplateItems
    // is true items are wrapped as {{item}}, otherwise they're already pre-wrapped
    // (e.g. callers pass `{{name}}` directly), so we use them as-is for matching.
    const validKeywords = useMemo(() => new Set(isFormatTemplateItems
        ? templateItems.map((item) => formatTemplate(item))
        : templateItems), [templateItems, isFormatTemplateItems]);
    const highlightExtensions = useMemo(() => [
        TemplateVarsHighlight.configure({ validKeywords }),
        ...(extraExtensions !== null && extraExtensions !== void 0 ? extraExtensions : []),
    ], [validKeywords, extraExtensions]);
    return (jsx(FieldWrapper, { id: id, label: label, hint: hint, isInvalid: isInvalid, isRequired: isRequired, tooltip: tooltip, description: description, fieldSize: fieldSize, labelPosition: labelPosition, children: jsx(TextEditor, { ...textEditorProps, isInvalid: isInvalid, "data-id": dataId, extraExtensions: highlightExtensions, toolbarExtras: [
                jsx(FieldTextEditorTemplaterMenu, { templateItems: templateItems, isFormatTemplateItems: isFormatTemplateItems, "data-id": `${dataId}-templater` }, "templates-menu"),
            ] }) }));
}

export { FieldInputTemplater, FieldTextAreaTemplater, FieldTextEditorTemplater, Templater, TemplaterMenu, highlightTemplateVars, useTemplaterBackdrop };
