// @hostaway/design-system/dist/components/VideoPlayer/index.cjs.js
'use strict';

var jsxRuntime = require('react/jsx-runtime');
var h = require('react');
var icons = require('@untitledui/icons');
var reactAriaComponents = require('react-aria-components');
var cx = require('../_shared/cx-CmkbyMOt.cjs.js');
require('tailwind-merge');

const PlayIcon = (props) => {
    return (jsxRuntime.jsx("svg", { width: "16", height: "16", viewBox: "0 0 16 16", fill: "none", ...props, children: jsxRuntime.jsx("path", { d: "M2.19995 2.86327C2.19995 1.61155 3.57248 0.844595 4.63851 1.50061L12.9856 6.63731C14.0009 7.26209 14.0009 8.73784 12.9856 9.36262L4.63851 14.4993C3.57247 15.1553 2.19995 14.3884 2.19995 13.1367V2.86327Z", fill: "currentColor" }) }));
};
const PauseIcon = (props) => {
    return (jsxRuntime.jsxs("svg", { width: "16", height: "16", viewBox: "0 0 16 16", fill: "none", ...props, children: [jsxRuntime.jsx("path", { d: "M2.19995 2.5C2.19995 2.22386 2.42381 2 2.69995 2H5.19995C5.47609 2 5.69995 2.22386 5.69995 2.5V13.5C5.69995 13.7761 5.47609 14 5.19995 14H2.69995C2.42381 14 2.19995 13.7761 2.19995 13.5V2.5Z", fill: "currentColor" }), jsxRuntime.jsx("path", { d: "M10.2 2.5C10.2 2.22386 10.4238 2 10.7 2H13.2C13.4761 2 13.7 2.22386 13.7 2.5V13.5C13.7 13.7761 13.4761 14 13.2 14H10.7C10.4238 14 10.2 13.7761 10.2 13.5V2.5Z", fill: "currentColor" })] }));
};

const VideoPlayer$1 = ({ src, type = 'video/mp4', autoPlay = false, thumbnailUrl, thumbnailAlt, showThumbnailOverlay, className, thumbnailButtonClassName, size = 'md', }) => {
    const videoRef = h.useRef(null);
    const containerRef = h.useRef(null);
    const progressRef = h.useRef(null);
    const [isPlaying, setIsPlaying] = h.useState(false);
    const [currentTime, setCurrentTime] = h.useState(0);
    const [duration, setDuration] = h.useState(0);
    const [buffered, setBuffered] = h.useState(null);
    const [hoverTime, setHoverTime] = h.useState(null);
    const [hoverPosition, setHoverPosition] = h.useState(null);
    const [showThumbnail, setShowThumbnail] = h.useState(true);
    const [isFullscreen, setIsFullscreen] = h.useState(false);
    const [volume, setVolume] = h.useState(1);
    const [isMuted, setIsMuted] = h.useState(false);
    const [previousVolume, setPreviousVolume] = h.useState(1);
    const [playbackRate, setPlaybackRate] = h.useState(1);
    const playbackSpeeds = [1, 1.25, 1.5, 1.75, 2];
    const formatTime = (time) => {
        const minutes = Math.max(Math.floor(time / 60), 0);
        const seconds = Math.max(Math.floor(time % 60), 0);
        return `${minutes}:${seconds.toString().padStart(2, '0')}`;
    };
    const togglePlay = () => {
        setIsPlaying((prevPlaying) => {
            var _a, _b;
            prevPlaying ? (_a = videoRef.current) === null || _a === void 0 ? void 0 : _a.pause() : (_b = videoRef.current) === null || _b === void 0 ? void 0 : _b.play();
            return !prevPlaying;
        });
        setShowThumbnail(false);
    };
    const handleProgressClick = (e) => {
        if (progressRef.current && videoRef.current) {
            const rect = progressRef.current.getBoundingClientRect();
            const pos = (e.clientX - rect.left) / rect.width;
            videoRef.current.currentTime = pos * duration;
        }
    };
    const handleProgressHover = (e) => {
        if (progressRef.current && videoRef.current) {
            const rect = progressRef.current.getBoundingClientRect();
            const pos = (e.clientX - rect.left) / rect.width;
            setHoverTime(pos * duration);
            setHoverPosition(e.clientX - rect.left);
        }
    };
    const handleProgress = () => {
        if (videoRef.current) {
            setBuffered(videoRef.current.buffered);
        }
    };
    const toggleFullscreen = () => {
        var _a, _b;
        if (!document.fullscreenElement) {
            (_b = (_a = videoRef.current) === null || _a === void 0 ? void 0 : _a.parentElement) === null || _b === void 0 ? void 0 : _b.requestFullscreen();
        }
        else {
            document.exitFullscreen();
        }
    };
    const toggleMute = h.useCallback(() => {
        if (isMuted) {
            setVolume(previousVolume);
            if (videoRef.current) {
                videoRef.current.volume = previousVolume;
            }
        }
        else {
            setPreviousVolume(volume);
            setVolume(0);
            if (videoRef.current) {
                videoRef.current.volume = 0;
            }
        }
        setIsMuted(!isMuted);
    }, [isMuted, previousVolume, volume]);
    const toggleSpeed = () => {
        const currentIndex = playbackSpeeds.indexOf(playbackRate);
        const nextIndex = (currentIndex + 1) % playbackSpeeds.length;
        const nextSpeed = playbackSpeeds[nextIndex];
        setPlaybackRate(nextSpeed);
        if (videoRef.current) {
            videoRef.current.playbackRate = nextSpeed;
        }
    };
    const handleVolumeChange = (e) => {
        const newVolume = parseFloat(e.target.value);
        setVolume(newVolume);
        if (videoRef.current) {
            videoRef.current.volume = newVolume;
        }
        setIsMuted(newVolume === 0);
    };
    h.useEffect(() => {
        const video = videoRef.current;
        if (!video)
            return;
        const handleTimeUpdate = () => setCurrentTime(video.currentTime);
        const handleDurationChange = () => setDuration(video.duration);
        const handleEnded = () => {
            setIsPlaying(false);
            setShowThumbnail(true);
        };
        const handleKeyDown = (e) => {
            if (e.code === 'Space' || e.code === 'KeyK') {
                e.preventDefault();
                togglePlay();
            }
            if (e.code === 'F') {
                e.preventDefault();
                toggleFullscreen();
            }
            if (e.code === 'KeyM') {
                e.preventDefault();
                toggleMute();
            }
            if (e.code === 'ArrowUp') {
                e.preventDefault();
                const newVolume = Math.min(video.volume + 0.1, 1);
                video.volume = newVolume;
                setVolume(newVolume);
                setIsMuted(newVolume === 0);
            }
            if (e.code === 'ArrowDown') {
                e.preventDefault();
                const newVolume = Math.max(video.volume - 0.1, 0);
                video.volume = newVolume;
                setVolume(newVolume);
                setIsMuted(newVolume === 0);
            }
            if (e.code === 'ArrowLeft') {
                e.preventDefault();
                video.currentTime = Math.max(video.currentTime - 10, 0);
            }
            if (e.code === 'ArrowRight') {
                e.preventDefault();
                video.currentTime = Math.min(video.currentTime + 10, video.duration);
            }
        };
        video.addEventListener('timeupdate', handleTimeUpdate);
        video.addEventListener('durationchange', handleDurationChange);
        video.addEventListener('ended', handleEnded);
        video.addEventListener('progress', handleProgress);
        video.addEventListener('keydown', handleKeyDown);
        return () => {
            video.removeEventListener('timeupdate', handleTimeUpdate);
            video.removeEventListener('durationchange', handleDurationChange);
            video.removeEventListener('ended', handleEnded);
            video.removeEventListener('progress', handleProgress);
            video.removeEventListener('keydown', handleKeyDown);
        };
    }, [toggleMute]);
    h.useEffect(() => {
        const handleFullscreenChange = () => {
            setIsFullscreen(!!document.fullscreenElement);
        };
        document.addEventListener('fullscreenchange', handleFullscreenChange);
        return () => {
            document.removeEventListener('fullscreenchange', handleFullscreenChange);
        };
    }, []);
    const renderBuffered = () => {
        if (!buffered)
            return null;
        const segments = [];
        for (let i = 0; i < buffered.length; i++) {
            const start = (buffered.start(i) / duration) * 100;
            const end = (buffered.end(i) / duration) * 100;
            segments.push(jsxRuntime.jsx("div", { className: "pointer-events-none absolute h-full min-w-2 rounded-full bg-fg-white/50", style: {
                    left: `${start}%`,
                    width: `${end - start}%`,
                } }, i));
        }
        return segments;
    };
    const canShowTrickPlayTime = hoverTime !== null && hoverPosition !== null;
    return (jsxRuntime.jsxs("div", { ref: containerRef, className: cx.cx('group/video relative', 
        // Add focus ring when video is focused
        'outline-focus-ring has-[video:focus-visible]:outline-2 has-[video:focus-visible]:outline-offset-4', className), children: [jsxRuntime.jsxs("div", { onClick: togglePlay, className: cx.cx('group absolute inset-0 z-10 cursor-pointer rounded-[inherit] transition-all duration-300 ease-in', thumbnailUrl && showThumbnail
                    ? 'visible opacity-100'
                    : 'pointer-events-none invisible opacity-0'), children: [jsxRuntime.jsx("img", { src: thumbnailUrl, alt: thumbnailAlt, className: "size-full object-cover" }), jsxRuntime.jsx("div", { className: "absolute inset-0 flex items-center justify-center", children: jsxRuntime.jsx("div", { className: cx.cx('flex size-16 items-center justify-center', size !== 'sm' &&
                                'rounded-full bg-alpha-white/30 backdrop-blur transition duration-100 ease-linear group-hover:bg-alpha-white/40 hover:bg-alpha-white/40', thumbnailUrl && showThumbnail
                                ? 'scale-100'
                                : 'scale-90 duration-500', thumbnailButtonClassName), children: jsxRuntime.jsx(PlayIcon, { className: "size-5 text-fg-white" }) }) }), showThumbnailOverlay && (jsxRuntime.jsx("div", { className: "absolute inset-0 rounded-[inherit] bg-black/10 outline-2 -outline-offset-1 outline-black/10" }))] }), jsxRuntime.jsxs("video", { ref: videoRef, tabIndex: 0, autoPlay: autoPlay, className: "h-full w-full cursor-pointer rounded-[inherit] bg-black outline-1 -outline-offset-1 outline-black/10", onClick: togglePlay, onDoubleClick: toggleFullscreen, children: [jsxRuntime.jsx("source", { src: src, type: type }), "Your browser does not support the video tag."] }), jsxRuntime.jsx("div", { className: cx.cx('absolute right-0 bottom-0 left-0 translate-y-4 transform bg-linear-to-t from-black/20 to-transparent opacity-0 transition duration-150 ease-in will-change-transform', 
                // Make it visible when video is hovered
                'group-hover/video:translate-y-0 group-hover/video:opacity-100 group-hover/video:duration-200 group-hover/video:ease-out', 
                // Make it visible when video is focused
                'group-has-[video:focus-visible]/video:translate-y-0 group-has-[video:focus-visible]/video:opacity-100 group-has-[video:focus-visible]/video:duration-200 group-has-[video:focus-visible]/video:ease-out', 
                // Sizes
                size === 'sm' && 'px-1 pt-6 pb-1', size === 'md' && 'px-5 pt-10 pb-4', size === 'lg' && 'px-8 pt-12 pb-6'), children: jsxRuntime.jsxs("div", { className: cx.cx('flex items-center', size === 'lg' ? 'gap-1' : 'gap-0.5'), children: [jsxRuntime.jsxs(VideoPlayerButton, { crossOffset: -12, tooltipPlacement: "top start", tooltipTitle: isPlaying ? 'Pause' : 'Play', tooltipDescription: "Space", tabIndex: -1, onClick: togglePlay, "aria-label": isPlaying ? 'Pause' : 'Play', className: "group/play relative", children: [jsxRuntime.jsx(PauseIcon, { className: cx.cx('size-4 transition duration-100 ease-linear group-active/play:scale-[0.8]', isPlaying
                                        ? 'scale-100 opacity-100'
                                        : 'pointer-events-none scale-[0.8] opacity-0') }), jsxRuntime.jsx(PlayIcon, { className: cx.cx('absolute size-4 transition duration-100 ease-linear group-active/play:scale-[0.8]', isPlaying
                                        ? 'pointer-events-none scale-[0.8] opacity-0'
                                        : 'scale-100 opacity-100') })] }), size !== 'lg' && (jsxRuntime.jsx(VideoPlayerButton, { tooltipPlacement: "top", tooltipTitle: isMuted ? 'Unmute' : 'Mute', tooltipDescription: "M", tabIndex: -1, onClick: toggleMute, "aria-label": isMuted ? 'Unmute' : 'Mute', className: "group/play relative", children: isMuted || volume === 0 ? (jsxRuntime.jsx(icons.VolumeX, { className: "size-4" })) : (jsxRuntime.jsx(icons.VolumeMax, { className: "size-4" })) })), size === 'lg' && (jsxRuntime.jsxs("div", { className: "flex items-center rounded-md pr-2 outline-hidden transition duration-100 ease-linear hover:bg-alpha-white/20 hover:backdrop-blur-sm focus-visible:outline-offset-2 focus-visible:outline-white", children: [jsxRuntime.jsx("button", { tabIndex: -1, onClick: toggleMute, "aria-label": isMuted ? 'Unmute' : 'Mute', className: "flex size-8 cursor-pointer items-center justify-center", children: isMuted || volume === 0 ? (jsxRuntime.jsx(icons.VolumeX, { className: "size-4 text-fg-white" })) : (jsxRuntime.jsx(icons.VolumeMax, { className: "size-4 text-fg-white" })) }), jsxRuntime.jsx("input", { type: "range", min: "0", max: "1", step: "0.1", value: volume, tabIndex: -1, onChange: handleVolumeChange, className: "h-1 w-11 cursor-pointer appearance-none rounded-full bg-alpha-white/30 [&::-moz-range-thumb]:h-3 [&::-moz-range-thumb]:w-3 [&::-moz-range-thumb]:appearance-none [&::-moz-range-thumb]:rounded-full [&::-moz-range-thumb]:border-0 [&::-moz-range-thumb]:bg-white [&::-webkit-slider-thumb]:h-3 [&::-webkit-slider-thumb]:w-3 [&::-webkit-slider-thumb]:translate-x-0 [&::-webkit-slider-thumb]:appearance-none [&::-webkit-slider-thumb]:rounded-full [&::-webkit-slider-thumb]:bg-fg-white", style: {
                                        backgroundImage: `linear-gradient(to right, white ${volume * 100}%, transparent ${volume * 100}%)`,
                                    } })] })), jsxRuntime.jsxs("div", { className: cx.cx('flex min-w-0 flex-1 items-center gap-2 px-2', 
                            // Visually hide progress bar when size is sm
                            size === 'sm' && 'pointer-events-none invisible opacity-0'), children: [jsxRuntime.jsx("span", { className: "pointer-events-none text-xs font-semibold text-white", children: formatTime(currentTime) }), jsxRuntime.jsx("div", { ref: progressRef, onClick: handleProgressClick, onMouseMove: handleProgressHover, className: "group/progress -my-8 flex-1 cursor-pointer py-8", children: jsxRuntime.jsxs("div", { className: "relative h-2 flex-1 rounded-full bg-fg-white/30", children: [renderBuffered(), jsxRuntime.jsx("div", { className: "pointer-events-none absolute h-full min-w-2 rounded-full bg-fg-white", style: { width: `${(currentTime / duration) * 100}%` } }), jsxRuntime.jsx("div", { className: "pointer-events-none absolute top-1/2 h-8 w-px -translate-y-1/2 bg-fg-white/30 opacity-0 transition duration-100 ease-linear group-hover/progress:opacity-100", style: {
                                                    left: hoverPosition || 0,
                                                } }), jsxRuntime.jsx("div", { className: "pointer-events-none absolute bottom-6 -translate-x-1/2 translate-y-2 transform text-xs font-semibold text-fg-white opacity-0 group-hover/progress:translate-y-0 group-hover/progress:opacity-100", style: {
                                                    left: hoverPosition || 0,
                                                    transition: 'opacity .3s cubic-bezier(0.4, 0, 0.6, 1), transform .3s cubic-bezier(0.4, 0, 0.6, 1), translate .3s cubic-bezier(0.4, 0, 0.6, 1)',
                                                }, children: canShowTrickPlayTime ? formatTime(hoverTime) : '' })] }) }), jsxRuntime.jsxs("span", { className: "pointer-events-none text-xs font-semibold text-white", children: ["-", formatTime(duration - currentTime)] })] }), size === 'lg' && (jsxRuntime.jsxs(VideoPlayerButton, { tooltipPlacement: "top", tooltipTitle: "Playback speed", tooltipDescription: "\u2190 \u2192", tabIndex: -1, onClick: toggleSpeed, "aria-label": `Change playback speed. Current speed: ${playbackRate}x`, className: "flex items-center", children: [jsxRuntime.jsx("span", { className: "text-xs font-semibold", children: playbackRate }), jsxRuntime.jsx("svg", { viewBox: "0 0 8 8", fill: "none", className: "mt-0.5 size-2", children: jsxRuntime.jsx("path", { d: "M6 2L2 6M2 2L6 6", stroke: "white", strokeWidth: "1.5", strokeLinecap: "round", strokeLinejoin: "round" }) })] })), jsxRuntime.jsx(VideoPlayerButton, { tooltipPlacement: "top end", crossOffset: 16, tooltipTitle: isFullscreen ? 'Minimize' : 'Fullscreen', tooltipDescription: isFullscreen ? 'Escape' : 'F', tabIndex: -1, onClick: toggleFullscreen, "aria-label": isFullscreen ? 'Exit fullscreen' : 'Enter fullscreen', children: isFullscreen ? (jsxRuntime.jsx(icons.Minimize01, { className: "size-4" })) : (jsxRuntime.jsx(icons.Maximize01, { className: "size-4" })) })] }) })] }));
};
const VideoPlayerButton = ({ children, tooltipPlacement = 'top', tooltipTitle, tooltipDescription, crossOffset, ...props }) => {
    return (jsxRuntime.jsxs(reactAriaComponents.TooltipTrigger, { delay: 350, closeDelay: 0, children: [jsxRuntime.jsx(reactAriaComponents.Focusable, { children: jsxRuntime.jsx("button", { ...props, className: cx.cx('flex h-8 min-w-8 cursor-pointer items-center justify-center rounded-md p-2 text-white outline-hidden transition duration-100 ease-linear before:absolute before:size-6 hover:bg-alpha-white/20 hover:backdrop-blur-sm focus-visible:outline-offset-2 focus-visible:outline-white', props.className), children: children }) }), jsxRuntime.jsxs(reactAriaComponents.Tooltip, { offset: 8, placement: tooltipPlacement, crossOffset: crossOffset, className: ({ isEntering, isExiting }) => cx.cx('dark-mode pointer-events-none flex gap-1.5 rounded-lg bg-alpha-white/20 p-1 pl-2 shadow-lg ring-1 ring-alpha-white/10 backdrop-blur-sm ring-inset', isEntering &&
                    'origin-bottom ease-out animate-in fade-in slide-in-from-bottom-0.5 zoom-in-95', isExiting &&
                    'origin-bottom ease-in animate-out fade-out slide-out-to-bottom-0.5 zoom-out-95'), children: [jsxRuntime.jsx("div", { className: "text-sm font-semibold text-white", children: tooltipTitle }), jsxRuntime.jsx("div", { className: "rounded bg-alpha-white/30 px-[3px] py-[1px] font-sans text-xs font-semibold text-white shadow-xs ring-1 ring-alpha-white/10 ring-inset", children: tooltipDescription })] })] }));
};

const styles = {
    sm: {
        maxWidth: 'max-w-80',
        rounded: 'rounded-lg',
    },
    md: {
        maxWidth: 'max-w-140',
        rounded: 'rounded-lg',
    },
    lg: {
        maxWidth: 'max-w-180',
        rounded: 'rounded-xl',
    },
};
function VideoPlayer({ className, size = 'sm', ...props }) {
    return (jsxRuntime.jsx(VideoPlayer$1, { ...props, size: size, className: cx.cx('aspect-video w-full overflow-hidden', styles[size].maxWidth, styles[size].rounded, className) }));
}

exports.VideoPlayer = VideoPlayer;
