// @hostaway/design-system/dist/components/ProgressIndicator/index.cjs.js
'use strict';

var ProgressBarIndicator = require('../_shared/ProgressBarIndicator-h-8ifFuN.cjs.js');
var jsxRuntime = require('react/jsx-runtime');
var cx = require('../_shared/cx-CmkbyMOt.cjs.js');
require('../_shared/progress-indicators-DCN4AeTu.cjs.js');
require('tailwind-merge');

const sizes = cx.sortCx({
    xxs: {
        strokeWidth: 6,
        radius: 29,
        valueClass: "text-sm font-semibold text-primary",
        labelClass: "text-xs font-medium text-tertiary",
        halfCircleTextPosition: "absolute bottom-0.5 text-center",
    },
    xs: {
        strokeWidth: 16,
        radius: 72,
        valueClass: "text-display-xs font-semibold text-primary",
        labelClass: "text-xs font-medium text-tertiary",
        halfCircleTextPosition: "absolute bottom-0.5 text-center",
    },
    sm: {
        strokeWidth: 20,
        radius: 90,
        valueClass: "text-display-sm font-semibold text-primary",
        labelClass: "text-xs font-medium text-tertiary",
        halfCircleTextPosition: "absolute bottom-1 text-center",
    },
    md: {
        strokeWidth: 24,
        radius: 108,
        valueClass: "text-display-md font-semibold text-primary",
        labelClass: "text-sm font-medium text-tertiary",
        halfCircleTextPosition: "absolute bottom-1 text-center",
    },
    lg: {
        strokeWidth: 28,
        radius: 126,
        valueClass: "text-display-lg font-semibold text-primary",
        labelClass: "text-sm font-medium text-tertiary",
        halfCircleTextPosition: "absolute bottom-0 text-center",
    },
});
const ProgressBarCircle = ({ value, min = 0, max = 100, size, label, valueFormatter }) => {
    const percentage = Math.round(((value - min) * 100) / (max - min));
    const sizeConfig = sizes[size];
    const { strokeWidth, radius, valueClass, labelClass } = sizeConfig;
    const diameter = 2 * (radius + strokeWidth / 2);
    const width = diameter;
    const height = diameter;
    const viewBox = `0 0 ${width} ${height}`;
    const cx$1 = diameter / 2;
    const cy = diameter / 2;
    const textPosition = label ? "absolute text-center" : "absolute text-primary";
    const strokeDashoffset = 100 - percentage;
    return (jsxRuntime.jsxs("div", { className: "flex flex-col items-center gap-0.5", children: [jsxRuntime.jsxs("div", { role: "progressbar", "aria-valuenow": value, "aria-valuemin": min, "aria-valuemax": max, className: "relative flex w-max items-center justify-center", children: [jsxRuntime.jsxs("svg", { className: "-rotate-90", width: width, height: height, viewBox: viewBox, children: [jsxRuntime.jsx("circle", { className: "stroke-bg-quaternary", cx: cx$1, cy: cy, r: radius, fill: "none", strokeWidth: strokeWidth, pathLength: "100", strokeDasharray: "100", strokeLinecap: "round" }), jsxRuntime.jsx("circle", { className: "stroke-fg-brand-primary", cx: cx$1, cy: cy, r: radius, fill: "none", strokeWidth: strokeWidth, pathLength: "100", strokeDasharray: "100", strokeLinecap: "round", strokeDashoffset: strokeDashoffset })] }), label && size !== "xxs" ? (jsxRuntime.jsxs("div", { className: "absolute text-center", children: [jsxRuntime.jsx("div", { className: labelClass, children: label }), jsxRuntime.jsx("div", { className: valueClass, children: valueFormatter ? valueFormatter(value, percentage) : `${percentage}%` })] })) : (jsxRuntime.jsx("span", { className: cx.cx(textPosition, valueClass), children: valueFormatter ? valueFormatter(value, percentage) : `${percentage}%` }))] }), label && size === "xxs" && jsxRuntime.jsx("div", { className: labelClass, children: label })] }));
};
const ProgressBarHalfCircle = ({ value, min = 0, max = 100, size, label, valueFormatter }) => {
    const percentage = Math.round(((value - min) * 100) / (max - min));
    const sizeConfig = sizes[size];
    const { strokeWidth, radius, valueClass, labelClass, halfCircleTextPosition } = sizeConfig;
    const width = 2 * (radius + strokeWidth / 2);
    const height = radius + strokeWidth;
    const viewBox = `0 0 ${width} ${height}`;
    const cx$1 = "50%";
    const cy = radius + strokeWidth / 2;
    const strokeDashoffset = -50 - (100 - percentage) / 2;
    return (jsxRuntime.jsxs("div", { className: "flex flex-col items-center gap-0.5", children: [jsxRuntime.jsxs("div", { role: "progressbar", "aria-valuenow": value, "aria-valuemin": min, "aria-valuemax": max, className: "relative flex w-max items-center justify-center", children: [jsxRuntime.jsxs("svg", { width: width, height: height, viewBox: viewBox, children: [jsxRuntime.jsx("circle", { className: "stroke-bg-quaternary", cx: cx$1, cy: cy, r: radius, fill: "none", strokeWidth: strokeWidth, pathLength: "100", strokeDasharray: "100", strokeDashoffset: "-50", strokeLinecap: "round" }), jsxRuntime.jsx("circle", { className: "origin-center -scale-x-100 stroke-fg-brand-primary", cx: cx$1, cy: cy, r: radius, fill: "none", strokeWidth: strokeWidth, pathLength: "100", strokeDasharray: "100", strokeDashoffset: strokeDashoffset, strokeLinecap: "round" })] }), label && size !== "xxs" ? (jsxRuntime.jsxs("div", { className: halfCircleTextPosition, children: [jsxRuntime.jsx("div", { className: labelClass, children: label }), jsxRuntime.jsx("div", { className: valueClass, children: valueFormatter ? valueFormatter(value, percentage) : `${percentage}%` })] })) : (jsxRuntime.jsx("span", { className: cx.cx(halfCircleTextPosition, valueClass), children: valueFormatter ? valueFormatter(value, percentage) : `${percentage}%` }))] }), label && size === "xxs" && jsxRuntime.jsx("div", { className: labelClass, children: label })] }));
};

function ProgressCircle({ isHalfCircle = false, ...props }) {
    if (isHalfCircle) {
        return jsxRuntime.jsx(ProgressBarHalfCircle, { ...props });
    }
    return jsxRuntime.jsx(ProgressBarCircle, { ...props });
}

exports.ProgressBar = ProgressBarIndicator.ProgressBar;
exports.ProgressCircle = ProgressCircle;
