// @hostaway/design-system/dist/components/ProgressIndicator/index.esm.js
export { P as ProgressBar } from '../_shared/ProgressBarIndicator-quOF1G__.esm.js';
import { jsxs, jsx } from 'react/jsx-runtime';
import { c as cx, s as sortCx } from '../_shared/cx-DnIR8RkH.esm.js';
import '../_shared/progress-indicators-OJSyQ_qE.esm.js';
import 'tailwind-merge';

const sizes = sortCx({
    xxs: {
        strokeWidth: 6,
        radius: 29,
        valueClass: "text-sm font-semibold text-primary",
        labelClass: "text-xs font-medium text-tertiary",
        halfCircleTextPosition: "absolute bottom-0.5 text-center",
    },
    xs: {
        strokeWidth: 16,
        radius: 72,
        valueClass: "text-display-xs font-semibold text-primary",
        labelClass: "text-xs font-medium text-tertiary",
        halfCircleTextPosition: "absolute bottom-0.5 text-center",
    },
    sm: {
        strokeWidth: 20,
        radius: 90,
        valueClass: "text-display-sm font-semibold text-primary",
        labelClass: "text-xs font-medium text-tertiary",
        halfCircleTextPosition: "absolute bottom-1 text-center",
    },
    md: {
        strokeWidth: 24,
        radius: 108,
        valueClass: "text-display-md font-semibold text-primary",
        labelClass: "text-sm font-medium text-tertiary",
        halfCircleTextPosition: "absolute bottom-1 text-center",
    },
    lg: {
        strokeWidth: 28,
        radius: 126,
        valueClass: "text-display-lg font-semibold text-primary",
        labelClass: "text-sm font-medium text-tertiary",
        halfCircleTextPosition: "absolute bottom-0 text-center",
    },
});
const ProgressBarCircle = ({ value, min = 0, max = 100, size, label, valueFormatter }) => {
    const percentage = Math.round(((value - min) * 100) / (max - min));
    const sizeConfig = sizes[size];
    const { strokeWidth, radius, valueClass, labelClass } = sizeConfig;
    const diameter = 2 * (radius + strokeWidth / 2);
    const width = diameter;
    const height = diameter;
    const viewBox = `0 0 ${width} ${height}`;
    const cx$1 = diameter / 2;
    const cy = diameter / 2;
    const textPosition = label ? "absolute text-center" : "absolute text-primary";
    const strokeDashoffset = 100 - percentage;
    return (jsxs("div", { className: "flex flex-col items-center gap-0.5", children: [jsxs("div", { role: "progressbar", "aria-valuenow": value, "aria-valuemin": min, "aria-valuemax": max, className: "relative flex w-max items-center justify-center", children: [jsxs("svg", { className: "-rotate-90", width: width, height: height, viewBox: viewBox, children: [jsx("circle", { className: "stroke-bg-quaternary", cx: cx$1, cy: cy, r: radius, fill: "none", strokeWidth: strokeWidth, pathLength: "100", strokeDasharray: "100", strokeLinecap: "round" }), jsx("circle", { className: "stroke-fg-brand-primary", cx: cx$1, cy: cy, r: radius, fill: "none", strokeWidth: strokeWidth, pathLength: "100", strokeDasharray: "100", strokeLinecap: "round", strokeDashoffset: strokeDashoffset })] }), label && size !== "xxs" ? (jsxs("div", { className: "absolute text-center", children: [jsx("div", { className: labelClass, children: label }), jsx("div", { className: valueClass, children: valueFormatter ? valueFormatter(value, percentage) : `${percentage}%` })] })) : (jsx("span", { className: cx(textPosition, valueClass), children: valueFormatter ? valueFormatter(value, percentage) : `${percentage}%` }))] }), label && size === "xxs" && jsx("div", { className: labelClass, children: label })] }));
};
const ProgressBarHalfCircle = ({ value, min = 0, max = 100, size, label, valueFormatter }) => {
    const percentage = Math.round(((value - min) * 100) / (max - min));
    const sizeConfig = sizes[size];
    const { strokeWidth, radius, valueClass, labelClass, halfCircleTextPosition } = sizeConfig;
    const width = 2 * (radius + strokeWidth / 2);
    const height = radius + strokeWidth;
    const viewBox = `0 0 ${width} ${height}`;
    const cx$1 = "50%";
    const cy = radius + strokeWidth / 2;
    const strokeDashoffset = -50 - (100 - percentage) / 2;
    return (jsxs("div", { className: "flex flex-col items-center gap-0.5", children: [jsxs("div", { role: "progressbar", "aria-valuenow": value, "aria-valuemin": min, "aria-valuemax": max, className: "relative flex w-max items-center justify-center", children: [jsxs("svg", { width: width, height: height, viewBox: viewBox, children: [jsx("circle", { className: "stroke-bg-quaternary", cx: cx$1, cy: cy, r: radius, fill: "none", strokeWidth: strokeWidth, pathLength: "100", strokeDasharray: "100", strokeDashoffset: "-50", strokeLinecap: "round" }), jsx("circle", { className: "origin-center -scale-x-100 stroke-fg-brand-primary", cx: cx$1, cy: cy, r: radius, fill: "none", strokeWidth: strokeWidth, pathLength: "100", strokeDasharray: "100", strokeDashoffset: strokeDashoffset, strokeLinecap: "round" })] }), label && size !== "xxs" ? (jsxs("div", { className: halfCircleTextPosition, children: [jsx("div", { className: labelClass, children: label }), jsx("div", { className: valueClass, children: valueFormatter ? valueFormatter(value, percentage) : `${percentage}%` })] })) : (jsx("span", { className: cx(halfCircleTextPosition, valueClass), children: valueFormatter ? valueFormatter(value, percentage) : `${percentage}%` }))] }), label && size === "xxs" && jsx("div", { className: labelClass, children: label })] }));
};

function ProgressCircle({ isHalfCircle = false, ...props }) {
    if (isHalfCircle) {
        return jsx(ProgressBarHalfCircle, { ...props });
    }
    return jsx(ProgressBarCircle, { ...props });
}

export { ProgressCircle };
