// @hostaway/design-system/dist/components/DatePicker/index.cjs.js
'use strict';

var jsxRuntime = require('react/jsx-runtime');
var h = require('react');
var reactAriaComponents = require('react-aria-components');
var DatePicker = require('../_shared/DatePicker-CZZjYER_.cjs.js');
var useControlledState = require('../_shared/useControlledState-B1fb2m75.cjs.js');
var icons = require('@untitledui/icons');
var cx = require('../_shared/cx-CmkbyMOt.cjs.js');
var BaseButton = require('../_shared/BaseButton-CchEmlms.cjs.js');
var useBreakpoint = require('../_shared/use-breakpoint-DnYPO9hT.cjs.js');
var FieldWrapper = require('../_shared/FieldWrapper-D6dXzMyx.cjs.js');
var Button = require('../_shared/Button-BE28ft73.cjs.js');
require('../_shared/context-Bh6OwrvA.cjs.js');
require('tailwind-merge');
require('../_shared/is-react-component-DkG20B86.cjs.js');
require('../_shared/tooltip-DDB8niG0.cjs.js');
require('../_shared/hint-text-LRIVuQnk.cjs.js');
require('../_shared/ConditionalTooltip-FLw9yEoi.cjs.js');

function Calendar({ value, defaultValue, onChange, ...rest }) {
    var _a;
    const [focusedValue, setFocusedValue] = h.useState((_a = value !== null && value !== void 0 ? value : defaultValue) !== null && _a !== void 0 ? _a : undefined);
    h.useEffect(() => {
        setFocusedValue(value !== null && value !== void 0 ? value : undefined);
    }, [value]);
    return (jsxRuntime.jsx(reactAriaComponents.DatePicker, { "aria-label": "calendar", value: value, defaultValue: defaultValue, onChange: (date) => {
            if (date) {
                onChange === null || onChange === void 0 ? void 0 : onChange(date);
                setFocusedValue(date);
            }
        }, children: jsxRuntime.jsx(DatePicker.Calendar, { ...rest, focusedValue: focusedValue, onFocusChange: setFocusedValue }) }));
}

const RangeCalendarContextProvider = ({ children }) => {
    const [value, onChange] = h.useState(null);
    const [focusedValue, onFocusChange] = h.useState();
    return jsxRuntime.jsx(reactAriaComponents.RangeCalendarContext.Provider, { value: { value, onChange, focusedValue, onFocusChange }, children: children });
};
const RangeCalendarTitle = ({ part }) => {
    const context = h.useContext(reactAriaComponents.RangeCalendarStateContext);
    if (!context) {
        throw new Error("<RangeCalendarTitle /> must be used within a <RangeCalendar /> component.");
    }
    const formatter = DatePicker.$896ba0a80a8f4d36$export$85fd5fdf27bacc79({
        month: "long",
        year: "numeric",
        calendar: context.visibleRange.start.calendar.identifier,
        timeZone: context.timeZone,
    });
    return part === "start"
        ? formatter.format(context.visibleRange.start.toDate(context.timeZone))
        : formatter.format(context.visibleRange.end.toDate(context.timeZone));
};
const MobilePresetButton = ({ value, children, ...props }) => {
    const context = h.useContext(reactAriaComponents.RangeCalendarStateContext);
    return (jsxRuntime.jsx(BaseButton.Button, { ...props, slot: null, size: "sm", color: "link-color", onClick: () => {
            context === null || context === void 0 ? void 0 : context.setValue(value);
            context === null || context === void 0 ? void 0 : context.setFocusedDate(value.start);
        }, children: children }));
};
const RangeCalendar = ({ presets, ...props }) => {
    const isDesktop = useBreakpoint.useBreakpoint("md");
    const context = reactAriaComponents.useSlottedContext(reactAriaComponents.RangeCalendarContext);
    const ContextWrapper = context ? h.Fragment : RangeCalendarContextProvider;
    return (jsxRuntime.jsx(ContextWrapper, { children: jsxRuntime.jsxs(reactAriaComponents.RangeCalendar, { className: "flex items-start", visibleDuration: {
                months: isDesktop ? 2 : 1,
            }, ...props, children: [jsxRuntime.jsxs("div", { className: "flex flex-col gap-3 px-6 py-5", children: [jsxRuntime.jsxs("header", { className: "relative flex items-center justify-between md:justify-start", children: [jsxRuntime.jsx(BaseButton.Button, { slot: "previous", iconLeading: icons.ChevronLeft, size: "sm", color: "tertiary", className: "size-8" }), jsxRuntime.jsx("h2", { className: "absolute top-1/2 left-1/2 -translate-x-1/2 -translate-y-1/2 text-sm font-semibold text-fg-secondary", children: jsxRuntime.jsx(RangeCalendarTitle, { part: "start" }) }), jsxRuntime.jsx(BaseButton.Button, { slot: "next", iconLeading: icons.ChevronRight, size: "sm", color: "tertiary", className: "size-8 md:hidden" })] }), !isDesktop && (jsxRuntime.jsxs("div", { className: "flex items-center gap-2 md:hidden", children: [jsxRuntime.jsx(DatePicker.DateInput, { slot: "start", className: "flex-1" }), jsxRuntime.jsx("div", { className: "text-md text-quaternary", children: "\u2013" }), jsxRuntime.jsx(DatePicker.DateInput, { slot: "end", className: "flex-1" })] })), !isDesktop && presets && (jsxRuntime.jsx("div", { className: "mt-1 flex justify-between gap-3 px-2 md:hidden", children: Object.values(presets).map((preset) => (jsxRuntime.jsx(MobilePresetButton, { value: preset.value, children: preset.label }, preset.label))) })), jsxRuntime.jsxs(reactAriaComponents.CalendarGrid, { weekdayStyle: "short", className: "w-max", children: [jsxRuntime.jsx(reactAriaComponents.CalendarGridHeader, { children: (day) => (jsxRuntime.jsx(reactAriaComponents.CalendarHeaderCell, { className: "border-b-4 border-transparent p-0", children: jsxRuntime.jsx("div", { className: "flex size-10 items-center justify-center text-sm font-medium text-secondary", children: day.slice(0, 2) }) })) }), jsxRuntime.jsx(reactAriaComponents.CalendarGridBody, { className: "[&_td]:p-0 [&_tr]:border-b-4 [&_tr]:border-transparent [&_tr:last-of-type]:border-none", children: (date) => jsxRuntime.jsx(DatePicker.CalendarCell, { date: date }) })] })] }), isDesktop && (jsxRuntime.jsxs("div", { className: "flex flex-col gap-3 border-l border-secondary px-6 py-5", children: [jsxRuntime.jsxs("header", { className: "relative flex items-center justify-end", children: [jsxRuntime.jsx("h2", { className: "absolute top-1/2 left-1/2 -translate-x-1/2 -translate-y-1/2 text-sm font-semibold text-fg-secondary", children: jsxRuntime.jsx(RangeCalendarTitle, { part: "end" }) }), jsxRuntime.jsx(BaseButton.Button, { slot: "next", iconLeading: icons.ChevronRight, size: "sm", color: "tertiary", className: "size-8" })] }), jsxRuntime.jsxs(reactAriaComponents.CalendarGrid, { weekdayStyle: "short", offset: { months: 1 }, className: "w-max", children: [jsxRuntime.jsx(reactAriaComponents.CalendarGridHeader, { children: (day) => (jsxRuntime.jsx(reactAriaComponents.CalendarHeaderCell, { className: "border-b-4 border-transparent p-0", children: jsxRuntime.jsx("div", { className: "flex size-10 items-center justify-center text-sm font-medium text-secondary", children: day.slice(0, 2) }) })) }), jsxRuntime.jsx(reactAriaComponents.CalendarGridBody, { className: "[&_td]:p-0 [&_tr]:border-b-4 [&_tr]:border-transparent [&_tr:last-of-type]:border-none", children: (date) => jsxRuntime.jsx(DatePicker.CalendarCell, { date: date }) })] })] }))] }) }));
};

const RangePresetButton = ({ value, className, children, ...props }) => {
    var _a, _b, _c, _d;
    const context = reactAriaComponents.useSlottedContext(reactAriaComponents.RangeCalendarContext);
    const isSelected = ((_b = (_a = context === null || context === void 0 ? void 0 : context.value) === null || _a === void 0 ? void 0 : _a.start) === null || _b === void 0 ? void 0 : _b.compare(value.start)) === 0 && ((_d = (_c = context === null || context === void 0 ? void 0 : context.value) === null || _c === void 0 ? void 0 : _c.end) === null || _d === void 0 ? void 0 : _d.compare(value.end)) === 0;
    return (jsxRuntime.jsx("button", { ...props, className: cx.cx("cursor-pointer rounded-md px-3 py-2 text-left text-sm font-medium outline-focus-ring transition duration-100 ease-linear focus-visible:outline-2 focus-visible:outline-offset-2", isSelected ? "bg-active text-secondary_hover hover:bg-secondary_hover" : "text-secondary hover:bg-primary_hover hover:text-secondary_hover", className), children: children }));
};

const DateRangePicker = ({ value: valueProp, defaultValue, onChange, onApply, onCancel, formatDate, t: textsProp, timeZone, 'data-id': dataId, isInvalid, isRequired, label, hint, labelPosition = 'left', tooltip, description, id, ...props }) => {
    const texts = {
        ...(textsProp !== null && textsProp !== void 0 ? textsProp : DatePicker.DEFAULT_TEXTS),
    };
    const tz = timeZone !== null && timeZone !== void 0 ? timeZone : DatePicker.$14e0f24ef4ac5c92$export$aa8b41735afcabd2();
    const { locale } = reactAriaComponents.useLocale();
    const formatter = DatePicker.$896ba0a80a8f4d36$export$85fd5fdf27bacc79({
        month: 'short',
        day: 'numeric',
        year: 'numeric',
    });
    const [value, setValue] = useControlledState.$458b0a5536c1a7cf$export$40bfa8c7b0832715(valueProp, defaultValue || null, onChange);
    const [focusedValue, setFocusedValue] = h.useState(null);
    const now = h.useMemo(() => DatePicker.$14e0f24ef4ac5c92$export$d0bdf45af03a6ea3(tz), [tz]);
    const highlightedDates = h.useMemo(() => [DatePicker.$14e0f24ef4ac5c92$export$d0bdf45af03a6ea3(tz)], [tz]);
    const formatDisplayDate = (date) => {
        const jsDate = date.toDate(tz);
        return formatDate ? formatDate(jsDate) : formatter.format(jsDate);
    };
    const formattedStartDate = (value === null || value === void 0 ? void 0 : value.start)
        ? formatDisplayDate(value.start)
        : texts.placeholder;
    const formattedEndDate = (value === null || value === void 0 ? void 0 : value.end)
        ? formatDisplayDate(value.end)
        : texts.placeholder;
    const presets = h.useMemo(() => ({
        today: { label: texts.presets.today, value: { start: now, end: now } },
        yesterday: {
            label: texts.presets.yesterday,
            value: {
                start: now.subtract({ days: 1 }),
                end: now.subtract({ days: 1 }),
            },
        },
        thisWeek: {
            label: texts.presets.thisWeek,
            value: { start: DatePicker.$14e0f24ef4ac5c92$export$42c81a444fbfb5d4(now, locale), end: DatePicker.$14e0f24ef4ac5c92$export$ef8b6d9133084f4e(now, locale) },
        },
        lastWeek: {
            label: texts.presets.lastWeek,
            value: {
                start: DatePicker.$14e0f24ef4ac5c92$export$42c81a444fbfb5d4(now, locale).subtract({ weeks: 1 }),
                end: DatePicker.$14e0f24ef4ac5c92$export$ef8b6d9133084f4e(now, locale).subtract({ weeks: 1 }),
            },
        },
        thisMonth: {
            label: texts.presets.thisMonth,
            value: { start: DatePicker.$14e0f24ef4ac5c92$export$a5a3b454ada2268e(now), end: DatePicker.$14e0f24ef4ac5c92$export$a2258d9c4118825c(now) },
        },
        lastMonth: {
            label: texts.presets.lastMonth,
            value: {
                start: DatePicker.$14e0f24ef4ac5c92$export$a5a3b454ada2268e(now).subtract({ months: 1 }),
                end: DatePicker.$14e0f24ef4ac5c92$export$a2258d9c4118825c(now).subtract({ months: 1 }),
            },
        },
        thisYear: {
            label: texts.presets.thisYear,
            value: {
                start: DatePicker.$14e0f24ef4ac5c92$export$a5a3b454ada2268e(now.set({ month: 1 })),
                end: DatePicker.$14e0f24ef4ac5c92$export$a2258d9c4118825c(now.set({ month: 12 })),
            },
        },
        lastYear: {
            label: texts.presets.lastYear,
            value: {
                start: DatePicker.$14e0f24ef4ac5c92$export$a5a3b454ada2268e(now.set({ month: 1 }).subtract({ years: 1 })),
                end: DatePicker.$14e0f24ef4ac5c92$export$a2258d9c4118825c(now.set({ month: 12 }).subtract({ years: 1 })),
            },
        },
        allTime: {
            label: texts.presets.allTime,
            value: {
                start: now.set({ year: 2000, month: 1, day: 1 }),
                end: now,
            },
        },
    }), [locale, now, texts.presets]);
    return (jsxRuntime.jsx(FieldWrapper.FieldWrapper, { isRequired: isRequired, label: label, hint: hint, isInvalid: isInvalid, labelPosition: labelPosition, tooltip: tooltip, description: description, id: id, children: jsxRuntime.jsxs(reactAriaComponents.DateRangePicker, { "aria-label": "Date range picker", shouldCloseOnSelect: false, ...props, value: value, onChange: setValue, children: [jsxRuntime.jsx(reactAriaComponents.Group, { children: jsxRuntime.jsx(Button.Button, { size: "sm", color: "secondary", iconLeading: icons.Calendar, className: cx.cx(isInvalid && 'ring-error_subtle'), "data-id": `${dataId}-calendar-icon-button`, children: !value ? (jsxRuntime.jsx("span", { className: "text-placeholder", children: texts.placeholderRange })) : (`${formattedStartDate} – ${formattedEndDate}`) }) }), jsxRuntime.jsx(reactAriaComponents.Popover, { placement: "bottom right", offset: 8, className: ({ isEntering, isExiting }) => cx.cx('will-change-transform', isEntering &&
                        'duration-150 ease-out animate-in fade-in placement-right:origin-left placement-right:slide-in-from-left-0.5 placement-top:origin-bottom placement-top:slide-in-from-bottom-0.5 placement-bottom:origin-top placement-bottom:slide-in-from-top-0.5', isExiting &&
                        'duration-100 ease-in animate-out fade-out placement-right:origin-left placement-right:slide-out-to-left-0.5 placement-top:origin-bottom placement-top:slide-out-to-bottom-0.5 placement-bottom:origin-top placement-bottom:slide-out-to-top-0.5'), children: jsxRuntime.jsx(reactAriaComponents.Dialog, { className: "flex rounded-2xl bg-primary shadow-xl ring ring-secondary_alt focus:outline-hidden", children: ({ close }) => (jsxRuntime.jsxs(jsxRuntime.Fragment, { children: [jsxRuntime.jsx("div", { className: "hidden w-38 flex-col gap-0.5 border-r border-solid border-secondary p-3 lg:flex", children: Object.values(presets).map((preset) => (jsxRuntime.jsx(RangePresetButton, { value: preset.value, onClick: () => {
                                            setValue(preset.value);
                                            setFocusedValue(preset.value.start);
                                        }, "data-id": `${dataId}-range-preset-button`, children: preset.label }, preset.label))) }), jsxRuntime.jsxs("div", { className: "flex flex-col", children: [jsxRuntime.jsx(RangeCalendar, { focusedValue: focusedValue, onFocusChange: setFocusedValue, highlightedDates: highlightedDates, presets: {
                                                lastWeek: presets.lastWeek,
                                                lastMonth: presets.lastMonth,
                                                lastYear: presets.lastYear,
                                            }, "data-id": `${dataId}-range-calendar` }), jsxRuntime.jsxs("div", { className: "flex justify-between gap-3 border-t border-secondary p-4", children: [jsxRuntime.jsxs("div", { className: "hidden items-center gap-3 md:flex", children: [jsxRuntime.jsx(DatePicker.DateInput, { slot: "start", className: "w-36", "data-id": `${dataId}-date-start-input` }), jsxRuntime.jsx("div", { className: "text-md text-quaternary", children: "\u2013" }), jsxRuntime.jsx(DatePicker.DateInput, { slot: "end", className: "w-36", "data-id": `${dataId}-date-end-input` })] }), jsxRuntime.jsxs("div", { className: "grid w-full grid-cols-2 gap-3 md:flex md:w-auto", children: [jsxRuntime.jsx(Button.Button, { size: "sm", color: "secondary", onClick: () => {
                                                                onCancel === null || onCancel === void 0 ? void 0 : onCancel();
                                                                close();
                                                            }, "data-id": `${dataId}-cancel-button`, children: texts.cancelBtn }), jsxRuntime.jsx(Button.Button, { size: "sm", color: "primary", onClick: () => {
                                                                onApply === null || onApply === void 0 ? void 0 : onApply();
                                                                close();
                                                            }, "data-id": `${dataId}-apply-button`, children: texts.applyBtn })] })] })] })] })) }) })] }) }));
};

exports.DatePicker = DatePicker.DatePicker;
exports.parseDate = DatePicker.$fae977aafc393c5c$export$6b862160d295c8e;
exports.Calendar = Calendar;
exports.DateRangePicker = DateRangePicker;
