// @hostaway/design-system/dist/components/DatePicker/index.esm.js
import { jsx, jsxs, Fragment as Fragment$1 } from 'react/jsx-runtime';
import { useState, useEffect, Fragment, useContext, useMemo } from 'react';
import { DatePicker, useSlottedContext, RangeCalendarContext, RangeCalendar as RangeCalendar$1, CalendarGrid, CalendarGridHeader, CalendarHeaderCell, CalendarGridBody, RangeCalendarStateContext, useLocale, DateRangePicker as DateRangePicker$1, Group, Popover, Dialog } from 'react-aria-components';
import { C as Calendar$1, a as DateInput, b as CalendarCell, $ as $896ba0a80a8f4d36$export$85fd5fdf27bacc79, c as DEFAULT_TEXTS, d as $14e0f24ef4ac5c92$export$aa8b41735afcabd2, e as $14e0f24ef4ac5c92$export$d0bdf45af03a6ea3, f as $14e0f24ef4ac5c92$export$a2258d9c4118825c, g as $14e0f24ef4ac5c92$export$a5a3b454ada2268e, h as $14e0f24ef4ac5c92$export$ef8b6d9133084f4e, i as $14e0f24ef4ac5c92$export$42c81a444fbfb5d4 } from '../_shared/DatePicker-FwY9F1pc.esm.js';
export { D as DatePicker, j as parseDate } from '../_shared/DatePicker-FwY9F1pc.esm.js';
import { $ as $458b0a5536c1a7cf$export$40bfa8c7b0832715 } from '../_shared/useControlledState-mV1mviQ6.esm.js';
import { ChevronLeft, ChevronRight, Calendar as Calendar$2 } from '@untitledui/icons';
import { c as cx } from '../_shared/cx-DnIR8RkH.esm.js';
import { B as Button } from '../_shared/BaseButton-CTtHjEZn.esm.js';
import { u as useBreakpoint } from '../_shared/use-breakpoint-D8xwrq2z.esm.js';
import { F as FieldWrapper } from '../_shared/FieldWrapper-UtX7P_J-.esm.js';
import { B as Button$1 } from '../_shared/Button-C3YUMQZU.esm.js';
import '../_shared/context-DUiWIuVv.esm.js';
import 'tailwind-merge';
import '../_shared/is-react-component-CYKUd65Q.esm.js';
import '../_shared/tooltip-l8auK4sw.esm.js';
import '../_shared/hint-text-Ck8Vj_nN.esm.js';
import '../_shared/ConditionalTooltip-DT2TnX8y.esm.js';

function Calendar({ value, defaultValue, onChange, ...rest }) {
    var _a;
    const [focusedValue, setFocusedValue] = useState((_a = value !== null && value !== void 0 ? value : defaultValue) !== null && _a !== void 0 ? _a : undefined);
    useEffect(() => {
        setFocusedValue(value !== null && value !== void 0 ? value : undefined);
    }, [value]);
    return (jsx(DatePicker, { "aria-label": "calendar", value: value, defaultValue: defaultValue, onChange: (date) => {
            if (date) {
                onChange === null || onChange === void 0 ? void 0 : onChange(date);
                setFocusedValue(date);
            }
        }, children: jsx(Calendar$1, { ...rest, focusedValue: focusedValue, onFocusChange: setFocusedValue }) }));
}

const RangeCalendarContextProvider = ({ children }) => {
    const [value, onChange] = useState(null);
    const [focusedValue, onFocusChange] = useState();
    return jsx(RangeCalendarContext.Provider, { value: { value, onChange, focusedValue, onFocusChange }, children: children });
};
const RangeCalendarTitle = ({ part }) => {
    const context = useContext(RangeCalendarStateContext);
    if (!context) {
        throw new Error("<RangeCalendarTitle /> must be used within a <RangeCalendar /> component.");
    }
    const formatter = $896ba0a80a8f4d36$export$85fd5fdf27bacc79({
        month: "long",
        year: "numeric",
        calendar: context.visibleRange.start.calendar.identifier,
        timeZone: context.timeZone,
    });
    return part === "start"
        ? formatter.format(context.visibleRange.start.toDate(context.timeZone))
        : formatter.format(context.visibleRange.end.toDate(context.timeZone));
};
const MobilePresetButton = ({ value, children, ...props }) => {
    const context = useContext(RangeCalendarStateContext);
    return (jsx(Button, { ...props, slot: null, size: "sm", color: "link-color", onClick: () => {
            context === null || context === void 0 ? void 0 : context.setValue(value);
            context === null || context === void 0 ? void 0 : context.setFocusedDate(value.start);
        }, children: children }));
};
const RangeCalendar = ({ presets, ...props }) => {
    const isDesktop = useBreakpoint("md");
    const context = useSlottedContext(RangeCalendarContext);
    const ContextWrapper = context ? Fragment : RangeCalendarContextProvider;
    return (jsx(ContextWrapper, { children: jsxs(RangeCalendar$1, { className: "flex items-start", visibleDuration: {
                months: isDesktop ? 2 : 1,
            }, ...props, children: [jsxs("div", { className: "flex flex-col gap-3 px-6 py-5", children: [jsxs("header", { className: "relative flex items-center justify-between md:justify-start", children: [jsx(Button, { slot: "previous", iconLeading: ChevronLeft, size: "sm", color: "tertiary", className: "size-8" }), jsx("h2", { className: "absolute top-1/2 left-1/2 -translate-x-1/2 -translate-y-1/2 text-sm font-semibold text-fg-secondary", children: jsx(RangeCalendarTitle, { part: "start" }) }), jsx(Button, { slot: "next", iconLeading: ChevronRight, size: "sm", color: "tertiary", className: "size-8 md:hidden" })] }), !isDesktop && (jsxs("div", { className: "flex items-center gap-2 md:hidden", children: [jsx(DateInput, { slot: "start", className: "flex-1" }), jsx("div", { className: "text-md text-quaternary", children: "\u2013" }), jsx(DateInput, { slot: "end", className: "flex-1" })] })), !isDesktop && presets && (jsx("div", { className: "mt-1 flex justify-between gap-3 px-2 md:hidden", children: Object.values(presets).map((preset) => (jsx(MobilePresetButton, { value: preset.value, children: preset.label }, preset.label))) })), jsxs(CalendarGrid, { weekdayStyle: "short", className: "w-max", children: [jsx(CalendarGridHeader, { children: (day) => (jsx(CalendarHeaderCell, { className: "border-b-4 border-transparent p-0", children: jsx("div", { className: "flex size-10 items-center justify-center text-sm font-medium text-secondary", children: day.slice(0, 2) }) })) }), jsx(CalendarGridBody, { className: "[&_td]:p-0 [&_tr]:border-b-4 [&_tr]:border-transparent [&_tr:last-of-type]:border-none", children: (date) => jsx(CalendarCell, { date: date }) })] })] }), isDesktop && (jsxs("div", { className: "flex flex-col gap-3 border-l border-secondary px-6 py-5", children: [jsxs("header", { className: "relative flex items-center justify-end", children: [jsx("h2", { className: "absolute top-1/2 left-1/2 -translate-x-1/2 -translate-y-1/2 text-sm font-semibold text-fg-secondary", children: jsx(RangeCalendarTitle, { part: "end" }) }), jsx(Button, { slot: "next", iconLeading: ChevronRight, size: "sm", color: "tertiary", className: "size-8" })] }), jsxs(CalendarGrid, { weekdayStyle: "short", offset: { months: 1 }, className: "w-max", children: [jsx(CalendarGridHeader, { children: (day) => (jsx(CalendarHeaderCell, { className: "border-b-4 border-transparent p-0", children: jsx("div", { className: "flex size-10 items-center justify-center text-sm font-medium text-secondary", children: day.slice(0, 2) }) })) }), jsx(CalendarGridBody, { className: "[&_td]:p-0 [&_tr]:border-b-4 [&_tr]:border-transparent [&_tr:last-of-type]:border-none", children: (date) => jsx(CalendarCell, { date: date }) })] })] }))] }) }));
};

const RangePresetButton = ({ value, className, children, ...props }) => {
    var _a, _b, _c, _d;
    const context = useSlottedContext(RangeCalendarContext);
    const isSelected = ((_b = (_a = context === null || context === void 0 ? void 0 : context.value) === null || _a === void 0 ? void 0 : _a.start) === null || _b === void 0 ? void 0 : _b.compare(value.start)) === 0 && ((_d = (_c = context === null || context === void 0 ? void 0 : context.value) === null || _c === void 0 ? void 0 : _c.end) === null || _d === void 0 ? void 0 : _d.compare(value.end)) === 0;
    return (jsx("button", { ...props, className: cx("cursor-pointer rounded-md px-3 py-2 text-left text-sm font-medium outline-focus-ring transition duration-100 ease-linear focus-visible:outline-2 focus-visible:outline-offset-2", isSelected ? "bg-active text-secondary_hover hover:bg-secondary_hover" : "text-secondary hover:bg-primary_hover hover:text-secondary_hover", className), children: children }));
};

const DateRangePicker = ({ value: valueProp, defaultValue, onChange, onApply, onCancel, formatDate, t: textsProp, timeZone, 'data-id': dataId, isInvalid, isRequired, label, hint, labelPosition = 'left', tooltip, description, id, ...props }) => {
    const texts = {
        ...(textsProp !== null && textsProp !== void 0 ? textsProp : DEFAULT_TEXTS),
    };
    const tz = timeZone !== null && timeZone !== void 0 ? timeZone : $14e0f24ef4ac5c92$export$aa8b41735afcabd2();
    const { locale } = useLocale();
    const formatter = $896ba0a80a8f4d36$export$85fd5fdf27bacc79({
        month: 'short',
        day: 'numeric',
        year: 'numeric',
    });
    const [value, setValue] = $458b0a5536c1a7cf$export$40bfa8c7b0832715(valueProp, defaultValue || null, onChange);
    const [focusedValue, setFocusedValue] = useState(null);
    const now = useMemo(() => $14e0f24ef4ac5c92$export$d0bdf45af03a6ea3(tz), [tz]);
    const highlightedDates = useMemo(() => [$14e0f24ef4ac5c92$export$d0bdf45af03a6ea3(tz)], [tz]);
    const formatDisplayDate = (date) => {
        const jsDate = date.toDate(tz);
        return formatDate ? formatDate(jsDate) : formatter.format(jsDate);
    };
    const formattedStartDate = (value === null || value === void 0 ? void 0 : value.start)
        ? formatDisplayDate(value.start)
        : texts.placeholder;
    const formattedEndDate = (value === null || value === void 0 ? void 0 : value.end)
        ? formatDisplayDate(value.end)
        : texts.placeholder;
    const presets = useMemo(() => ({
        today: { label: texts.presets.today, value: { start: now, end: now } },
        yesterday: {
            label: texts.presets.yesterday,
            value: {
                start: now.subtract({ days: 1 }),
                end: now.subtract({ days: 1 }),
            },
        },
        thisWeek: {
            label: texts.presets.thisWeek,
            value: { start: $14e0f24ef4ac5c92$export$42c81a444fbfb5d4(now, locale), end: $14e0f24ef4ac5c92$export$ef8b6d9133084f4e(now, locale) },
        },
        lastWeek: {
            label: texts.presets.lastWeek,
            value: {
                start: $14e0f24ef4ac5c92$export$42c81a444fbfb5d4(now, locale).subtract({ weeks: 1 }),
                end: $14e0f24ef4ac5c92$export$ef8b6d9133084f4e(now, locale).subtract({ weeks: 1 }),
            },
        },
        thisMonth: {
            label: texts.presets.thisMonth,
            value: { start: $14e0f24ef4ac5c92$export$a5a3b454ada2268e(now), end: $14e0f24ef4ac5c92$export$a2258d9c4118825c(now) },
        },
        lastMonth: {
            label: texts.presets.lastMonth,
            value: {
                start: $14e0f24ef4ac5c92$export$a5a3b454ada2268e(now).subtract({ months: 1 }),
                end: $14e0f24ef4ac5c92$export$a2258d9c4118825c(now).subtract({ months: 1 }),
            },
        },
        thisYear: {
            label: texts.presets.thisYear,
            value: {
                start: $14e0f24ef4ac5c92$export$a5a3b454ada2268e(now.set({ month: 1 })),
                end: $14e0f24ef4ac5c92$export$a2258d9c4118825c(now.set({ month: 12 })),
            },
        },
        lastYear: {
            label: texts.presets.lastYear,
            value: {
                start: $14e0f24ef4ac5c92$export$a5a3b454ada2268e(now.set({ month: 1 }).subtract({ years: 1 })),
                end: $14e0f24ef4ac5c92$export$a2258d9c4118825c(now.set({ month: 12 }).subtract({ years: 1 })),
            },
        },
        allTime: {
            label: texts.presets.allTime,
            value: {
                start: now.set({ year: 2000, month: 1, day: 1 }),
                end: now,
            },
        },
    }), [locale, now, texts.presets]);
    return (jsx(FieldWrapper, { isRequired: isRequired, label: label, hint: hint, isInvalid: isInvalid, labelPosition: labelPosition, tooltip: tooltip, description: description, id: id, children: jsxs(DateRangePicker$1, { "aria-label": "Date range picker", shouldCloseOnSelect: false, ...props, value: value, onChange: setValue, children: [jsx(Group, { children: jsx(Button$1, { size: "sm", color: "secondary", iconLeading: Calendar$2, className: cx(isInvalid && 'ring-error_subtle'), "data-id": `${dataId}-calendar-icon-button`, children: !value ? (jsx("span", { className: "text-placeholder", children: texts.placeholderRange })) : (`${formattedStartDate} – ${formattedEndDate}`) }) }), jsx(Popover, { placement: "bottom right", offset: 8, className: ({ isEntering, isExiting }) => cx('will-change-transform', isEntering &&
                        'duration-150 ease-out animate-in fade-in placement-right:origin-left placement-right:slide-in-from-left-0.5 placement-top:origin-bottom placement-top:slide-in-from-bottom-0.5 placement-bottom:origin-top placement-bottom:slide-in-from-top-0.5', isExiting &&
                        'duration-100 ease-in animate-out fade-out placement-right:origin-left placement-right:slide-out-to-left-0.5 placement-top:origin-bottom placement-top:slide-out-to-bottom-0.5 placement-bottom:origin-top placement-bottom:slide-out-to-top-0.5'), children: jsx(Dialog, { className: "flex rounded-2xl bg-primary shadow-xl ring ring-secondary_alt focus:outline-hidden", children: ({ close }) => (jsxs(Fragment$1, { children: [jsx("div", { className: "hidden w-38 flex-col gap-0.5 border-r border-solid border-secondary p-3 lg:flex", children: Object.values(presets).map((preset) => (jsx(RangePresetButton, { value: preset.value, onClick: () => {
                                            setValue(preset.value);
                                            setFocusedValue(preset.value.start);
                                        }, "data-id": `${dataId}-range-preset-button`, children: preset.label }, preset.label))) }), jsxs("div", { className: "flex flex-col", children: [jsx(RangeCalendar, { focusedValue: focusedValue, onFocusChange: setFocusedValue, highlightedDates: highlightedDates, presets: {
                                                lastWeek: presets.lastWeek,
                                                lastMonth: presets.lastMonth,
                                                lastYear: presets.lastYear,
                                            }, "data-id": `${dataId}-range-calendar` }), jsxs("div", { className: "flex justify-between gap-3 border-t border-secondary p-4", children: [jsxs("div", { className: "hidden items-center gap-3 md:flex", children: [jsx(DateInput, { slot: "start", className: "w-36", "data-id": `${dataId}-date-start-input` }), jsx("div", { className: "text-md text-quaternary", children: "\u2013" }), jsx(DateInput, { slot: "end", className: "w-36", "data-id": `${dataId}-date-end-input` })] }), jsxs("div", { className: "grid w-full grid-cols-2 gap-3 md:flex md:w-auto", children: [jsx(Button$1, { size: "sm", color: "secondary", onClick: () => {
                                                                onCancel === null || onCancel === void 0 ? void 0 : onCancel();
                                                                close();
                                                            }, "data-id": `${dataId}-cancel-button`, children: texts.cancelBtn }), jsx(Button$1, { size: "sm", color: "primary", onClick: () => {
                                                                onApply === null || onApply === void 0 ? void 0 : onApply();
                                                                close();
                                                            }, "data-id": `${dataId}-apply-button`, children: texts.applyBtn })] })] })] })] })) }) })] }) }));
};

export { Calendar, DateRangePicker };
