// @hostaway/design-system/dist/components/PhoneNumberInput/index.cjs.js
'use strict';

var jsxRuntime = require('react/jsx-runtime');
var useControlledState = require('../_shared/useControlledState-B1fb2m75.cjs.js');
var h = require('react');
var reactAriaComponents = require('react-aria-components');
var icons = require('@untitledui/icons');
var PopoverSearchInput = require('../_shared/PopoverSearchInput-BFtG7lQx.cjs.js');
var input = require('../_shared/input-CVSUknGT.cjs.js');
var inputGroup = require('../_shared/input-group-DhYuXcvD.cjs.js');
var selectNative = require('../_shared/select-native-CrEHIRS9.cjs.js');
var cx = require('../_shared/cx-CmkbyMOt.cjs.js');
var FlagIcon = require('../_shared/FlagIcon-BbhvmpcI.cjs.js');
var FieldWrapper = require('../_shared/FieldWrapper-D6dXzMyx.cjs.js');
require('../_shared/MyCustomIcon-DYQF4lJT.cjs.js');
require('../_shared/Icon-BF_rHwZP.cjs.js');
require('../_shared/InputField-DaibKVKA.cjs.js');
require('../_shared/hint-text-LRIVuQnk.cjs.js');
require('../_shared/label-D0WtNYRO.cjs.js');
require('../_shared/tooltip-DDB8niG0.cjs.js');
require('tailwind-merge');
require('../_shared/ConditionalTooltip-FLw9yEoi.cjs.js');

// Helper: country id to emoji flag
function isoToFlag(id) {
    const code = FlagIcon.aliasCountryId(id);
    if (code.length !== 2)
        return '';
    const A = 127397; // regional indicator base
    const chars = code
        .split('')
        .map((c) => String.fromCodePoint(A + c.charCodeAt(0)));
    return chars.join('');
}
const countries = [
    { id: 'us', label: 'United States', code: '+1' },
    { id: 'ca', label: 'Canada', code: '+1' },
    { id: 'uk', label: 'United Kingdom', code: '+44', mask: '.. .... .........' },
    { id: 'de', label: 'Germany', code: '+49' },
    { id: 'fr', label: 'France', code: '+33', mask: '. .. .. .......' },
    { id: 'it', label: 'Italy', code: '+39' },
    { id: 'es', label: 'Spain', code: '+34', mask: '.. ... ... ...' },
    { id: 'nl', label: 'Netherlands', code: '+31' },
    { id: 'be', label: 'Belgium', code: '+32' },
    { id: 'ch', label: 'Switzerland', code: '+41' },
    { id: 'at', label: 'Austria', code: '+43' },
    { id: 'se', label: 'Sweden', code: '+46' },
    { id: 'no', label: 'Norway', code: '+47' },
    { id: 'dk', label: 'Denmark', code: '+45' },
    { id: 'fi', label: 'Finland', code: '+358' },
    { id: 'pl', label: 'Poland', code: '+48' },
    { id: 'ie', label: 'Ireland', code: '+353' },
    { id: 'pt', label: 'Portugal', code: '+351' },
    { id: 'gr', label: 'Greece', code: '+30' },
    { id: 'jp', label: 'Japan', code: '+81' },
    { id: 'au', label: 'Australia', code: '+61' },
    { id: 'nz', label: 'New Zealand', code: '+64' },
    { id: 'br', label: 'Brazil', code: '+55' },
    { id: 'mx', label: 'Mexico', code: '+52' },
    { id: 'ar', label: 'Argentina', code: '+54', mask: '. .. .... ....' },
    { id: 'cl', label: 'Chile', code: '+56' },
    { id: 'co', label: 'Colombia', code: '+57' },
    { id: 'pe', label: 'Peru', code: '+51' },
    { id: 'ru', label: 'Russia', code: '+7' },
    { id: 'ua', label: 'Ukraine', code: '+380' },
    { id: 'tr', label: 'Turkey', code: '+90' },
    { id: 'il', label: 'Israel', code: '+972' },
    { id: 'eg', label: 'Egypt', code: '+20' },
    { id: 'za', label: 'South Africa', code: '+27' },
].map((c) => ({ ...c, flag: isoToFlag(c.id) }));
// Match the longest country code at the start of a value like '+351 ...'
function matchCountryByValue(all, fullValue) {
    if (!fullValue)
        return undefined;
    const v = fullValue.trim();
    if (!v.startsWith('+'))
        return undefined;
    // Sort codes by length desc to avoid partial matches like +3 before +351
    const sorted = [...all].sort((a, b) => b.code.length - a.code.length);
    return sorted.find((c) => v.startsWith(c.code));
}
// Match a country against a search query. Numeric queries match the dial code;
// text queries match label, ISO id, and any searchAliases.
function matchesCountrySearch(country, query) {
    var _a;
    const trimmed = query.trim().toLowerCase().replace(/^\+/, '');
    if (!trimmed)
        return true;
    if (/^\d+$/.test(trimmed)) {
        return country.code.replace(/\+/g, '').includes(trimmed);
    }
    return [country.label, country.id, ...((_a = country.searchAliases) !== null && _a !== void 0 ? _a : [])].some((field) => field.toLowerCase().includes(trimmed));
}
// Format raw digits according to a country mask (`.` = digit slot, other chars are literals).
function applyMask(digits, mask) {
    let out = '';
    let di = 0;
    for (let mi = 0; mi < mask.length && di < digits.length; mi++) {
        if (mask[mi] === '.') {
            out += digits[di++];
        }
        else {
            out += mask[mi];
        }
    }
    // Append any digits beyond the mask length unchanged.
    if (di < digits.length)
        out += digits.slice(di);
    return out;
}
// Ensure value starts with '+' and allow only '+', digits, spaces, dashes, parentheses
function normalizeValue(raw, fallbackDial) {
    let v = raw.replace(/[^\d+\s\-()]/g, '');
    // Keep only the first '+' and ensure it exists
    v = v.replace(/\+/g, '');
    v = `+${v}`;
    // If user typed nothing after '+', seed with selected code
    if (v === '+')
        return fallbackDial + ' ';
    return v;
}
const PhoneNumberInput = ({ label, hint, placeholder, isRequired, isDisabled, isInvalid, isWarning, tooltip, description, id, fieldSize = 'md', labelPosition = 'left', alignLabel, size = 'sm', className, value, defaultValue, onChange, selectedCountry, defaultSelectedCountry = countries[0], onCountryChange, countries: customCountries = countries, countryLabelMode = 'flagName', variant = 'default', searchable = false, texts, ref, 'data-id': dataId, ...rest }) => {
    var _a, _b, _c, _d, _e;
    const searchPlaceholder = (_a = texts === null || texts === void 0 ? void 0 : texts.searchPlaceholder) !== null && _a !== void 0 ? _a : 'Search';
    const searchNotFoundLabel = (_b = texts === null || texts === void 0 ? void 0 : texts.searchNotFound) !== null && _b !== void 0 ? _b : 'No countries found';
    const [searchQuery, setSearchQuery] = h.useState('');
    // Clear the search query when the consumer swaps the country list (e.g.
    // region change). Compared by id-list stringification so inline-array refs
    // from a parent re-render do not spuriously clear the query.
    const countriesKey = customCountries.map((c) => c.id).join('|');
    h.useEffect(() => {
        setSearchQuery('');
    }, [countriesKey]);
    const filteredCountries = h.useMemo(() => {
        if (!searchable || !searchQuery.trim())
            return customCountries;
        return customCountries.filter((c) => matchesCountrySearch(c, searchQuery));
    }, [customCountries, searchQuery, searchable]);
    // Country selection (controlled/uncontrolled)
    const [country, setCountry] = useControlledState.$458b0a5536c1a7cf$export$40bfa8c7b0832715(selectedCountry, defaultSelectedCountry, (c) => {
        onCountryChange === null || onCountryChange === void 0 ? void 0 : onCountryChange(c);
    });
    // Full phone value including the '+' and dial code
    const [fullValue, setFullValue] = useControlledState.$458b0a5536c1a7cf$export$40bfa8c7b0832715(value, defaultValue || country.code + ' ', (val) => {
        onChange === null || onChange === void 0 ? void 0 : onChange(val);
    });
    // Reconcile country with externally-driven value changes ONLY when the
    // consumer controls `value` but leaves `selectedCountry` uncontrolled. In all
    // other paths, `handleInputChange` and `handleCountrySelect` already keep
    // them in sync — running the effect there risks repeated `onCountryChange`
    // calls (controlled-country case) and stale-dep re-runs (inline countries).
    const shouldReconcileCountry = value !== undefined && selectedCountry === undefined;
    h.useEffect(() => {
        if (!shouldReconcileCountry)
            return;
        const matched = matchCountryByValue(customCountries, fullValue);
        if (matched && matched.id !== country.id) {
            setCountry(matched);
        }
    }, [
        shouldReconcileCountry,
        fullValue,
        customCountries,
        country.id,
        setCountry,
    ]);
    const selectOptions = customCountries.map((c) => {
        var _a;
        const iso = c.id.toUpperCase();
        const optionLabel = countryLabelMode === 'iso'
            ? iso
            : countryLabelMode === 'name'
                ? c.label
                : `${(_a = c.flag) !== null && _a !== void 0 ? _a : ''} ${c.label}`.trim();
        return { label: optionLabel, value: c.id };
    });
    const effectivePlaceholder = placeholder !== null && placeholder !== void 0 ? placeholder : `${(_c = country === null || country === void 0 ? void 0 : country.code) !== null && _c !== void 0 ? _c : '+'} (555) 000-0000`;
    const handleInputChange = (val) => {
        let normalized = normalizeValue(val, country.code);
        // Auto-switch country (and flag) when the typed prefix matches a known code.
        const matched = matchCountryByValue(customCountries, normalized);
        if (matched) {
            if (matched.id !== country.id)
                setCountry(matched);
            const rest = normalized.slice(matched.code.length);
            const digits = rest.replace(/\D/g, '');
            if (matched.mask && digits) {
                normalized = `${matched.code} ${applyMask(digits, matched.mask)}`;
            }
            else if (rest && !rest.startsWith(' ')) {
                // Insert a space between the dial code and the national number.
                normalized = `${matched.code} ${rest}`;
            }
        }
        setFullValue(normalized);
    };
    const handleCountrySelect = (iso) => {
        const next = customCountries.find((c) => c.id === iso) || customCountries[0];
        if (next.id === country.id)
            return;
        // Extract national digits from the existing value (drop the previous dial code).
        const currentMatch = matchCountryByValue(customCountries, fullValue);
        const remainder = currentMatch
            ? fullValue.slice(currentMatch.code.length)
            : fullValue;
        const digits = remainder.replace(/\D/g, '');
        const formatted = digits
            ? next.mask
                ? `${next.code} ${applyMask(digits, next.mask)}`
                : `${next.code} ${digits}`
            : `${next.code} `;
        setCountry(next);
        setFullValue(formatted);
    };
    const countryPopoverContents = (jsxRuntime.jsxs(reactAriaComponents.Popover, { placement: "bottom start", offset: 4, className: cx.cx('flex max-h-64 w-72 flex-col overflow-hidden rounded-lg bg-primary shadow-lg ring-1 ring-secondary_alt outline-hidden', searchable ? 'py-0' : 'py-1'), children: [searchable && (jsxRuntime.jsx(PopoverSearchInput.PopoverSearchInput, { "data-id": `${dataId}-search`, value: searchQuery, onChange: setSearchQuery, placeholder: searchPlaceholder })), searchable && filteredCountries.length === 0 ? (jsxRuntime.jsx("div", { className: "px-3 py-4 text-sm text-tertiary", role: "status", children: searchNotFoundLabel })) : (jsxRuntime.jsx(reactAriaComponents.ListBox, { items: filteredCountries, className: "flex-1 overflow-y-auto outline-hidden", children: (c) => (jsxRuntime.jsxs(reactAriaComponents.ListBoxItem, { id: c.id, textValue: `${c.label} ${c.code}`, className: cx.cx('flex cursor-pointer items-center gap-2 px-3 py-2 text-md text-primary outline-hidden', 'data-[focused]:bg-primary_hover data-[selected]:bg-active'), children: [jsxRuntime.jsx(FlagIcon.FlagIcon, { flag: c.id }), jsxRuntime.jsx("span", { className: "flex-1 truncate", children: c.label }), jsxRuntime.jsx("span", { className: "text-tertiary", children: c.code })] })) }))] }));
    const content = variant === 'flag' ? (jsxRuntime.jsxs("div", { className: cx.cx('flex w-full items-center rounded-lg bg-primary shadow-xs ring-1 ring-primary ring-inset transition duration-100 ease-linear', 'focus-within:ring-2 focus-within:ring-brand', isDisabled && 'cursor-not-allowed bg-disabled_subtle', isInvalid && 'ring-error_subtle focus-within:ring-error', className), children: [jsxRuntime.jsxs(reactAriaComponents.Select, { "aria-label": "Country", selectedKey: country === null || country === void 0 ? void 0 : country.id, onSelectionChange: (k) => handleCountrySelect(String(k)), isDisabled: isDisabled, children: [jsxRuntime.jsxs(reactAriaComponents.Button, { className: cx.cx('flex h-full items-center gap-1 outline-hidden cursor-pointer rounded-l-lg', size === 'sm' ? 'pl-3 pr-2 py-2' : 'pl-3.5 pr-2 py-2.5', 'data-[focus-visible]:ring-2 data-[focus-visible]:ring-brand data-[focus-visible]:ring-inset', isDisabled && 'cursor-not-allowed'), "data-id": `${dataId}-country`, children: [jsxRuntime.jsx(FlagIcon.FlagIcon, { flag: country.id }), jsxRuntime.jsx(icons.ChevronDown, { "aria-hidden": "true", className: "size-4 stroke-[2.5px] text-fg-quaternary" })] }), countryPopoverContents] }), jsxRuntime.jsx("input", { ...rest, id: id, ref: ref, type: "tel", inputMode: "tel", disabled: isDisabled, "aria-invalid": isInvalid || undefined, required: isRequired, placeholder: effectivePlaceholder, value: fullValue, onChange: (e) => handleInputChange(e.target.value), "data-id": `${dataId}-input`, className: cx.cx('flex-1 bg-transparent text-md text-primary outline-hidden placeholder:text-placeholder px-2 py-2.5', isDisabled && 'cursor-not-allowed text-disabled') })] })) : (jsxRuntime.jsx(inputGroup.InputGroup, { "data-id": dataId, isDisabled: isDisabled, isInvalid: isInvalid, className: className, value: fullValue, onChange: handleInputChange, leadingAddon: searchable ? (jsxRuntime.jsxs(reactAriaComponents.Select, { "aria-label": "Country", selectedKey: country === null || country === void 0 ? void 0 : country.id, onSelectionChange: (k) => handleCountrySelect(String(k)), isDisabled: isDisabled, children: [jsxRuntime.jsxs(reactAriaComponents.Button, { className: cx.cx('flex h-full items-center gap-1 outline-hidden cursor-pointer rounded-l-lg', size === 'sm' ? 'pl-3 pr-2 py-2' : 'pl-3.5 pr-2 py-2.5', 'data-[focus-visible]:ring-2 data-[focus-visible]:ring-brand data-[focus-visible]:ring-inset', isDisabled && 'cursor-not-allowed'), "data-id": `${dataId}-country`, children: [jsxRuntime.jsx("span", { className: "truncate text-md text-primary", children: (_e = (_d = selectOptions.find((o) => o.value === (country === null || country === void 0 ? void 0 : country.id))) === null || _d === void 0 ? void 0 : _d.label) !== null && _e !== void 0 ? _e : country === null || country === void 0 ? void 0 : country.label }), jsxRuntime.jsx(icons.ChevronDown, { "aria-hidden": "true", className: "size-4 stroke-[2.5px] text-fg-quaternary" })] }), countryPopoverContents] })) : (jsxRuntime.jsx(selectNative.NativeSelect, { "aria-label": "Country", options: selectOptions, value: country === null || country === void 0 ? void 0 : country.id, onChange: (e) => handleCountrySelect(e.target.value) })), children: jsxRuntime.jsx(input.InputBase, { ...rest, id: id, ref: ref, isDisabled: isDisabled, isInvalid: isInvalid, placeholder: effectivePlaceholder, inputMode: "tel", type: "tel", "data-id": `${dataId}-input` }) }));
    return (jsxRuntime.jsx(FieldWrapper.FieldWrapper, { id: id, label: label, hint: hint, isInvalid: isInvalid, isRequired: isRequired, isWarning: isWarning, tooltip: tooltip, description: description, fieldSize: fieldSize, labelPosition: labelPosition, alignLabel: alignLabel, children: content }));
};

exports.PhoneNumberInput = PhoneNumberInput;
exports.countries = countries;
