// @hostaway/design-system/dist/components/Table/index.esm.js
import { jsxs, jsx, Fragment } from 'react/jsx-runtime';
import { useState, useEffect, useMemo, useCallback, useRef, isValidElement } from 'react';
import { match } from 'ts-pattern';
import { s as sortCx, c as cx } from '../_shared/cx-DnIR8RkH.esm.js';
import { T as TableCard, a as Table$1 } from '../_shared/table-ByI6IeP_.esm.js';
import { Settings01, SearchLg, ChevronRight, ChevronDown, Menu01 } from '@untitledui/icons';
import { D as Dropdown } from '../_shared/BaseDropdown-CG2Me0TI.esm.js';
import { C as Checkbox } from '../_shared/Checkbox-DH2haenf.esm.js';
import { ParagraphSpacing } from '@untitledui/icons/ParagraphSpacing';
import { Lock01 } from '@untitledui/icons/Lock01';
import { Pin01 } from '@untitledui/icons/Pin01';
import { I as InputField } from '../_shared/InputField-8Udd41T0.esm.js';
import { B as Button } from '../_shared/Button-C3YUMQZU.esm.js';
import { ArrowDown } from '@untitledui/icons/ArrowDown';
import { HelpCircle } from '@untitledui/icons/HelpCircle';
import { ChevronSelectorVertical } from '@untitledui/icons/ChevronSelectorVertical';
import { useTableOptions, Column, Group } from 'react-aria-components';
import { u as useTruncation } from '../_shared/FieldWrapper-UtX7P_J-.esm.js';
import { C as ConditionalTooltip } from '../_shared/ConditionalTooltip-DT2TnX8y.esm.js';
import { T as Tooltip } from '../_shared/Tooltip-BtHtZ-Fm.esm.js';
import { E as EmptyState } from '../_shared/empty-state-DURw77A3.esm.js';
import { B as Button$1 } from '../_shared/BaseButton-CTtHjEZn.esm.js';
import { P as Pagination } from '../_shared/pagination-base-mHEbsM2P.esm.js';
import { ArrowLeft } from '@untitledui/icons/ArrowLeft';
import { ArrowRight } from '@untitledui/icons/ArrowRight';
import { u as useBreakpoint } from '../_shared/use-breakpoint-D8xwrq2z.esm.js';
import { _ as _Select } from '../_shared/BaseSelect-CHht82wi.esm.js';
import '../_shared/MyCustomIcon-BRHTusqN.esm.js';
import { B as ButtonUtility } from '../_shared/ButtonUtility-CLhBXgkj.esm.js';
import { L as LoadingIndicator } from '../_shared/LoadingIndicator-D57yM39W.esm.js';
import 'tailwind-merge';
import '../_shared/badges-ihdaM96E.esm.js';
import '../_shared/dot-icon-DOJ8mWXT.esm.js';
import '../_shared/checkbox-oR9JZpTn.esm.js';
import '../_shared/tooltip-l8auK4sw.esm.js';
import '../_shared/input-BOpcbRQD.esm.js';
import '../_shared/hint-text-Ck8Vj_nN.esm.js';
import '../_shared/label-pgzo6IFb.esm.js';
import '../_shared/FileIcon-DaGn1dck.esm.js';
import '../_shared/is-react-component-CYKUd65Q.esm.js';
import '../_shared/index-UrSxsuvd.esm.js';
import '../_shared/PopoverSearchInput-D81SVeJp.esm.js';
import '../_shared/avatar-DeJPrYaB.esm.js';
import '../_shared/Icon-BYFFIpjv.esm.js';

const DEFAULT_TEXTS = {
    pagination: {
        labelNext: 'Next',
        labelPrevious: 'Previous',
        labelItemsPerPage: 'Items per page',
    },
    loading: {
        labelLoading: 'Loading...',
    },
    columnsSelector: {
        settingsAriaLabel: 'Manage column visibility',
        selectAllCheckboxLabel: 'Select All',
        findColumnPlaceholder: 'Find Column',
        toggleCheckboxAriaLabel: 'Toggle',
        selectedColumnsLabel: 'Selected Columns',
        reorderColumnAriaLabel: 'Reorder',
    },
};

const DROP_INDICATOR_CLASSES = [
    /* ── Drop indicator line (applied to <table>) ── */
    `[&_.react-aria-DropIndicator[data-drop-target]>td]:relative`,
    `[&_.react-aria-DropIndicator[data-drop-target]>td]:after:content-[""]`,
    `[&_.react-aria-DropIndicator[data-drop-target]>td]:after:absolute`,
    `[&_.react-aria-DropIndicator[data-drop-target]>td]:after:inset-x-0`,
    `[&_.react-aria-DropIndicator[data-drop-target]>td]:after:-top-px`,
    `[&_.react-aria-DropIndicator[data-drop-target]>td]:after:h-0.5`,
    `[&_.react-aria-DropIndicator[data-drop-target]>td]:after:bg-brand-solid`,
    /* ── Hide drop indicators inside parent/child groups ── */
    `[&_.react-aria-DropIndicator:has(+[data-child-row])]:hidden`,
    `[&_[data-child-row]:not([data-last-child])+.react-aria-DropIndicator]:hidden`,
    /* ── Hide drop indicator above the dragged row (no-op drop) ── */
    `[&_.react-aria-DropIndicator:has(+[data-dragging])]:hidden`,
].join(' ');

/**
 * Width constraints applied to each column-size tier. The `<th>` and the
 * cell-content wrapper both consume this so the rendered widths stay aligned.
 */
const COL_SIZE_CLASS = sortCx({
    fill: '',
    xs: 'w-[24px]',
    sm: 'min-w-[64px] max-w-[120px]',
    md: 'min-w-[120px] max-w-[200px]',
    lg: 'min-w-[200px] max-w-[360px]',
});
/**
 * Pixel row-height per table size. Exposed as the `--row-height` CSS variable
 * so `content-visibility` reservations match the actual rendered row.
 */
const TABLE_ROW_HEIGHT = {
    sm: 56,
    md: 72,
};
/** Inline padding applied to each `<th>` based on the table size. */
const TABLE_HEAD_PADDING = {
    sm: 'pr-4 first:pl-3 last:pr-3',
    md: 'pr-6 first:pl-4 last:pr-4',
};
/**
 * Margin/padding applied to the inner `<div>` of every body cell. Implemented
 * via child selectors because the cell itself has no padding (the inner div
 * carries the size-constrained min/max widths).
 */
const TABLE_CELL_PADDING = {
    sm: '[&>div]:my-2 [&>div]:mr-4 first:[&>div]:ml-3 last:[&>div]:mr-3',
    md: '[&>div]:my-4 [&>div]:mr-6 first:[&>div]:ml-4 last:[&>div]:mr-4',
};
/** Padding applied to footer cells. Mirrors the body-cell horizontal rhythm. */
const TABLE_FOOTER_CELL_PADDING = {
    sm: 'px-4 py-2',
    md: 'px-6 py-2',
};
/** Padding for the `tableCardHeader` slot wrapper. */
const TABLE_CARD_HEADER_PADDING = {
    sm: 'py-4 md:px-5',
    md: 'py-5 md:px-6',
};
/**
 * Horizontal indent of the vertical connector line in collapsible-tree rows.
 * Must land directly under the parent row's expand chevron, whose position
 * shifts with selection-checkbox and drag-handle presence.
 */
const TREE_LINE_INDENT = (size, hasSelection, hasDragHandle) => match({ size, hasSelection, hasDragHandle })
    .with({ hasSelection: true, hasDragHandle: true }, () => 'left-11')
    .with({ hasSelection: true }, () => 'left-5')
    .with({ hasDragHandle: true, size: 'sm' }, () => 'left-11')
    .with({ hasDragHandle: true, size: 'md' }, () => 'left-12')
    .with({ size: 'sm' }, () => 'left-5')
    .with({ size: 'md' }, () => 'left-6')
    .exhaustive();

function isColumnSelected(column, columnsSelection) {
    if (column.alwaysVisible) {
        return true;
    }
    if (column.isSelected !== undefined) {
        return column.isSelected;
    }
    if (columnsSelection !== undefined &&
        columnsSelection[String(column.id)] !== undefined) {
        return columnsSelection[String(column.id)];
    }
    return column.defaultVisibility === undefined || column.defaultVisibility;
}

function useColumns(columnDefs, columnsSelectorEnabled, columnOrder, columnsSelection) {
    const [columns, setColumns] = useState(columnDefs);
    useEffect(() => {
        setColumns((prev) => {
            if (prev.length !== columnDefs.length) {
                return columnDefs;
            }
            const changed = columnDefs.some((col, i) => col.id !== prev[i].id || col.isSelected !== prev[i].isSelected);
            return changed ? columnDefs : prev;
        });
    }, [columnDefs]);
    const visibleColumns = useMemo(() => {
        const cols = columnsSelectorEnabled
            ? columns.filter((c) => isColumnSelected(c, columnsSelection))
            : columns;
        const left = cols.filter((c) => c.pinned === 'left');
        const right = cols.filter((c) => c.pinned === 'right');
        let center = cols.filter((c) => !c.pinned);
        // Apply column ordering to non-pinned columns
        if (columnOrder && Object.keys(columnOrder).length > 0) {
            const maxExplicit = Math.max(...Object.values(columnOrder), -1);
            center = [...center]
                .map((col, i) => ({
                col,
                sortKey: col.id != null && col.id in columnOrder
                    ? columnOrder[col.id]
                    : maxExplicit + 1 + i,
            }))
                .sort((a, b) => a.sortKey - b.sortKey)
                .map((item) => item.col);
        }
        if (left.length === 0 && right.length === 0) {
            return center;
        }
        return [...left, ...center, ...right];
    }, [columns, columnsSelectorEnabled, columnOrder, columnsSelection]);
    return { columns, setColumns, visibleColumns };
}

function useStickyColumns(visibleColumns, { enabled }) {
    // Track the wrapper element via state (driven by a callback ref) instead of
    // a plain object ref. Consumers can render the table conditionally — e.g.
    // an empty state on first paint that's later swapped for the table — so the
    // ref can attach *after* the first effect runs. Putting the element in
    // state adds it to the effect's deps, ensuring the effect re-runs as soon
    // as the wrapper mounts.
    const [wrapper, setWrapper] = useState(null);
    const wrapperRef = useCallback((node) => {
        // Skip the state update when disabled so consumers (e.g. tables with no
        // pinned columns) don't pay for an extra render on mount.
        if (!enabled) {
            return;
        }
        setWrapper(node);
    }, [enabled]);
    const styleElRef = useRef(null);
    const scopeIdRef = useRef('');
    useEffect(() => {
        if (!wrapper || !enabled)
            return;
        const table = wrapper.querySelector('table');
        if (!table)
            return;
        const scrollContainer = table.parentElement;
        if (!scrollContainer)
            return;
        const pinnedLeftIndexes = [];
        const pinnedRightIndexes = [];
        visibleColumns.forEach((col, i) => {
            if (col.pinned === 'left')
                pinnedLeftIndexes.push(i);
            if (col.pinned === 'right')
                pinnedRightIndexes.push(i);
        });
        if (pinnedLeftIndexes.length === 0 && pinnedRightIndexes.length === 0)
            return;
        // Assign a unique scope to this table so CSS rules don't leak across instances
        if (!scopeIdRef.current) {
            scopeIdRef.current = `sticky-${Math.random().toString(36).slice(2, 8)}`;
        }
        const scopeId = scopeIdRef.current;
        table.setAttribute('data-sticky', scopeId);
        const buildStyles = () => {
            // --- Read: measure cell widths from the header row ---
            const headerCells = table.querySelectorAll('thead th, thead td');
            // react-aria-components may auto-inject leading cells (selection
            // checkbox when selectionMode is set, drag handles, etc.) that don't
            // correspond to entries in `visibleColumns`. They appear before the
            // user's columns in the DOM, so every visibleColumns index needs to be
            // shifted by this amount when targeting nth-child / measuring widths.
            const domOffset = Math.max(0, headerCells.length - visibleColumns.length);
            // Treat auto-injected leading cells as implicit pinned-left so they
            // stay visible alongside user pinned-left columns when scrolling.
            // Their absence from visibleColumns means we'd otherwise leave them
            // floating without a stick rule.
            const autoLeftIndexes = pinnedLeftIndexes.length > 0
                ? Array.from({ length: domOffset }, (_, i) => i)
                : [];
            const leftDomIndexes = [
                ...autoLeftIndexes,
                ...pinnedLeftIndexes.map((i) => i + domOffset),
            ];
            const rightDomIndexes = pinnedRightIndexes.map((i) => i + domOffset);
            const leftWidths = leftDomIndexes.map((i) => { var _a, _b; return (_b = (_a = headerCells[i]) === null || _a === void 0 ? void 0 : _a.offsetWidth) !== null && _b !== void 0 ? _b : 0; });
            const rightWidths = rightDomIndexes.map((i) => { var _a, _b; return (_b = (_a = headerCells[i]) === null || _a === void 0 ? void 0 : _a.offsetWidth) !== null && _b !== void 0 ? _b : 0; });
            // Cumulative left offsets
            const leftOffsets = [];
            let accLeft = 0;
            for (const w of leftWidths) {
                leftOffsets.push(accLeft);
                accLeft += w;
            }
            // Cumulative right offsets (accumulated from the end)
            const rightOffsets = [];
            let accRight = 0;
            for (let i = rightWidths.length - 1; i >= 0; i--) {
                rightOffsets[i] = accRight;
                accRight += rightWidths[i];
            }
            // --- Write: build CSS rules that apply to all rows via selectors ---
            const scope = `table[data-sticky="${scopeId}"]`;
            const col = (domIndex) => `${scope} tr > :nth-child(${domIndex + 1})`;
            const rules = [];
            // Rows need an opaque background so sticky cells can inherit it
            rules.push(`${scope} thead tr { background-color: var(--color-bg-secondary); }`);
            rules.push(`${scope} tbody tr { background-color: var(--color-bg-primary); }`);
            for (let i = 0; i < leftDomIndexes.length; i++) {
                rules.push(`${col(leftDomIndexes[i])} { position: sticky; left: ${leftOffsets[i]}px; z-index: 20; background-color: inherit; }`);
            }
            for (let i = 0; i < rightDomIndexes.length; i++) {
                rules.push(`${col(rightDomIndexes[i])} { position: sticky; right: ${rightOffsets[i]}px; z-index: 20; background-color: inherit; }`);
            }
            // Shadow on last left-pinned column (visible only when scrolled from start)
            if (leftDomIndexes.length > 0) {
                const lastLeft = leftDomIndexes[leftDomIndexes.length - 1];
                rules.push(`${scope}[data-scrolled-start] tr > :nth-child(${lastLeft + 1})::before { content: ''; position: absolute; top: 0; bottom: 0; right: -4px; width: 0.25rem; background: linear-gradient(to right, rgb(0 0 0 / 0.06), transparent); pointer-events: none; }`);
            }
            // Shadow on first right-pinned column (visible only when not scrolled to end)
            if (rightDomIndexes.length > 0) {
                const firstRight = rightDomIndexes[0];
                rules.push(`${scope}[data-scrolled-end] tr > :nth-child(${firstRight + 1})::before { content: ''; position: absolute; top: 0; bottom: 0; left: -4px; width: 0.25rem; background: linear-gradient(to left, rgb(0 0 0 / 0.06), transparent); pointer-events: none; }`);
            }
            if (!styleElRef.current) {
                const style = document.createElement('style');
                wrapper.appendChild(style);
                styleElRef.current = style;
            }
            styleElRef.current.textContent = rules.join('\n');
        };
        const updateShadows = () => {
            const { scrollLeft, scrollWidth, clientWidth } = scrollContainer;
            table.toggleAttribute('data-scrolled-start', scrollLeft > 0);
            table.toggleAttribute('data-scrolled-end', scrollLeft + clientWidth < scrollWidth - 1);
        };
        buildStyles();
        updateShadows();
        scrollContainer.addEventListener('scroll', updateShadows, {
            passive: true,
        });
        // Recalculate widths on resize — no MutationObserver needed since
        // CSS nth-child rules automatically apply to new/removed rows
        let rafId = 0;
        const resizeObserver = new ResizeObserver(() => {
            cancelAnimationFrame(rafId);
            rafId = requestAnimationFrame(buildStyles);
        });
        resizeObserver.observe(scrollContainer);
        return () => {
            var _a;
            cancelAnimationFrame(rafId);
            resizeObserver.disconnect();
            scrollContainer.removeEventListener('scroll', updateShadows);
            (_a = styleElRef.current) === null || _a === void 0 ? void 0 : _a.remove();
            styleElRef.current = null;
            table.removeAttribute('data-sticky');
            table.removeAttribute('data-scrolled-start');
            table.removeAttribute('data-scrolled-end');
        };
    }, [wrapper, visibleColumns, enabled]);
    return wrapperRef;
}

function isValidRowId(value) {
    return ((typeof value === 'string' && value.length > 0) ||
        (typeof value === 'number' && !Number.isNaN(value)));
}
function resolveRowId(row, index, getRowId) {
    const explicitId = typeof getRowId === 'function' ? getRowId(row, index) : undefined;
    if (isValidRowId(explicitId))
        return explicitId;
    if (typeof row === 'object' && isValidRowId(row.id))
        return row.id;
    if ((typeof row === 'number' || typeof row === 'string') && isValidRowId(row))
        return row;
    return index;
}
const DEFAULT_META = {
    isParent: false,
    isChild: false,
    isLastChild: false,
};
function useTableItems(rowData, getRowId, getChildRows, collapsible) {
    const [collapsedKeys, setCollapsedKeys] = useState(() => new Set());
    const toggleExpanded = useCallback((key) => {
        setCollapsedKeys((prev) => {
            const next = new Set(prev);
            if (next.has(key))
                next.delete(key);
            else
                next.add(key);
            return next;
        });
    }, []);
    if (!getChildRows) {
        return {
            items: rowData.map((row, index) => ({
                __row: row,
                __index: index,
                __meta: DEFAULT_META,
                id: resolveRowId(row, index, getRowId),
            })),
            hasChildRows: false,
            collapsedKeys: undefined,
            toggleExpanded: undefined,
        };
    }
    let flatIndex = 0;
    let anyChildren = false;
    const items = [];
    rowData.forEach((row, sourceIndex) => {
        var _a;
        const id = resolveRowId(row, sourceIndex, getRowId);
        const children = (_a = getChildRows(row)) !== null && _a !== void 0 ? _a : [];
        const isParent = children.length > 0;
        if (isParent)
            anyChildren = true;
        items.push({
            __row: row,
            __index: flatIndex++,
            __meta: { isParent, isChild: false, isLastChild: false },
            id,
        });
        const isExpanded = !collapsible || !collapsedKeys.has(id);
        if (isParent && isExpanded) {
            children.forEach((child, childIndex) => {
                items.push({
                    __row: child,
                    __index: flatIndex++,
                    __meta: {
                        isParent: false,
                        isChild: true,
                        isLastChild: childIndex === children.length - 1,
                    },
                    id: resolveRowId(child, childIndex, getRowId),
                });
            });
        }
    });
    return {
        items,
        hasChildRows: anyChildren,
        collapsedKeys: collapsible ? collapsedKeys : undefined,
        toggleExpanded: collapsible ? toggleExpanded : undefined,
    };
}

function ColumnCheckboxList({ columns, allColumnDefs, columnsSelection, onColumnsSelectedChange, onColumnsSelectionChange, texts, dataId, }) {
    const { selectedAmount, unselectedAmount } = columns.reduce((acc, c) => {
        const isSelected = isColumnSelected(c, columnsSelection);
        if (isSelected) {
            acc.selectedAmount += 1;
        }
        else {
            acc.unselectedAmount += 1;
        }
        return acc;
    }, {
        selectedAmount: 0,
        unselectedAmount: 0,
    });
    const isAllSelected = selectedAmount === columns.length;
    const isAllUnselected = unselectedAmount === columns.length;
    const isAllIndeterminate = !isAllSelected && !isAllUnselected;
    // Sort checked items to the top
    const sorted = [...columns].sort((a, b) => {
        const aSelected = !isColumnSelected(a, columnsSelection) ? 0 : 1;
        const bSelected = !isColumnSelected(b, columnsSelection) ? 0 : 1;
        return bSelected - aSelected;
    });
    return (jsxs("div", { className: "overflow-y-auto max-h-120 pb-2", children: [jsx("div", { className: "px-3 py-1.5", children: jsx(Checkbox, { isSelected: isAllSelected, isIndeterminate: isAllIndeterminate, label: texts.columnsSelector.selectAllCheckboxLabel, onChange: () => {
                        const newSelected = !isAllSelected;
                        const columnIds = new Set(columns.map((c) => c.id));
                        const cols = allColumnDefs.map((c) => c.alwaysVisible || !columnIds.has(c.id)
                            ? c
                            : { ...c, isSelected: newSelected });
                        onColumnsSelectionChange(cols);
                        onColumnsSelectedChange === null || onColumnsSelectedChange === void 0 ? void 0 : onColumnsSelectedChange(allColumnDefs.reduce((acc, c) => ({
                            ...acc,
                            [String(c.id)]: c.alwaysVisible || newSelected,
                        }), {}));
                    }, "data-id": `${dataId}-selectAll` }) }), sorted.map((column) => {
                const isVisible = isColumnSelected(column, columnsSelection);
                return (jsx("div", { className: "px-3 py-1.5", children: jsx(Checkbox, { isSelected: isVisible, "aria-label": `${texts.columnsSelector.toggleCheckboxAriaLabel} ${column.label}`, label: column.label, onChange: () => {
                            const cols = allColumnDefs.map((c) => {
                                if (c.id === column.id) {
                                    return {
                                        ...c,
                                        isSelected: column.isSelected === undefined ? false : !isVisible,
                                    };
                                }
                                return c;
                            });
                            onColumnsSelectionChange(cols);
                            const selectedCols = allColumnDefs.reduce((acc, c) => {
                                const isSelected = isColumnSelected(c, columnsSelection);
                                return {
                                    ...acc,
                                    [String(c.id)]: c.id === column.id ? !isSelected : isSelected,
                                };
                            }, {});
                            onColumnsSelectedChange === null || onColumnsSelectedChange === void 0 ? void 0 : onColumnsSelectedChange(selectedCols);
                        }, "data-id": `${dataId}-checkbox-${column.id}` }) }, column.id));
            })] }));
}

function sortByOrder(cols, columnOrder) {
    if (!columnOrder || Object.keys(columnOrder).length === 0)
        return cols;
    return [...cols].sort((a, b) => {
        const aOrder = a.id != null && a.id in columnOrder ? columnOrder[a.id] : Infinity;
        const bOrder = b.id != null && b.id in columnOrder ? columnOrder[b.id] : Infinity;
        return aOrder - bOrder;
    });
}
function buildOrderMap(ids) {
    const map = {};
    ids.forEach((id, i) => {
        map[id] = i;
    });
    return map;
}
function SelectedColumnItem({ column, isPinned, onDragStart, onDragOver, onDrop, onDragEnd, onKeyDown, isDragOver, dropSide, isDragging, dataId, }) {
    return (jsxs("div", { tabIndex: isPinned ? undefined : 0, role: isPinned ? undefined : 'option', "aria-label": column.label, "aria-roledescription": isPinned ? undefined : 'sortable', className: cx('relative flex items-center gap-1.5 rounded-md px-1.5 py-1 text-sm text-secondary outline-none', isPinned
            ? 'text-quaternary'
            : 'cursor-grab focus-visible:ring-2 focus-visible:ring-brand-solid', isDragging && 'opacity-40'), draggable: !isPinned, onDragStart: onDragStart, onDragOver: onDragOver, onDrop: onDrop, onDragEnd: onDragEnd, onKeyDown: onKeyDown, "data-id": `${dataId}-${isPinned ? 'pinned' : 'reorderable'}-${column.id}`, ...(!isPinned ? { 'data-reorderable': '' } : {}), children: [isDragOver && dropSide === 'top' && (jsx("div", { className: "absolute top-0 right-1 left-1 h-0.5 -translate-y-1/2 rounded-full bg-brand-solid" })), isPinned ? (jsx(Lock01, { className: "size-4 shrink-0 text-fg-quaternary" })) : (jsx(ParagraphSpacing, { className: "size-4 shrink-0 text-fg-quaternary" })), jsx("span", { className: "flex-1 truncate", children: column.label }), isPinned && jsx(Pin01, { className: "size-4 shrink-0 text-fg-quaternary" }), isDragOver && dropSide === 'bottom' && (jsx("div", { className: "absolute bottom-0 right-1 left-1 h-0.5 translate-y-1/2 rounded-full bg-brand-solid" }))] }));
}
function SelectedColumnsPanel({ columns, columnOrder, onColumnOrderChange, texts, dataId, }) {
    const [draggedId, setDraggedId] = useState(null);
    const [dragOverId, setDragOverId] = useState(null);
    const [dropSide, setDropSide] = useState(null);
    const ordered = useMemo(() => sortByOrder(columns, columnOrder), [columns, columnOrder]);
    const pinnedLeft = ordered.filter((c) => c.pinned === 'left');
    const pinnedRight = ordered.filter((c) => c.pinned === 'right');
    const reorderable = ordered.filter((c) => !c.pinned);
    const handleDragStart = useCallback((columnId) => (e) => {
        e.dataTransfer.effectAllowed = 'move';
        e.dataTransfer.setData('text/plain', String(columnId));
        setDraggedId(columnId);
    }, []);
    const handleDragOver = useCallback((columnId) => (e) => {
        e.preventDefault();
        e.dataTransfer.dropEffect = 'move';
        const rect = e.currentTarget.getBoundingClientRect();
        const midpoint = rect.top + rect.height / 2;
        setDropSide(e.clientY < midpoint ? 'top' : 'bottom');
        setDragOverId(columnId);
    }, []);
    const handleDrop = useCallback((targetId) => (e) => {
        e.preventDefault();
        if (draggedId == null || draggedId === targetId) {
            setDraggedId(null);
            setDragOverId(null);
            setDropSide(null);
            return;
        }
        const ids = reorderable
            .map((c) => c.id)
            .filter((id) => id != null);
        const withoutDragged = ids.filter((id) => id !== draggedId);
        const targetIndex = withoutDragged.indexOf(targetId);
        if (targetIndex === -1) {
            setDraggedId(null);
            setDragOverId(null);
            setDropSide(null);
            return;
        }
        const rect = e.currentTarget.getBoundingClientRect();
        const midpoint = rect.top + rect.height / 2;
        const insertAfter = e.clientY >= midpoint;
        const insertIndex = insertAfter ? targetIndex + 1 : targetIndex;
        withoutDragged.splice(insertIndex, 0, draggedId);
        onColumnOrderChange === null || onColumnOrderChange === void 0 ? void 0 : onColumnOrderChange(buildOrderMap(withoutDragged));
        setDraggedId(null);
        setDragOverId(null);
        setDropSide(null);
    }, [draggedId, reorderable, onColumnOrderChange]);
    const handleDragEnd = useCallback(() => {
        setDraggedId(null);
        setDragOverId(null);
        setDropSide(null);
    }, []);
    const noopDrag = useCallback((e) => {
        e.preventDefault();
    }, []);
    const listRef = useRef(null);
    const handleKeyboardReorder = useCallback((columnId, e) => {
        if (e.key !== 'ArrowUp' && e.key !== 'ArrowDown')
            return;
        e.preventDefault();
        const ids = reorderable
            .map((c) => c.id)
            .filter((id) => id != null);
        const currentIndex = ids.indexOf(columnId);
        if (currentIndex === -1)
            return;
        const swapIndex = e.key === 'ArrowUp' ? currentIndex - 1 : currentIndex + 1;
        if (swapIndex < 0 || swapIndex >= ids.length)
            return;
        const newIds = [...ids];
        [newIds[currentIndex], newIds[swapIndex]] = [
            newIds[swapIndex],
            newIds[currentIndex],
        ];
        onColumnOrderChange === null || onColumnOrderChange === void 0 ? void 0 : onColumnOrderChange(buildOrderMap(newIds));
        // Restore focus to the moved item after React re-renders
        requestAnimationFrame(() => {
            var _a, _b;
            const items = (_a = listRef.current) === null || _a === void 0 ? void 0 : _a.querySelectorAll('[data-reorderable]');
            (_b = items === null || items === void 0 ? void 0 : items[swapIndex]) === null || _b === void 0 ? void 0 : _b.focus();
        });
    }, [reorderable, onColumnOrderChange]);
    return (jsxs("div", { className: "flex flex-col", children: [jsx("div", { className: "px-3 py-2 text-xs font-semibold text-quaternary", children: texts.columnsSelector.selectedColumnsLabel }), jsxs("div", { ref: listRef, className: "flex-1 overflow-y-auto px-1.5 pb-2", children: [pinnedLeft.map((col) => col.id != null ? (jsx(SelectedColumnItem, { column: col, isPinned: true, onDragStart: noopDrag, onDragOver: noopDrag, onDrop: noopDrag, onDragEnd: handleDragEnd, isDragOver: false, dropSide: null, isDragging: false, dataId: `${dataId}-pinnedLeft` }, col.id)) : null), reorderable.map((col) => col.id != null ? (jsx(SelectedColumnItem, { column: col, isPinned: false, onDragStart: handleDragStart(col.id), onDragOver: handleDragOver(col.id), onDrop: handleDrop(col.id), onDragEnd: handleDragEnd, onKeyDown: (e) => handleKeyboardReorder(col.id, e), isDragOver: dragOverId === col.id && draggedId !== col.id, dropSide: dragOverId === col.id ? dropSide : null, isDragging: draggedId === col.id, dataId: `${dataId}-reorderable` }, col.id)) : null), pinnedRight.map((col) => col.id != null ? (jsx(SelectedColumnItem, { column: col, isPinned: true, onDragStart: noopDrag, onDragOver: noopDrag, onDrop: noopDrag, onDragEnd: handleDragEnd, isDragOver: false, dropSide: null, isDragging: false, dataId: `${dataId}-pinnedRight` }, col.id)) : null)] })] }));
}

function TableColumnsSelector({ columnDefs, columnsSelection, onColumnsSelectedChange, onColumnsSelectionChange, columnsSelectorTabs, columnOrder, onColumnOrderChange, t, 'data-id': dataId, }) {
    var _a, _b, _c, _d;
    const texts = {
        columnsSelector: {
            ...DEFAULT_TEXTS.columnsSelector,
            ...t === null || t === void 0 ? void 0 : t.columnsSelector,
        },
    };
    const [filter, setFilter] = useState('');
    const [activeTab, setActiveTab] = useState((_b = (_a = columnsSelectorTabs === null || columnsSelectorTabs === void 0 ? void 0 : columnsSelectorTabs[0]) === null || _a === void 0 ? void 0 : _a.id) !== null && _b !== void 0 ? _b : '');
    const tablistRef = useRef(null);
    const toggleableColumns = columnDefs.filter((c) => !c.alwaysVisible);
    const filteredColumns = filter
        ? toggleableColumns.filter((c) => { var _a; return ((_a = c.label) === null || _a === void 0 ? void 0 : _a.toLowerCase().indexOf(filter.toLowerCase())) !== -1; })
        : toggleableColumns;
    const getTabColumns = (tabId) => filteredColumns.filter((c) => { var _a, _b; return ((_a = c.category) !== null && _a !== void 0 ? _a : (_b = columnsSelectorTabs === null || columnsSelectorTabs === void 0 ? void 0 : columnsSelectorTabs[0]) === null || _b === void 0 ? void 0 : _b.id) === tabId; });
    const visibleTabs = columnsSelectorTabs === null || columnsSelectorTabs === void 0 ? void 0 : columnsSelectorTabs.filter((tab) => toggleableColumns.some((c) => { var _a, _b; return ((_a = c.category) !== null && _a !== void 0 ? _a : (_b = columnsSelectorTabs === null || columnsSelectorTabs === void 0 ? void 0 : columnsSelectorTabs[0]) === null || _b === void 0 ? void 0 : _b.id) === tab.id; }));
    const hasTabs = visibleTabs && visibleTabs.length > 0;
    const selectedColumns = columnDefs.filter((c) => c.label && isColumnSelected(c, columnsSelection));
    const getTabSelectedCount = (tabId) => toggleableColumns.filter((c) => {
        var _a, _b;
        return ((_a = c.category) !== null && _a !== void 0 ? _a : (_b = columnsSelectorTabs === null || columnsSelectorTabs === void 0 ? void 0 : columnsSelectorTabs[0]) === null || _b === void 0 ? void 0 : _b.id) === tabId &&
            isColumnSelected(c, columnsSelection);
    }).length;
    // Arrow key navigation for tablist
    const handleTabKeyDown = (e) => {
        var _a, _b;
        if (!visibleTabs) {
            return;
        }
        const tabIds = visibleTabs.map((t) => t.id);
        const currentIndex = tabIds.indexOf(activeTab);
        if (currentIndex === -1) {
            return;
        }
        let nextIndex = null;
        if (e.key === 'ArrowDown' || e.key === 'ArrowRight') {
            nextIndex = (currentIndex + 1) % tabIds.length;
        }
        else if (e.key === 'ArrowUp' || e.key === 'ArrowLeft') {
            nextIndex = (currentIndex - 1 + tabIds.length) % tabIds.length;
        }
        if (nextIndex != null) {
            e.preventDefault();
            setActiveTab(tabIds[nextIndex]);
            const buttons = (_a = tablistRef.current) === null || _a === void 0 ? void 0 : _a.querySelectorAll('[role="tab"]');
            (_b = buttons === null || buttons === void 0 ? void 0 : buttons[nextIndex]) === null || _b === void 0 ? void 0 : _b.focus();
        }
    };
    // Non-tabbed layout: two panels (checkboxes + selected columns)
    if (!hasTabs) {
        return (jsxs(Dropdown.Root, { children: [jsx(Button, { size: "sm", color: "tertiary", "aria-label": texts.columnsSelector.settingsAriaLabel, className: "bg-transparent hover:bg-transparent focus:bg-transparent !w-17", "data-id": `${dataId}-settings`, children: jsx(Settings01, { className: "size-4 text-quaternary" }) }), jsxs(Dropdown.Popover, { className: "flex !overflow-hidden w-[480px] max-w-[calc(100vw-1rem)]", children: [jsxs("div", { className: "flex min-w-0 flex-1 flex-col border-r border-secondary", children: [jsx("div", { className: "border-b border-secondary p-2", children: jsx(InputField, { icon: SearchLg, value: filter, onChange: (val) => {
                                            setFilter(val);
                                        }, placeholder: texts.columnsSelector.findColumnPlaceholder, className: "text-sm", "data-id": `${dataId}-filter` }) }), jsx("div", { className: "flex-1 overflow-y-auto", children: jsx(ColumnCheckboxList, { columns: filteredColumns, allColumnDefs: columnDefs, columnsSelection: columnsSelection, onColumnsSelectedChange: onColumnsSelectedChange, onColumnsSelectionChange: onColumnsSelectionChange, texts: texts, dataId: `${dataId}-checkboxList` }) })] }), jsx("div", { className: "flex w-2/5 flex-col", children: jsx(SelectedColumnsPanel, { columns: selectedColumns, columnOrder: columnOrder, onColumnOrderChange: onColumnOrderChange, texts: texts, dataId: `${dataId}-selectedColumns` }) })] })] }));
    }
    // Tabbed layout: three-panel with vertical tabs
    const resolvedActiveTab = visibleTabs.some((t) => t.id === activeTab)
        ? activeTab
        : ((_d = (_c = visibleTabs[0]) === null || _c === void 0 ? void 0 : _c.id) !== null && _d !== void 0 ? _d : '');
    const activeTabColumns = getTabColumns(resolvedActiveTab);
    return (jsxs(Dropdown.Root, { children: [jsx(Button, { size: "sm", color: "tertiary", "aria-label": texts.columnsSelector.settingsAriaLabel, className: "bg-transparent hover:bg-transparent focus:bg-transparent !w-17", "data-id": `${dataId}-tabbed-settings`, children: jsx(Settings01, { className: "size-4" }) }), jsxs(Dropdown.Popover, { className: "flex !overflow-hidden w-[775px] max-w-[calc(100vw-1rem)]", children: [jsx("div", { ref: tablistRef, className: "flex w-[32%] min-w-0 flex-col overflow-hidden border-r border-secondary bg-primary", role: "tablist", "aria-orientation": "vertical", "aria-label": "Column categories", children: visibleTabs.map((tab) => (jsxs(Button, { type: "button", role: "tab", color: "tertiary", size: "sm", "aria-selected": resolvedActiveTab === tab.id, tabIndex: 0, className: cx('!flex w-full items-center !px-3 !py-2 text-left text-sm !rounded-none focus-visible:ring-2 focus-visible:ring-inset focus-visible:ring-brand-solid [&>[data-text]]:flex [&>[data-text]]:w-full [&>[data-text]]:min-w-0 [&>[data-text]]:items-center [&>[data-text]]:justify-between', resolvedActiveTab === tab.id
                                ? 'bg-secondary !font-medium !text-secondary_hover'
                                : '!font-normal !text-primary'), onPress: () => setActiveTab(tab.id), onKeyDown: handleTabKeyDown, "data-id": `${dataId}-tab-${tab.id}`, children: [jsx("span", { className: "truncate", children: tab.label }), getTabSelectedCount(tab.id) > 0 && (jsx("span", { className: "inline-flex min-w-4 items-center justify-center rounded border border-secondary px-0.5 text-[10px] leading-4 text-tertiary", children: getTabSelectedCount(tab.id) }))] }, tab.id))) }), jsxs("div", { className: "flex min-w-0 flex-1 flex-col border-r border-secondary", role: "tabpanel", "aria-labelledby": `${dataId}-tab-${resolvedActiveTab}`, children: [jsx("div", { className: "border-b border-secondary p-2", children: jsx(InputField, { icon: SearchLg, value: filter, onChange: (val) => {
                                        setFilter(val);
                                    }, placeholder: texts.columnsSelector.findColumnPlaceholder, className: "text-sm", "data-id": `${dataId}-tabbed-filter` }) }), jsx("div", { className: "flex-1 overflow-y-auto", children: jsx(ColumnCheckboxList, { columns: activeTabColumns, allColumnDefs: columnDefs, columnsSelection: columnsSelection, onColumnsSelectedChange: onColumnsSelectedChange, onColumnsSelectionChange: onColumnsSelectionChange, texts: texts, dataId: `${dataId}-tabbed-checkboxList` }) })] }), jsx("div", { className: "flex w-[32%] min-w-0 flex-col", children: jsx(SelectedColumnsPanel, { columns: selectedColumns, columnOrder: columnOrder, onColumnOrderChange: onColumnOrderChange, texts: texts, dataId: `${dataId}-tabbed-selectedColumns` }) })] })] }));
}

function TableHeadLabel({ label, shouldHideLabel, }) {
    const { ref, isTruncated } = useTruncation('x');
    return (jsx(ConditionalTooltip, { text: label, isTruncated: isTruncated, children: jsx("span", { ref: ref, className: cx('inline-block', shouldHideLabel
                ? 'sr-only'
                : 'text-xs font-semibold whitespace-nowrap text-quaternary text-left text-ellipsis overflow-hidden'), children: label }) }));
}

function TableHead({ className, tooltip, label, shouldHideLabel, children, size, 'data-id': dataId, ...props }) {
    const { selectionBehavior } = useTableOptions();
    return (jsx(Column, { ...props, className: (state) => cx('relative p-0 outline-hidden focus-visible:z-1 focus-visible:ring-2 focus-visible:ring-focus-ring focus-visible:ring-offset-bg-primary focus-visible:ring-inset', selectionBehavior === 'toggle' && 'nth-2:pl-3', state.allowsSorting && 'cursor-pointer', typeof className === 'function' ? className(state) : className), children: (state) => {
            var _a;
            return (jsxs(Group, { className: cx('flex items-center gap-1', size && COL_SIZE_CLASS[size]), children: [jsxs("div", { className: "flex items-center gap-1 overflow-hidden", children: [label && (jsx(TableHeadLabel, { label: label, shouldHideLabel: shouldHideLabel, "data-id": `${dataId}-label` })), typeof children === 'function' ? children(state) : children] }), tooltip && (jsx(Tooltip, { title: tooltip.title, description: tooltip.description, arrow: (_a = tooltip.arrow) !== null && _a !== void 0 ? _a : false, placement: "top", triggerProps: {
                            'aria-label': typeof tooltip.title === 'string' ? tooltip.title : 'Help',
                            className: 'text-fg-quaternary transition duration-100 ease-linear hover:text-fg-quaternary_hover focus:text-fg-quaternary_hover',
                        }, "data-id": `${dataId}-tooltip`, children: jsx(HelpCircle, { size: 16 }) })), state.allowsSorting &&
                        (state.sortDirection ? (jsx(ArrowDown, { className: cx('size-3 stroke-[3px] text-fg-quaternary shrink-0', state.sortDirection === 'ascending' && 'rotate-180') })) : (jsx(ChevronSelectorVertical, { size: 12, strokeWidth: 3, className: "text-fg-quaternary shrink-0" })))] }));
        } }));
}

function TableEmptyState({ emptyState }) {
    return (jsx("div", { className: "flex items-center justify-center overflow-hidden px-8 pt-10 pb-12", children: jsxs(EmptyState, { size: "sm", children: [jsx(EmptyState.Header, { pattern: "circle", className: "text-gray-200", children: jsx(EmptyState.FeaturedIcon, { color: "gray", theme: "modern-neue" }) }), (emptyState === null || emptyState === void 0 ? void 0 : emptyState.title) || (emptyState === null || emptyState === void 0 ? void 0 : emptyState.description) ? (jsxs(EmptyState.Content, { children: [jsx(EmptyState.Title, { children: emptyState.title }), jsx(EmptyState.Description, { children: emptyState.description })] })) : null, (emptyState === null || emptyState === void 0 ? void 0 : emptyState.footer) ? (jsx(EmptyState.Footer, { children: emptyState === null || emptyState === void 0 ? void 0 : emptyState.footer })) : null] }) }));
}

function getPageSizeOptions(sizes) {
    return sizes.map((size) => ({
        label: String(size),
        id: String(size),
    }));
}
const DEFAULT_LIMITS = [20, 50, 100];
function PageSizeSelect({ paginationPageSize, paginationPageSizeSelector, onPaginationChanged, onPageLimitChange, label, className, }) {
    const limits = useMemo(() => Array.isArray(paginationPageSizeSelector)
        ? getPageSizeOptions(paginationPageSizeSelector)
        : getPageSizeOptions(DEFAULT_LIMITS), [paginationPageSizeSelector]);
    return (jsxs("div", { className: cx('flex items-center gap-2', className), children: [jsxs("p", { className: "font-semibold text-secondary text-sm", children: [label, ":"] }), jsx(_Select, { className: "w-20", defaultSelectedKey: String(paginationPageSize), "aria-labelledby": "select-items-per-page", items: limits, onSelectionChange: (value) => {
                    const limit = Number(value);
                    onPaginationChanged === null || onPaginationChanged === void 0 ? void 0 : onPaginationChanged({ newPageSize: limit });
                    onPageLimitChange === null || onPageLimitChange === void 0 ? void 0 : onPageLimitChange(limit);
                }, children: (item) => jsx(_Select.Item, { id: item.id, children: item.label }) })] }));
}

function TablePagination({ paginationCurrentPage, paginationTotalItems, paginationPageSize, paginationPageSizeSelector, paginationText, onPaginationChanged, onPageLimitChange, onPageNumberChange, t, }) {
    const texts = { ...DEFAULT_TEXTS.pagination, ...t === null || t === void 0 ? void 0 : t.pagination };
    const isDesktop = useBreakpoint('md');
    return (jsxs("div", { className: "flex w-full items-center justify-between gap-3 border-t border-secondary px-4 py-3 md:px-6 md:pt-3 md:pb-4", children: [paginationText ? (jsx("p", { className: "flex flex-1 font-semibold text-secondary text-sm", children: paginationText })) : paginationPageSizeSelector ? (jsx(PageSizeSelect, { paginationPageSize: paginationPageSize, paginationPageSizeSelector: paginationPageSizeSelector, onPaginationChanged: onPaginationChanged, onPageLimitChange: onPageLimitChange, label: texts.labelItemsPerPage })) : null, jsxs(Pagination.Root, { page: paginationCurrentPage, total: Math.ceil(paginationTotalItems / paginationPageSize), onPageChange: (page) => {
                    onPaginationChanged === null || onPaginationChanged === void 0 ? void 0 : onPaginationChanged({ newPage: page });
                    onPageNumberChange === null || onPageNumberChange === void 0 ? void 0 : onPageNumberChange(page);
                }, className: "flex items-center justify-between gap-3", children: [jsx("div", { className: "flex flex-1 justify-start", children: jsx(Pagination.PrevTrigger, { asChild: true, children: jsx(Button$1, { iconLeading: ArrowLeft, color: "secondary", size: "sm", children: isDesktop ? texts.labelPrevious : null }) }) }), jsx(Pagination.Context, { children: ({ pages }) => (jsx("div", { className: "hidden items-center justify-center gap-0.5 md:flex", children: pages.map((page, index) => page.type === 'page' ? (jsx(Pagination.Item, { asChild: true, ...page, children: jsx(Button$1, { color: "tertiary", size: "sm", className: cx('font-medium', page.isCurrent && 'bg-primary_hover text-secondary'), children: page.value }) }, index)) : (jsx(Pagination.Ellipsis, { className: "flex size-10 shrink-0 items-center justify-center text-tertiary", children: "\u2026" }, index))) })) }), jsx("div", { className: "flex flex-1 justify-end", children: jsx(Pagination.NextTrigger, { asChild: true, children: jsx(Button$1, { iconTrailing: ArrowRight, color: "secondary", size: "sm", children: isDesktop ? texts.labelNext : null }) }) })] }), paginationText && paginationPageSizeSelector ? (jsx(PageSizeSelect, { paginationPageSize: paginationPageSize, paginationPageSizeSelector: paginationPageSizeSelector, onPaginationChanged: onPaginationChanged, onPageLimitChange: onPageLimitChange, label: texts.labelItemsPerPage, className: "flex-1 justify-end" })) : null] }));
}

function TreeCell({ meta, itemId, collapsible, isCollapsed, toggleExpanded, children, treeLineLeft, 'data-id': dataId, }) {
    const { isChild, isLastChild, isParent } = meta;
    const showToggle = collapsible && isParent;
    const showTreeLines = (isParent && !isCollapsed) || isChild;
    const handleToggleClick = (e) => {
        e.stopPropagation();
        toggleExpanded === null || toggleExpanded === void 0 ? void 0 : toggleExpanded(itemId);
    };
    return (jsxs("div", { className: "flex items-center min-w-0", children: [jsxs("div", { className: cx('w-4 aspect-square flex', isChild ? 'mr-8' : 'mr-2'), children: [showToggle && (jsx(ButtonUtility, { "data-id": `${dataId}-toggle`, size: "xs", color: "tertiary", icon: isCollapsed ? ChevronRight : ChevronDown, className: "p-0", onClick: handleToggleClick })), showTreeLines && (jsxs(Fragment, { children: [jsx("div", { className: cx('absolute w-4 -translate-x-[0.5px] translate-y-[0.5px] bottom-0 border-l border-primary', showToggle && 'mt-2', isChild ? 'top-0' : 'top-1/2', isLastChild && 'bottom-1/2 border-b rounded-bl', treeLineLeft) }), isChild && !isLastChild && (jsx("div", { className: cx('absolute top-1/2 -translate-y-1/2 w-4 h-px bg-border-primary', treeLineLeft) }))] }))] }), jsx("div", { className: "min-w-0 flex-1", children: children })] }));
}

const FOOTER_ROW_ID = '__hostaway-table-footer__';
function toAriaDirection(dir) {
    return match(dir)
        .with('asc', () => 'ascending')
        .with('desc', () => 'descending')
        .with('ascending', () => 'ascending')
        .with('descending', () => 'descending')
        .exhaustive();
}
function toConsumerDirection(ariaDir, inputDir) {
    return match(inputDir)
        .with('asc', 'desc', () => (ariaDir === 'ascending' ? 'asc' : 'desc'))
        .with('ascending', 'descending', () => ariaDir)
        .exhaustive();
}
function Table({ getRowId, columnDefs = [], defaultColDef, rowData = [], tableCardHeader = null, pagination = false, paginationCurrentPage = 1, paginationTotalItems = 0, paginationPageSize = 20, paginationPageSizeSelector = true, onPaginationChanged = undefined, onPageNumberChange = undefined, onPageLimitChange = undefined, rowProps, emptyState, isLoading = false, columnsSelectorEnabled = false, columnsSelectorTabs, t, paginationText, onColumnsSelectionChange, columnOrder, onColumnOrderChange, columnsSelection, onColumnsSelectedChange, 'data-id': dataId, sortDescriptor, onSortChange, rounded = true, card = true, getChildRows, collapsible = false, showFooter = false, size = 'md', dragHandleVisible = false, ...tableProps }) {
    const texts = {
        loading: { ...DEFAULT_TEXTS.loading, ...t === null || t === void 0 ? void 0 : t.loading },
    };
    const isColumnOrderControlled = columnOrder !== undefined;
    const [internalColumnOrder, setInternalColumnOrder] = useState({});
    const resolvedColumnOrder = isColumnOrderControlled
        ? columnOrder
        : internalColumnOrder;
    const handleColumnOrderChange = (newOrder) => {
        if (!isColumnOrderControlled) {
            setInternalColumnOrder(newOrder);
        }
        onColumnOrderChange === null || onColumnOrderChange === void 0 ? void 0 : onColumnOrderChange(newOrder);
    };
    const { columns, setColumns, visibleColumns } = useColumns(columnDefs, columnsSelectorEnabled, resolvedColumnOrder, columnsSelection);
    const hasPinnedColumns = visibleColumns.some((c) => c.pinned);
    const stickyCallback = useStickyColumns(visibleColumns, {
        enabled: hasPinnedColumns,
    });
    const wrapperRef = useRef(null);
    const stickyRef = useCallback((node) => {
        stickyCallback(node);
        wrapperRef.current = node;
    }, [stickyCallback]);
    const { items, hasChildRows, collapsedKeys, toggleExpanded } = useTableItems(rowData, getRowId, getChildRows, collapsible);
    const itemsWithFooter = showFooter
        ? [
            ...items,
            {
                __row: undefined,
                __index: items.length,
                __meta: { isParent: false, isChild: false, isLastChild: false },
                id: FOOTER_ROW_ID,
            },
        ]
        : items;
    const isEmpty = !isLoading && (!rowData || rowData.length === 0);
    const rowHeight = TABLE_ROW_HEIGHT[size];
    // Restrict drag to the handle button only. Re-runs on isEmpty toggles
    // because the wrapper unmounts in the empty state, so the ref needs to
    // be re-attached when data arrives.
    useEffect(() => {
        if (!dragHandleVisible)
            return;
        const el = wrapperRef.current;
        if (!el)
            return;
        let pointerTarget = null;
        const onPointerDown = (e) => {
            pointerTarget = e.target;
        };
        const onDragStart = (e) => {
            if (!(pointerTarget === null || pointerTarget === void 0 ? void 0 : pointerTarget.closest('[data-drag-handle]'))) {
                e.preventDefault();
            }
        };
        const onDragOver = (e) => e.preventDefault();
        el.addEventListener('pointerdown', onPointerDown, { capture: true });
        el.addEventListener('dragstart', onDragStart, { capture: true });
        el.addEventListener('dragover', onDragOver);
        return () => {
            el.removeEventListener('pointerdown', onPointerDown, { capture: true });
            el.removeEventListener('dragstart', onDragStart, { capture: true });
            el.removeEventListener('dragover', onDragOver);
        };
    }, [dragHandleVisible, isEmpty]);
    return (jsxs(TableCard.Root, { size: size, className: cx('relative flex flex-col', !rounded && 'rounded-none', !card && 'shadow-none ring-0'), children: [tableCardHeader ? (jsx("div", { className: cx('relative flex flex-col items-start gap-4 border-b border-secondary bg-primary px-4 md:flex-row', TABLE_CARD_HEADER_PADDING[size]), children: tableCardHeader })) : null, isEmpty ? (jsx(TableEmptyState, { emptyState: emptyState })) : (jsxs("div", { className: "relative flex min-h-0 flex-1 flex-col", children: [columnsSelectorEnabled && (jsx("div", { className: cx('absolute right-0 z-51', size === 'sm' ? 'top-0.5' : 'top-1.5'), children: jsx(TableColumnsSelector, { columnDefs: columns, columnsSelection: columnsSelection, onColumnsSelectedChange: onColumnsSelectedChange, onColumnsSelectionChange: (cols) => {
                                setColumns(cols);
                                onColumnsSelectionChange === null || onColumnsSelectionChange === void 0 ? void 0 : onColumnsSelectionChange(cols);
                            }, columnsSelectorTabs: columnsSelectorTabs, columnOrder: resolvedColumnOrder, onColumnOrderChange: handleColumnOrderChange, t: t, "data-id": `${dataId}-columnSelector` }) })), jsx("div", { ref: stickyRef, className: cx('min-h-0 flex-1 [&>div]:h-full [&>div]:overflow-auto! [&_thead]:sticky [&_thead]:top-0 [&_thead]:z-50', columnsSelectorEnabled && '[&_thead_th:last-child]:pr-15'), style: {
                            '--row-height': `${rowHeight}px`,
                        }, children: jsxs(Table$1, { ...tableProps, size: size, disabledKeys: [FOOTER_ROW_ID], sortDescriptor: sortDescriptor
                                ? {
                                    ...sortDescriptor,
                                    direction: toAriaDirection(sortDescriptor.direction),
                                }
                                : undefined, onSortChange: onSortChange
                                ? (ariaDescriptor) => {
                                    onSortChange({
                                        column: ariaDescriptor.column,
                                        direction: sortDescriptor
                                            ? toConsumerDirection(ariaDescriptor.direction, sortDescriptor.direction)
                                            : ariaDescriptor.direction,
                                    });
                                }
                                : undefined, className: cx('[&_tbody_tr]:[content-visibility:auto] [&_tbody_tr]:[contain-intrinsic-block-size:var(--row-height)]', DROP_INDICATOR_CLASSES, !dragHandleVisible &&
                                '[&[data-allows-dragging]_tbody_tr]:cursor-grab'), children: [jsx(Table$1.Header, { children: visibleColumns.map(({ cellRenderer: _, ...col }) => (jsx(TableHead, { ...defaultColDef, ...col, className: (state) => cx(TABLE_HEAD_PADDING[size], typeof col.className === 'function'
                                            ? col.className(state)
                                            : col.className), "data-id": `${dataId}-head-${col.id}` }, col.id))) }), jsx(Table$1.Body, { items: isLoading ? [] : itemsWithFooter, children: (item) => {
                                        if (item.id === FOOTER_ROW_ID) {
                                            return (jsx(Table$1.Row, { id: item.id, className: "bg-secondary hover:bg-secondary sticky bottom-0 z-20 !h-11 font-semibold uppercase", isFooterRow: true, children: visibleColumns.map((col, colIndex) => {
                                                    var _a, _b;
                                                    if (!col.id)
                                                        return jsx(Table$1.Cell, {}, colIndex);
                                                    const cellContent = (_b = (_a = col.footerCell) === null || _a === void 0 ? void 0 : _a.call(col)) !== null && _b !== void 0 ? _b : null;
                                                    return (jsx(Table$1.Cell, { "data-id": `${dataId}-footer-cell-${col.id}`, className: TABLE_FOOTER_CELL_PADDING[size], children: cellContent }, col.id));
                                                }) }));
                                        }
                                        const { __meta: meta } = item;
                                        const isCollapsed = collapsible && !!(collapsedKeys === null || collapsedKeys === void 0 ? void 0 : collapsedKeys.has(item.id));
                                        const isLastDataRow = item.__index === items.length - 1;
                                        return (jsx(Table$1.Row, { ...rowProps, "data-child-row": meta.isChild || undefined, "data-last-child": meta.isLastChild || undefined, className: (state) => cx(typeof (rowProps === null || rowProps === void 0 ? void 0 : rowProps.className) === 'function'
                                                ? rowProps.className(state)
                                                : rowProps === null || rowProps === void 0 ? void 0 : rowProps.className, tableProps.onRowAction && 'cursor-pointer', meta.isParent &&
                                                !isCollapsed &&
                                                '[&>td]:after:bg-transparent', meta.isChild &&
                                                !meta.isLastChild &&
                                                '[&>td]:after:bg-transparent', showFooter &&
                                                isLastDataRow &&
                                                '[&>td]:after:bg-transparent'), id: item.id, children: visibleColumns.map((col, colIndex) => {
                                                if (!col.id)
                                                    return jsx(Table$1.Cell, {}, colIndex);
                                                let cellContent;
                                                if (typeof col.cellRenderer === 'function') {
                                                    cellContent = col.cellRenderer(item.__row);
                                                }
                                                else {
                                                    const value = typeof item.__row === 'object'
                                                        ? item.__row[col.id]
                                                        : item.__row;
                                                    cellContent = isValidElement(value)
                                                        ? value
                                                        : String(value);
                                                }
                                                const isSelectionMode = tableProps.selectionMode !== undefined &&
                                                    tableProps.selectionMode !== 'none';
                                                if (colIndex === 0 && hasChildRows) {
                                                    cellContent = (jsx(TreeCell, { meta: meta, itemId: item.id, collapsible: collapsible, isCollapsed: isCollapsed, toggleExpanded: toggleExpanded, treeLineLeft: TREE_LINE_INDENT(size, isSelectionMode, dragHandleVisible), "data-id": `${dataId}-tree-cell-${item.id}`, children: cellContent }));
                                                }
                                                const showDragHandle = dragHandleVisible && colIndex === 0;
                                                if (showDragHandle) {
                                                    return (jsx(Table$1.Cell, { className: TABLE_CELL_PADDING[size], children: jsxs("div", { className: cx('flex items-center', col.size && COL_SIZE_CLASS[col.size], meta.isChild && 'pl-6'), children: [!meta.isChild && (jsx(ButtonUtility, { "data-id": `${dataId}-drag-${item.id}`, icon: Menu01, size: "xs", slot: "drag", color: "tertiary", "data-drag-handle": true, style: { pointerEvents: 'auto' }, className: "p-0 pr-2 cursor-grab data-pressed:cursor-grabbing" })), cellContent] }) }, col.id));
                                                }
                                                return (jsx(Table$1.Cell, { className: TABLE_CELL_PADDING[size], children: jsx("div", { className: cx('break-words', col.size && COL_SIZE_CLASS[col.size]), children: cellContent }) }, col.id));
                                            }) }));
                                    } })] }) }), isLoading ? (jsx("div", { className: "flex items-center justify-center py-10", children: jsx(LoadingIndicator, { "data-id": `${dataId}-loading`, size: "sm", label: texts.loading.labelLoading }) })) : null] })), pagination ? (jsx(TablePagination, { paginationCurrentPage: paginationCurrentPage, paginationTotalItems: paginationTotalItems, paginationPageSize: paginationPageSize, paginationPageSizeSelector: paginationPageSizeSelector, paginationText: paginationText, onPaginationChanged: onPaginationChanged, onPageNumberChange: onPageNumberChange, onPageLimitChange: onPageLimitChange, t: t })) : null] }));
}

const MAX_INLINE_ACTIONS = 2;
function isCustom(item) {
    return 'type' in item && item.type === 'custom';
}
function TableActions({ actions, 'data-id': dataId, 'aria-label': ariaLabel = 'Row actions', maxInlineActions = MAX_INLINE_ACTIONS, }) {
    const inlineActions = actions.slice(0, maxInlineActions);
    const overflowActions = actions.slice(maxInlineActions);
    const isSeparator = (action) => 'type' in action;
    return (jsxs("div", { role: "group", "aria-label": ariaLabel, className: "flex justify-end gap-1", children: [inlineActions.map((action) => {
                if (isCustom(action)) {
                    return action.render();
                }
                return !isSeparator(action) ? (jsx(ButtonUtility, { "data-id": action.dataId, icon: action.icon, tooltip: action.label, color: "tertiary", onClick: action.onClick, isDisabled: action.isDisabled }, action.dataId)) : null;
            }), overflowActions.length > 0 && (jsxs(Dropdown.Root, { children: [jsx(Dropdown.DotsButton, { "data-id": `${dataId}-trigger`, horizontal: true }), jsx(Dropdown.Popover, { className: "w-min", children: jsx(Dropdown.Menu, { children: overflowActions.map((action, index) => {
                                if (isSeparator(action)) {
                                    return (jsx(Dropdown.Separator, {}, `${dataId}-separator-${index}`));
                                }
                                if (isCustom(action)) {
                                    return (jsxs(Dropdown.Item, { "data-id": action.dataId, unstyled: true, children: [action.render(), ";"] }, `${dataId}-custom-action-${index}`));
                                }
                                return (jsx(Dropdown.Item, { "data-id": action.dataId, icon: action.icon, onClick: action.onClick, isDisabled: action.isDisabled, tooltip: action.tooltip, children: action.label }, action.dataId));
                            }) }) })] }))] }));
}

export { Table, TableActions };
