// @hostaway/design-system/dist/components/Slider/index.cjs.js
'use strict';

var jsxRuntime = require('react/jsx-runtime');
var reactAriaComponents = require('react-aria-components');
var cx = require('../_shared/cx-CmkbyMOt.cjs.js');
require('tailwind-merge');

const styles = cx.sortCx({
    default: 'hidden',
    bottom: 'absolute top-2 left-1/2 -translate-x-1/2 translate-y-full text-md font-medium text-primary',
    'top-floating': 'absolute -top-2 left-1/2 -translate-x-1/2 -translate-y-full rounded-lg bg-primary px-3 py-2 text-xs font-semibold text-secondary shadow-lg ring-1 ring-secondary_alt',
});
const Slider$1 = ({ labelPosition = 'default', minValue = 0, maxValue = 100, labelFormatter, formatOptions, ...rest }) => {
    // Format thumb value as percentage by default.
    const defaultFormatOptions = {
        style: 'percent',
        maximumFractionDigits: 0,
    };
    return (jsxRuntime.jsxs(reactAriaComponents.Slider, { ...rest, minValue, maxValue, formatOptions: formatOptions !== null && formatOptions !== void 0 ? formatOptions : defaultFormatOptions, children: [jsxRuntime.jsx(reactAriaComponents.Label, {}), jsxRuntime.jsx(reactAriaComponents.SliderTrack, { className: "relative h-6 w-full", children: ({ state: { values, getThumbValue, getThumbPercent, getFormattedValue }, }) => {
                    const left = values.length === 1 ? 0 : getThumbPercent(0);
                    const width = values.length === 1
                        ? getThumbPercent(0)
                        : getThumbPercent(1) - left;
                    return (jsxRuntime.jsxs(jsxRuntime.Fragment, { children: [jsxRuntime.jsx("span", { className: "absolute top-1/2 h-2 w-full -translate-y-1/2 rounded-full bg-quaternary" }), jsxRuntime.jsx("span", { className: "absolute top-1/2 h-2 w-full -translate-y-1/2 rounded-full bg-brand-solid", style: {
                                    left: `${left * 100}%`,
                                    width: `${width * 100}%`,
                                } }), values.map((_, index) => {
                                return (jsxRuntime.jsx(reactAriaComponents.SliderThumb, { index: index, className: ({ isFocusVisible, isDragging }) => cx.cx('top-1/2 box-border size-6 cursor-grab rounded-full bg-slider-handle-bg shadow-md ring-2 ring-slider-handle-border ring-inset', isFocusVisible &&
                                        'outline-2 outline-offset-2 outline-focus-ring', isDragging && 'cursor-grabbing'), children: jsxRuntime.jsx(reactAriaComponents.SliderOutput, { className: cx.cx('whitespace-nowrap', styles[labelPosition]), children: labelFormatter
                                            ? labelFormatter(getThumbValue(index))
                                            : getFormattedValue(getThumbValue(index) / 100) }) }, index));
                            })] }));
                } })] }));
};

function Slider(props) {
    return jsxRuntime.jsx(Slider$1, { ...props });
}

exports.Slider = Slider;
