// @hostaway/design-system/dist/components/Filter/index.cjs.js
'use strict';

var jsxRuntime = require('react/jsx-runtime');
var h = require('react');
var reactAriaComponents = require('react-aria-components');
var FilterLines = require('@untitledui/icons/FilterLines');
var SlashCircle01 = require('@untitledui/icons/SlashCircle01');
var CheckCircle = require('@untitledui/icons/CheckCircle');
var ChevronDown = require('@untitledui/icons/ChevronDown');
var XClose = require('@untitledui/icons/XClose');
var badges = require('../_shared/badges-xpScvf89.cjs.js');
var RadioButton = require('../_shared/RadioButton-xPnVW2w3.cjs.js');
var SearchLg = require('@untitledui/icons/SearchLg');
var cx = require('../_shared/cx-CmkbyMOt.cjs.js');
var InputField = require('../_shared/InputField-DaibKVKA.cjs.js');
var Checkbox = require('../_shared/Checkbox-lgLTtE3V.cjs.js');
var Button = require('../_shared/Button-BE28ft73.cjs.js');
var Tooltip = require('../_shared/Tooltip-CymVDnjP.cjs.js');
require('@untitledui/icons');
require('../_shared/dot-icon-C5ORy9WX.cjs.js');
require('../_shared/label-D0WtNYRO.cjs.js');
require('../_shared/MyCustomIcon-DYQF4lJT.cjs.js');
require('../_shared/Icon-BF_rHwZP.cjs.js');
require('../_shared/tooltip-DDB8niG0.cjs.js');
require('tailwind-merge');
require('../_shared/input-CVSUknGT.cjs.js');
require('../_shared/hint-text-LRIVuQnk.cjs.js');
require('../_shared/FieldWrapper-D6dXzMyx.cjs.js');
require('../_shared/ConditionalTooltip-FLw9yEoi.cjs.js');
require('../_shared/checkbox-D6ve3vaM.cjs.js');
require('../_shared/BaseButton-CchEmlms.cjs.js');
require('../_shared/is-react-component-DkG20B86.cjs.js');

function useChipOverflow({ containerRef, chipCount, chipOverflow, isExpanded, }) {
    const [visibleCount, setVisibleCount] = h.useState(chipOverflow && !isExpanded ? 0 : chipCount);
    const [clipWidth, setClipWidth] = h.useState(undefined);
    const [measureTick, setMeasureTick] = h.useState(0);
    const prevExpandedRef = h.useRef(isExpanded);
    // Track the last two measured counts to detect oscillation (e.g. 3 → 2 → 3 → 2).
    // When the count alternates between two values at a boundary width, we settle on
    // the lower value to stop the infinite resize loop.
    const prevCountRef = h.useRef(null);
    const prevPrevCountRef = h.useRef(null);
    const prevChipCountRef = h.useRef(chipCount);
    if (prevChipCountRef.current !== chipCount) {
        prevChipCountRef.current = chipCount;
        prevCountRef.current = null;
        prevPrevCountRef.current = null;
    }
    const measure = h.useCallback(() => {
        const container = containerRef.current;
        if (!container || !chipOverflow || isExpanded) {
            prevCountRef.current = null;
            prevPrevCountRef.current = null;
            setVisibleCount(chipCount);
            setClipWidth(undefined);
            return;
        }
        const chipElements = Array.from(container.children);
        if (chipElements.length === 0) {
            prevCountRef.current = null;
            prevPrevCountRef.current = null;
            setVisibleCount(chipCount);
            setClipWidth(undefined);
            return;
        }
        // Temporarily remove clip constraint so we measure against the full available width
        const prevMaxWidth = container.style.maxWidth;
        container.style.maxWidth = 'none';
        const containerRect = container.getBoundingClientRect();
        let count = 0;
        for (const chip of chipElements) {
            if (chip.getBoundingClientRect().right <= containerRect.right) {
                count++;
            }
            else {
                break;
            }
        }
        // Detect oscillation: if the count is bouncing between two values,
        // settle on the lower one to break the loop.
        if (prevPrevCountRef.current !== null &&
            count === prevPrevCountRef.current &&
            count !== prevCountRef.current) {
            count = Math.min(count, prevCountRef.current);
        }
        prevPrevCountRef.current = prevCountRef.current;
        prevCountRef.current = count;
        // Compute a maxWidth that ends exactly at the last visible chip
        const newClipWidth = count > 0 && count < chipElements.length
            ? chipElements[count - 1].getBoundingClientRect().right -
                containerRect.left
            : undefined;
        // Restore before React re-renders
        container.style.maxWidth = prevMaxWidth;
        setVisibleCount(count);
        setClipWidth(newClipWidth);
    }, [containerRef, chipCount, chipOverflow, isExpanded]);
    h.useLayoutEffect(() => {
        // When transitioning from isExpanded → collapsed, reset visibleCount to 0
        // so the "+X more" button renders and takes space during measurement.
        // Then bump measureTick to trigger a second pass that actually measures.
        if (prevExpandedRef.current && !isExpanded && chipOverflow) {
            prevExpandedRef.current = isExpanded;
            setVisibleCount(0);
            setClipWidth(undefined);
            setMeasureTick((t) => t + 1);
            return;
        }
        prevExpandedRef.current = isExpanded;
        measure();
    }, [measure, isExpanded, chipOverflow, measureTick]);
    h.useEffect(() => {
        const container = containerRef.current;
        if (!container || !chipOverflow)
            return;
        let rafId;
        const observer = new ResizeObserver(() => {
            cancelAnimationFrame(rafId);
            rafId = requestAnimationFrame(() => {
                measure();
            });
        });
        observer.observe(container);
        return () => {
            cancelAnimationFrame(rafId);
            observer.disconnect();
        };
    }, [containerRef, chipOverflow, measure]);
    const hiddenCount = chipOverflow && !isExpanded ? Math.max(0, chipCount - visibleCount) : 0;
    return { visibleCount, hiddenCount, clipWidth };
}

function deepEqual(a, b) {
    if (a === b)
        return true;
    if (a === null || b === null)
        return false;
    if (typeof a !== 'object' || typeof b !== 'object')
        return false;
    if (Array.isArray(a) !== Array.isArray(b))
        return false;
    const keysA = Object.keys(a);
    const keysB = Object.keys(b);
    if (keysA.length !== keysB.length)
        return false;
    return keysA.every((key) => deepEqual(a[key], b[key]));
}

function isEmptyObject(value) {
    if (Array.isArray(value))
        return value.length === 0;
    const keys = Object.keys(value);
    if (keys.length === 0)
        return true;
    return keys.every((k) => {
        const v = value[k];
        if (Array.isArray(v)) {
            return v.length === 0;
        }
        return !v;
    });
}

const FILTER_VARIANTS = {
    DEFAULT: 'default',
    MINIMAL: 'minimal',
};
function isFieldApplied(value, emptyValue) {
    if (emptyValue !== undefined)
        return !deepEqual(value, emptyValue);
    if (value == null)
        return false;
    if (value === '')
        return false;
    if (Array.isArray(value))
        return value.length > 0;
    if (typeof value === 'object')
        return !isEmptyObject(value);
    return true;
}
function getValueCount(value) {
    if (Array.isArray(value))
        return value.length;
    if (value !== null && typeof value === 'object') {
        // Count non-nullish keys as "active" entries
        return Object.values(value).filter((v) => v != null && v !== '').length;
    }
    return 1;
}
function Filter({ initValues, groups = {}, className, onSubmit, onClearAll, onFieldChange, chipOverflow = false, setMoreLabel, isEditable = true, alwaysShowClearAll = false, variant = FILTER_VARIANTS.DEFAULT, t = { ...defaultTexts }, 'data-id': dataId, }) {
    var _a;
    const [isOpen, setIsOpen] = h.useState(false);
    const [values, setValues] = h.useState(() => ({ ...initValues }));
    const [selectedGroup, setSelectedGroup] = h.useState(() => Object.keys(groups)[0] || '');
    const [isExpanded, setIsExpanded] = h.useState(false);
    const chipsRef = h.useRef(null);
    const prevInitValuesRef = h.useRef(initValues);
    const applied = h.useMemo(() => Object.entries(groups).flatMap(([groupKey, group]) => {
        return group.fields.flatMap((field) => {
            var _a, _b;
            const fieldValue = initValues[field.name];
            const { Component, props } = field.field;
            const baseChip = {
                fieldName: field.name,
                Component,
                isEditable,
                fieldProps: props,
                onClose: () => {
                    setValues((current) => {
                        const changed = !deepEqual(current[field.name], fieldValue);
                        if (changed)
                            onSubmit({ ...current });
                        return current;
                    });
                },
            };
            if (field.chips) {
                const descriptors = field.chips(fieldValue);
                if (descriptors.length === 0)
                    return [];
                return descriptors.map((desc) => {
                    var _a;
                    return ({
                        ...baseChip,
                        Component: (_a = desc.Component) !== null && _a !== void 0 ? _a : baseChip.Component,
                        id: `${groupKey}-${String(field.name)}-${desc.key}`,
                        label: desc.label,
                        chevron: desc.chevron,
                        onRemove: desc.partialEmpty
                            ? () => {
                                const merged = {
                                    ...fieldValue,
                                    ...desc.partialEmpty,
                                };
                                const next = field.emptyValue !== undefined &&
                                    deepEqual(merged, field.emptyValue)
                                    ? field.emptyValue
                                    : merged;
                                onSubmit({ ...initValues, [field.name]: next });
                            }
                            : undefined,
                    });
                });
            }
            if (!isFieldApplied(fieldValue, field.emptyValue))
                return [];
            const count = getValueCount(fieldValue);
            const labelSuffix = ` (${count})`;
            const listProps = props;
            const resolvedLabel = (listProps === null || listProps === void 0 ? void 0 : listProps.isMulti) ? (jsxRuntime.jsxs(jsxRuntime.Fragment, { children: [group.label, jsxRuntime.jsx("span", { className: "text-quaternary", children: labelSuffix })] })) : (((_b = (_a = listProps === null || listProps === void 0 ? void 0 : listProps.options) === null || _a === void 0 ? void 0 : _a.find((item) => deepEqual(item.value, fieldValue))) === null || _b === void 0 ? void 0 : _b.label) || group.label);
            return [
                {
                    ...baseChip,
                    id: `${groupKey}-${String(field.name)}`,
                    label: resolvedLabel,
                    chevron: false,
                    onRemove: () => {
                        var _a;
                        onSubmit({
                            ...initValues,
                            [field.name]: (_a = field.emptyValue) !== null && _a !== void 0 ? _a : null,
                        });
                    },
                },
            ];
        });
    }), [groups, initValues, onSubmit, isEditable]);
    const { hiddenCount, clipWidth } = useChipOverflow({
        containerRef: chipsRef,
        chipCount: applied.length,
        chipOverflow,
        isExpanded,
    });
    h.useEffect(() => {
        if (applied.length === 0)
            setIsExpanded(false);
    }, [applied.length]);
    const handleClearAll = () => {
        const clearedValues = {};
        Object.keys(groups).forEach((groupKey) => {
            const group = groups[groupKey];
            group.fields.forEach((field) => {
                if (field.emptyValue !== undefined) {
                    clearedValues[field.name] = field.emptyValue;
                }
                else {
                    const current = initValues[field.name];
                    if (typeof current === 'object' &&
                        current !== null &&
                        !Array.isArray(current)) {
                        // Reset each key to null, preserving object shape
                        clearedValues[field.name] = Object.fromEntries(Object.keys(current).map((k) => [k, null]));
                    }
                    else {
                        clearedValues[field.name] = null;
                    }
                }
            });
        });
        if (onClearAll) {
            onClearAll();
        }
        else {
            onSubmit(clearedValues);
        }
        // Reset values so reopening the panel shows cleared state immediately,
        // even if the consumer updates initValues asynchronously after onClearAll.
        setValues(clearedValues);
        setIsOpen(false);
    };
    const handleApply = () => {
        onSubmit(values);
        setIsOpen(false);
    };
    h.useEffect(() => {
        if (prevInitValuesRef.current !== initValues) {
            prevInitValuesRef.current = initValues;
            setValues({ ...initValues });
        }
    }, [initValues]);
    return (jsxRuntime.jsxs("div", { className: cx.cx('flex gap-2 items-start', chipOverflow && isExpanded && 'flex-wrap', className), children: [jsxRuntime.jsxs(reactAriaComponents.MenuTrigger, { isOpen: isOpen, onOpenChange: setIsOpen, children: [isEditable && (jsxRuntime.jsx(Button.Button, { iconLeading: FilterLines.FilterLines, color: "secondary", className: cx.cx(variant === FILTER_VARIANTS.MINIMAL && 'h-10'), onPress: () => setIsOpen(true), "data-id": `${dataId}-trigger`, "aria-label": variant === FILTER_VARIANTS.MINIMAL
                            ? `${t.btnFilter}, ${applied.length}`
                            : undefined, children: variant === FILTER_VARIANTS.MINIMAL ? (jsxRuntime.jsx(badges.Badge, { color: applied.length > 0 ? 'brand' : 'gray', className: "h-5 leading-none", children: applied.length })) : (t.btnFilter) })), jsxRuntime.jsxs(reactAriaComponents.Popover, { containerPadding: 0, className: "rounded-lg bg-primary shadow-lg ring ring-secondary_alt", children: [jsxRuntime.jsxs("div", { className: "grid grid-cols-[250px_250px] border-b border-secondary_alt", children: [jsxRuntime.jsx(reactAriaComponents.Menu, { className: "flex flex-col gap-0.5 border-r border-secondary_alt px-1.5 py-1 outline-hidden select-none overflow-y-auto min-h-82 max-h-120", selectionMode: "multiple", children: Object.keys(groups).map((key) => {
                                            var _a, _b;
                                            const group = groups[key];
                                            const count = (_b = (_a = group === null || group === void 0 ? void 0 : group.fields) === null || _a === void 0 ? void 0 : _a.reduce((acc, field) => {
                                                const value = values[field.name];
                                                if (field.chips) {
                                                    return acc + field.chips(value).length;
                                                }
                                                if (!isFieldApplied(value, field.emptyValue))
                                                    return acc;
                                                if (field.count) {
                                                    const fieldCount = field.count(value);
                                                    return acc + fieldCount;
                                                }
                                                if (Array.isArray(value))
                                                    return acc + value.length;
                                                if (value !== null && typeof value === 'object') {
                                                    // Count non-nullish keys as "active" entries
                                                    return Object.values(value).filter((v) => v != null && v !== '').length;
                                                }
                                                return acc + 1;
                                            }, 0)) !== null && _b !== void 0 ? _b : 0;
                                            return (jsxRuntime.jsxs(reactAriaComponents.MenuItem, { id: key, className: cx.cx('flex justify-between cursor-pointer rounded-md px-2.5 py-2 hover:bg-primary_hover font-normal w-full text-left text-secondary text-sm outline-hidden', selectedGroup === key &&
                                                    'bg-primary_hover text-secondary_hover'), onPress: () => setSelectedGroup(key), children: [jsxRuntime.jsx("span", { children: groups[key].label }), jsxRuntime.jsx(badges.Badge, { type: "modern", className: cx.cx('invisible', count > 0 && 'visible'), children: count })] }, key));
                                        }) }), jsxRuntime.jsx("div", { className: "px-2 py-2 min-h-82 max-h-120 overflow-y-auto", children: jsxRuntime.jsx("div", { className: "flex flex-col gap-4", children: (_a = groups[selectedGroup]) === null || _a === void 0 ? void 0 : _a.fields.map((field) => {
                                                const { Component, props } = field.field;
                                                if (!Component)
                                                    return null;
                                                return (jsxRuntime.jsx(Component, { ...props, "data-id": `${dataId}-${String(field.name)}`, context: "panel", value: values[field.name], onChange: (value) => {
                                                        setValues((prev) => {
                                                            const updated = { ...prev, [field.name]: value };
                                                            const extra = onFieldChange === null || onFieldChange === void 0 ? void 0 : onFieldChange(field.name, value, prev);
                                                            return extra ? { ...updated, ...extra } : updated;
                                                        });
                                                    } }, String(field.name)));
                                            }) }) })] }), jsxRuntime.jsxs("div", { className: "flex p-4 gap-4", children: [jsxRuntime.jsx(Button.Button, { iconLeading: SlashCircle01.SlashCircle01, color: "secondary", size: "sm", className: "w-full", onPress: handleClearAll, "data-id": `${dataId}-clearAll`, children: t.btnClearAll }), jsxRuntime.jsx(Button.Button, { iconLeading: CheckCircle.CheckCircle, size: "sm", className: "w-full", onPress: handleApply, "data-id": `${dataId}-apply`, children: t.btnApply })] })] })] }), variant !== FILTER_VARIANTS.MINIMAL &&
                applied.length > 0 &&
                isEditable && (jsxRuntime.jsx("div", { "aria-hidden": "true", className: "h-7 w-px bg-border-secondary self-center shrink-0" })), variant !== FILTER_VARIANTS.MINIMAL && applied.length > 0 ? (jsxRuntime.jsxs("div", { className: cx.cx('flex items-center gap-2', chipOverflow && !isExpanded && 'min-w-0', chipOverflow && isExpanded && 'contents'), children: [jsxRuntime.jsx("div", { ref: chipsRef, role: "group", "aria-label": t.btnAppliedFilters, className: cx.cx('flex items-center gap-2', chipOverflow && !isExpanded
                            ? 'flex-1 flex-nowrap overflow-clip [overflow-clip-margin:6px] min-w-0'
                            : 'contents'), style: chipOverflow && !isExpanded && clipWidth !== undefined
                            ? { maxWidth: clipWidth }
                            : undefined, children: applied.map((chip) => (jsxRuntime.jsxs("div", { className: "relative", children: [jsxRuntime.jsxs(reactAriaComponents.MenuTrigger, { onOpenChange: (open) => {
                                        if (!open)
                                            chip.onClose();
                                    }, children: [jsxRuntime.jsx(Button.Button, { color: "secondary", "data-id": `${dataId}-chipButton`, iconTrailing: isEditable && !chip.onRemove && chip.chevron
                                                ? ChevronDown.ChevronDown
                                                : undefined, children: jsxRuntime.jsx("span", { className: isEditable && chip.onRemove ? 'mr-6' : '', children: chip.label }) }), jsxRuntime.jsx(reactAriaComponents.Popover, { placement: "bottom left", className: "min-w-60 max-w-80 rounded-lg bg-primary shadow-lg ring ring-secondary_alt", children: jsxRuntime.jsx("div", { className: "px-1.5 py-1 max-h-92 overflow-y-auto", children: jsxRuntime.jsx(ChipPopoverContent, { chip: chip, value: values[chip.fieldName], setValues: setValues, onFieldChange: onFieldChange, "data-id": `${dataId}-chip-${String(chip.fieldName)}`, isEditable: isEditable }) }) })] }), isEditable && chip.onRemove && (jsxRuntime.jsx(reactAriaComponents.Button, { "aria-label": t.btnRemoveFilter, className: "absolute right-3 top-[calc(50%_-_10px)] z-10 flex cursor-pointer items-center justify-center rounded-lg p-0 text-fg-quaternary", onPress: chip.onRemove, children: jsxRuntime.jsx(XClose.XClose, { "aria-hidden": "true", className: "size-5" }) }))] }, chip.id))) }), chipOverflow && (hiddenCount > 0 || isExpanded) && (jsxRuntime.jsx(Button.Button, { color: "tertiary", className: "shrink-0", onPress: () => setIsExpanded((prev) => !prev), "data-id": `${dataId}-toggleExpand`, children: isExpanded
                            ? t.btnCollapse
                            : setMoreLabel
                                ? setMoreLabel(hiddenCount)
                                : `+${hiddenCount} more` })), isEditable && (jsxRuntime.jsx(Button.Button, { "aria-label": `${t.btnClearAll} ${t.btnAppliedFilters}`.toLowerCase(), color: "tertiary", className: "shrink-0", onPress: handleClearAll, "data-id": `${dataId}-chipsClearAll`, children: t.btnClearAll }))] })) : null, applied.length === 0 && isEditable && alwaysShowClearAll && (jsxRuntime.jsxs(jsxRuntime.Fragment, { children: [jsxRuntime.jsx("div", { "aria-hidden": "true", className: "h-7 w-px bg-border-secondary self-center shrink-0" }), jsxRuntime.jsx(Button.Button, { "aria-label": `${t.btnClearAll} ${t.btnAppliedFilters}`.toLowerCase(), color: "tertiary", className: "shrink-0", onPress: handleClearAll, "data-id": `${dataId}-chipsClearAll`, children: t.btnClearAll })] }))] }));
}
function ChipPopoverContent({ chip, value, setValues, onFieldChange, 'data-id': dataId, isEditable, }) {
    return (jsxRuntime.jsx(chip.Component, { ...chip.fieldProps, "data-id": dataId, context: "chip", value: value, onChange: (newValue) => {
            if (!isEditable)
                return;
            setValues((prev) => {
                const updated = { ...prev, [chip.fieldName]: newValue };
                const extra = onFieldChange === null || onFieldChange === void 0 ? void 0 : onFieldChange(chip.fieldName, newValue, prev);
                return extra ? { ...updated, ...extra } : updated;
            });
        } }));
}
// Sized to fit inside the panel (max-h-120 wrapper) and chip (max-h-92 wrapper)
// popovers respectively. Keep in sync with those wrappers.
const LIST_MAX_H = {
    panel: 'max-h-116',
    chip: 'max-h-80',
};
function List({ options = [], onChange, value, isMulti = false, withSearch = false, searchPlaceholder, context, 'data-id': dataId, }) {
    // Chip popover should close on pick when single-select; multi-select stays
    // open so users can toggle several. Panel context always keeps the inner
    // menu in "multiple" mode to avoid bubbling close-on-action up to the
    // surrounding panel MenuTrigger.
    const shouldAutoClose = !isMulti && context === 'chip';
    const [searchTerm, setSearchTerm] = h.useState('');
    const filteredOptions = options.filter((item) => {
        return (item.label.toLowerCase().includes(searchTerm.trim().toLowerCase()) ||
            (item.supportingText &&
                item.supportingText
                    .toLowerCase()
                    .includes(searchTerm.trim().toLowerCase())));
    });
    const multiValues = Array.isArray(value)
        ? value
        : [];
    return (jsxRuntime.jsxs("div", { className: cx.cx('flex flex-col gap-1', LIST_MAX_H[context !== null && context !== void 0 ? context : 'chip']), children: [withSearch && (jsxRuntime.jsx(InputField.InputField, { placeholder: searchPlaceholder, icon: SearchLg.SearchLg, value: searchTerm, onChange: setSearchTerm, "data-id": `${dataId}-searchInput` })), jsxRuntime.jsx(reactAriaComponents.RadioGroup, { value: value != null ? String(value) : '', onChange: () => { }, "aria-hidden": "true", className: "contents", children: jsxRuntime.jsx(reactAriaComponents.Menu, { className: "h-min overflow-y-auto overscroll-none outline-hidden select-none flex flex-col gap-0.5", selectionMode: shouldAutoClose ? 'single' : 'multiple', selectedKeys: multiValues, children: filteredOptions.map((item) => {
                        const isSelected = isMulti
                            ? multiValues.includes(item.value)
                            : value === item.value;
                        return (jsxRuntime.jsxs(reactAriaComponents.MenuItem, { id: item.value, className: cx.cx('w-full px-2.5 py-2 cursor-pointer outline-hidden flex items-center gap-2 text-secondary hover:bg-primary_hover rounded-md', isSelected && 'bg-primary_hover text-secondary_hover'), onPress: () => {
                                if (isMulti) {
                                    const newValue = isSelected
                                        ? multiValues.filter((v) => v !== item.value)
                                        : [...multiValues, item.value];
                                    onChange(newValue.length > 0 ? newValue : null);
                                }
                                else {
                                    const updatedValue = (isSelected ? null : item.value);
                                    onChange(updatedValue);
                                }
                            }, "data-id": `${dataId}-option-${item.value}`, children: [isMulti ? (jsxRuntime.jsx(Checkbox.Checkbox, { isSelected: isSelected, onChange: () => { }, className: "pointer-events-none shrink-0", "aria-hidden": "true", "data-id": `${dataId}-checkbox-${item.value}` })) : (jsxRuntime.jsx(RadioButton.RadioButton, { value: String(item.value), className: "pointer-events-none shrink-0", "aria-hidden": "true", "data-id": `${dataId}-radio-${item.value}` })), jsxRuntime.jsxs("div", { className: "flex flex-col gap-y-1 text-sm flex-1 min-w-0", children: [jsxRuntime.jsx("span", { className: "font-medium text-primary", children: item.label }), item.supportingText && (jsxRuntime.jsx(Tooltip.Tooltip, { "data-id": `${dataId}-option-supporting-text-tooltip-${item.value}`, title: item.supportingText, triggerProps: {
                                                className: 'w-full overflow-hidden text-ellipsis text-nowrap text-left',
                                            }, children: jsxRuntime.jsx("span", { className: "font-normal text-quaternary", children: item.supportingText }) }))] })] }, item.value));
                    }) }) })] }));
}
Filter.List = List;
const defaultTexts = {
    btnFilter: 'Filter',
    btnClearAll: 'Clear all',
    btnApply: 'Apply filters',
    btnCollapse: 'Collapse',
    btnRemoveFilter: 'Remove filter',
    btnAppliedFilters: 'Applied filters',
};

exports.FILTER_VARIANTS = FILTER_VARIANTS;
exports.Filter = Filter;
