// @hostaway/design-system/dist/components/Tag/index.cjs.js
'use strict';

var jsxRuntime = require('react/jsx-runtime');
var h = require('react');
var reactAriaComponents = require('react-aria-components');
var avatar = require('../_shared/avatar-C5INEgx9.cjs.js');
var dotIcon = require('../_shared/dot-icon-C5ORy9WX.cjs.js');
var cx = require('../_shared/cx-CmkbyMOt.cjs.js');
var tagCloseX = require('../_shared/tag-close-x-DKfF84Fm.cjs.js');
require('@untitledui/icons');
require('tailwind-merge');

const FlagIcon = ({ flag }) => (jsxRuntime.jsx("img", { src: `https://www.untitledui.com/images/flags/${flag}.svg`, className: "size-4 max-w-none rounded-full", alt: `${flag} flag` }));

const TagGroupContext = h.createContext({
    size: 'sm',
});
const TagGroup = ({ label, size = 'sm', selectionMode: _selectionMode, disallowEmptySelection: _disallowEmptySelection, children, ...otherProps }) => {
    return (jsxRuntime.jsx(TagGroupContext.Provider, { value: { size }, children: jsxRuntime.jsx(reactAriaComponents.TagGroup, { "aria-label": label, ...otherProps, children: children }) }));
};
const TagList = reactAriaComponents.TagList;
const styles = {
    sm: {
        root: {
            base: 'px-2 py-0.75 text-xs font-medium',
            withAvatar: 'pl-1',
            withDot: 'pl-1.5',
            withCount: 'pr-1',
            withClose: 'pr-1',
            withFlag: 'pl-1.25',
        },
        content: 'gap-1',
        count: 'px-1 text-xs font-medium',
    },
    md: {
        root: {
            base: 'px-2.25 py-0.5 text-sm font-medium',
            withAvatar: 'pl-1.25',
            withDot: 'pl-1.75',
            withCount: 'pr-0.75',
            withClose: 'pr-1',
            withFlag: 'pl-1.25',
        },
        content: 'gap-1.25',
        count: 'px-1.25 text-xs font-medium',
    },
    lg: {
        root: {
            base: 'px-2.5 py-1 text-sm font-medium',
            withAvatar: 'pl-1.75',
            withDot: 'pl-2.25',
            withCount: 'pr-1',
            withClose: 'pr-1',
            withFlag: 'pl-1.25',
        },
        content: 'gap-1.5',
        count: 'px-1.5 text-sm font-medium',
    },
};
const Tag = ({ id, avatarSrc, avatarContrastBorder, dot, dotClassName, isDisabled, count, className, children, onClose, flag, }) => {
    const context = h.useContext(TagGroupContext);
    const leadingContent = avatarSrc ? (jsxRuntime.jsx(avatar.Avatar, { size: "xxs", src: avatarSrc, alt: "Avatar", contrastBorder: avatarContrastBorder })) : dot ? (jsxRuntime.jsx(dotIcon.Dot, { className: cx.cx('text-fg-success-secondary', dotClassName), size: "sm" })) : flag ? (jsxRuntime.jsx(FlagIcon, { flag: flag })) : null;
    return (jsxRuntime.jsx(reactAriaComponents.Tag, { id: id, isDisabled: isDisabled, textValue: typeof children === 'string' ? children : undefined, className: (state) => cx.cx('flex cursor-default items-center gap-[3px] rounded-sm border-primary bg-primary text-secondary ring-1 ring-primary ring-inset focus:outline-hidden focus-visible:outline-2 focus-visible:outline-offset-2 focus-visible:outline-focus-ring', styles[context.size].root.base, 
        // With avatar
        avatarSrc && styles[context.size].root.withAvatar, 
        // With flag
        flag && styles[context.size].root.withFlag, 
        // With X button
        (onClose || state.allowsRemoving) &&
            styles[context.size].root.withClose, 
        // With dot
        dot && styles[context.size].root.withDot, 
        // With count
        typeof count === 'number' && styles[context.size].root.withCount, 
        // Disabled
        isDisabled && 'cursor-not-allowed', typeof className === 'function' ? className(state) : className), children: ({ allowsRemoving }) => (jsxRuntime.jsxs(jsxRuntime.Fragment, { children: [jsxRuntime.jsxs("div", { className: cx.cx('flex items-center gap-1', styles[context.size].content), children: [leadingContent, children, typeof count === 'number' && (jsxRuntime.jsx("span", { className: cx.cx('flex items-center justify-center rounded-[3px] bg-tertiary text-center', styles[context.size].count), children: count }))] }), (onClose || allowsRemoving) && (jsxRuntime.jsx(tagCloseX.TagCloseX, { size: context.size, onPress: () => id && (onClose === null || onClose === void 0 ? void 0 : onClose(id.toString())) }))] })) }));
};

function TagsGroup({ children, ...props }) {
    return jsxRuntime.jsx(TagGroup, { ...props, children: children });
}
TagsGroup.TagList = TagList;
TagsGroup.Tag = Tag;

exports.TagsGroup = TagsGroup;
