// @hostaway/design-system/dist/components/ProgressSteps/index.cjs.js
'use strict';

var jsxRuntime = require('react/jsx-runtime');
var cx = require('../_shared/cx-CmkbyMOt.cjs.js');
var featuredIcon = require('../_shared/featured-icon-CeUxkT8F.cjs.js');
var h = require('react');
require('../_shared/CountrySelect-CKyGROoJ.cjs.js');
require('recharts');
require('../_shared/MyCustomIcon-DYQF4lJT.cjs.js');
require('react-aria-components');
require('tailwind-merge');
require('../_shared/is-react-component-DkG20B86.cjs.js');
require('../_shared/BaseSelect-0MCwMQdF.cjs.js');
require('@untitledui/icons');
require('../_shared/PopoverSearchInput-BFtG7lQx.cjs.js');
require('../_shared/InputField-DaibKVKA.cjs.js');
require('../_shared/input-CVSUknGT.cjs.js');
require('../_shared/hint-text-LRIVuQnk.cjs.js');
require('../_shared/label-D0WtNYRO.cjs.js');
require('../_shared/tooltip-DDB8niG0.cjs.js');
require('../_shared/FieldWrapper-D6dXzMyx.cjs.js');
require('../_shared/ConditionalTooltip-FLw9yEoi.cjs.js');
require('../_shared/avatar-C5INEgx9.cjs.js');
require('../_shared/FlagIcon-BbhvmpcI.cjs.js');
require('../_shared/Icon-BF_rHwZP.cjs.js');

const statuses = {
    incomplete: {
        icon: 'bg-disabled_subtle ring-[1.5px] ring-inset ring-disabled_subtle',
        dot: 'bg-fg-disabled_subtle',
        connector: 'border-secondary',
        title: 'text-secondary',
        description: 'text-tertiary',
    },
    current: {
        icon: 'bg-brand-solid ring-2 ring-focus-ring ring-offset-bg-primary ring-offset-2',
        dot: 'bg-fg-white',
        connector: 'border-secondary',
        title: 'text-brand-secondary',
        description: 'text-brand-tertiary',
    },
    complete: {
        icon: 'bg-brand-solid',
        dot: 'hidden',
        connector: '',
        title: 'text-secondary',
        description: 'text-tertiary',
    },
};
const IconOnly = ({ status = 'incomplete', size = 'sm', }) => {
    return (jsxRuntime.jsxs("span", { className: cx.cx('z-10 flex items-center justify-center rounded-full', statuses[status].icon, size === 'sm' ? 'size-6' : 'size-8'), children: [jsxRuntime.jsx("span", { className: cx.cx('rounded-full', statuses[status].dot, size === 'sm' ? 'size-2' : 'size-2.5') }), status === 'complete' && (jsxRuntime.jsx("svg", { width: "12", height: "12", viewBox: "0 0 12 12", fill: "none", className: cx.cx(size === 'sm' ? 'size-3' : 'size-4'), children: jsxRuntime.jsx("path", { d: "M10 3L4.5 8.5L2 6", stroke: "white", strokeWidth: "2", strokeLinecap: "round", strokeLinejoin: "round" }) }))] }));
};
const IconTop = ({ status = 'incomplete', connector = true, size = 'sm', title, description, ...otherProps }) => {
    return (jsxRuntime.jsxs("div", { className: cx.cx('flex w-full flex-col items-center justify-center gap-4', size === 'sm' && 'gap-3'), children: [jsxRuntime.jsxs("div", { className: "relative flex w-full flex-col items-center self-stretch", children: [jsxRuntime.jsx(IconOnly, { ...otherProps, size: size, status: status }), connector && (jsxRuntime.jsx("span", { className: cx.cx('absolute top-1/2 left-[53%] z-0 w-full flex-1 -translate-y-1/2 rounded-xs border-t-2', statuses[status].connector) }))] }), jsxRuntime.jsxs("div", { className: cx.cx('flex w-full flex-col items-start gap-0.5 self-stretch', size === 'sm' && 'gap-0'), children: [jsxRuntime.jsx("p", { className: cx.cx('w-full text-center', statuses[status].title, size === 'sm' ? 'text-sm font-semibold' : 'text-md font-semibold'), children: title }), jsxRuntime.jsx("p", { className: cx.cx('w-full text-center', statuses[status].description, size === 'sm' ? 'text-sm' : 'text-md'), children: description })] })] }));
};
const IconLeft = ({ status = 'incomplete', connector = true, size = 'sm', title, description, ...otherProps }) => {
    return (jsxRuntime.jsxs("div", { className: cx.cx('group flex h-max flex-row items-start justify-start gap-4', size === 'sm' && 'gap-3'), children: [jsxRuntime.jsxs("div", { className: "flex flex-col items-center self-stretch", children: [jsxRuntime.jsx(IconOnly, { ...otherProps, size: size, status: status }), connector && (jsxRuntime.jsx("span", { className: cx.cx('flex-1 rounded-xs border-l-2', statuses[status].connector, size === 'sm' ? 'my-1' : 'my-1.5') }))] }), jsxRuntime.jsxs("div", { className: cx.cx('flex flex-col items-start', size === 'sm'
                    ? 'pt-0.5 not-group-last:pb-6'
                    : 'pt-1 not-group-last:pb-8'), children: [jsxRuntime.jsx("p", { className: cx.cx(statuses[status].title, size === 'sm' ? 'text-sm font-semibold' : 'text-md font-semibold'), children: title }), jsxRuntime.jsx("p", { className: cx.cx(statuses[status].description, size === 'sm' ? 'text-sm' : 'text-md'), children: description })] })] }));
};
const IconTopNumber = ({ status = 'incomplete', connector = true, size = 'sm', title, description, step, }) => {
    const statuses = {
        incomplete: {
            icon: 'bg-primary ring-1 ring-inset ring-secondary text-disabled',
            title: 'text-secondary',
            description: 'text-tertiary',
        },
        current: {
            icon: 'bg-primary ring-1 ring-inset ring-secondary text-secondary',
            title: 'text-secondary',
            description: 'text-tertiary',
        },
        complete: {
            icon: 'bg-success-solid text-fg-white',
            title: 'text-secondary',
            description: 'text-tertiary',
        },
    };
    return (jsxRuntime.jsxs("div", { className: cx.cx('flex w-full flex-col items-center justify-center gap-4', size === 'sm' && 'gap-3'), children: [jsxRuntime.jsxs("div", { className: "relative flex w-full flex-col items-center self-stretch", children: [jsxRuntime.jsx("span", { className: cx.cx('z-10 flex items-center justify-center rounded-full', statuses[status].icon, status === 'incomplete' && 'opacity-60', size === 'sm' ? 'size-6' : 'size-8'), children: status === 'complete' ? (jsxRuntime.jsx("svg", { width: "12", height: "12", viewBox: "0 0 12 12", fill: "none", className: cx.cx(size === 'sm' ? 'size-3' : 'size-4'), children: jsxRuntime.jsx("path", { d: "M10 3L4.5 8.5L2 6", stroke: "white", strokeWidth: "2", strokeLinecap: "round", strokeLinejoin: "round" }) })) : (jsxRuntime.jsx("span", { className: cx.cx('font-semibold', size === 'sm' ? 'text-xs' : 'text-sm'), children: step })) }), connector && (jsxRuntime.jsx("svg", { className: "absolute top-1/2 left-[53%] z-0 h-[2.5px] w-full flex-1 -translate-y-1/2 max-md:hidden", children: jsxRuntime.jsx("line", { x1: "1.2", y1: "1.2", x2: "100%", y2: "1.2", className: "stroke-border-primary", stroke: "black", strokeWidth: "2.4", strokeDasharray: "0,6", strokeLinecap: "round" }) }))] }), jsxRuntime.jsxs("div", { className: cx.cx('flex w-full flex-col items-start gap-0.5 self-stretch', status === 'incomplete' && 'opacity-60', size === 'sm' && 'gap-0'), children: [jsxRuntime.jsx("p", { className: cx.cx('w-full text-center', statuses[status].title, size === 'sm' ? 'text-sm font-semibold' : 'text-md font-semibold'), children: title }), jsxRuntime.jsx("p", { className: cx.cx('w-full text-center', statuses[status].description, size === 'sm' ? 'text-sm' : 'text-md'), children: description })] })] }));
};
const IconLeftNumber = ({ status = 'incomplete', connector = true, size = 'sm', title, description, step, }) => {
    const statuses = {
        incomplete: {
            icon: 'bg-primary ring-1 ring-inset ring-secondary text-disabled',
            title: 'text-secondary',
            description: 'text-tertiary',
        },
        current: {
            icon: 'bg-primary ring-1 ring-inset ring-secondary text-secondary',
            title: 'text-secondary',
            description: 'text-tertiary',
        },
        complete: {
            icon: 'bg-success-solid text-fg-white',
            title: 'text-secondary',
            description: 'text-tertiary',
        },
    };
    return (jsxRuntime.jsxs("div", { className: cx.cx('group flex h-max flex-row items-start justify-start gap-4', status === 'incomplete' && 'opacity-60', size === 'sm' && 'gap-3'), children: [jsxRuntime.jsxs("div", { className: "flex flex-col items-center self-stretch", children: [jsxRuntime.jsx("span", { className: cx.cx('z-10 flex shrink-0 items-center justify-center rounded-full', statuses[status].icon, status === 'incomplete' && 'opacity-60', size === 'sm' ? 'size-6' : 'size-8'), children: status === 'complete' ? (jsxRuntime.jsx("svg", { width: "12", height: "12", viewBox: "0 0 12 12", fill: "none", className: cx.cx(size === 'sm' ? 'size-3' : 'size-4'), children: jsxRuntime.jsx("path", { d: "M10 3L4.5 8.5L2 6", stroke: "white", strokeWidth: "2", strokeLinecap: "round", strokeLinejoin: "round" }) })) : (jsxRuntime.jsx("span", { className: cx.cx('font-semibold', size === 'sm' ? 'text-xs' : 'text-sm'), children: step })) }), connector && (jsxRuntime.jsx("div", { className: cx.cx('relative flex h-full w-full justify-center self-center overflow-hidden', status === 'current' && 'opacity-60', size === 'sm' ? 'my-1' : 'my-1.5'), children: jsxRuntime.jsx("svg", { className: "absolute", width: "3", children: jsxRuntime.jsx("line", { x1: "1.2", y1: "1.2", x2: "1.2", y2: "100%", className: "stroke-border-primary", stroke: "black", strokeWidth: "2.4", strokeDasharray: "0,6", strokeLinecap: "round" }) }) }))] }), jsxRuntime.jsxs("div", { className: cx.cx('flex flex-col items-start', size === 'sm'
                    ? 'pt-0.5 not-group-last:pb-6'
                    : 'pt-1 not-group-last:pb-8'), children: [jsxRuntime.jsx("p", { className: cx.cx(statuses[status].title, size === 'sm' ? 'text-sm font-semibold' : 'text-md font-semibold'), children: title }), jsxRuntime.jsx("p", { className: cx.cx(statuses[status].description, size === 'sm' ? 'text-sm' : 'text-md'), children: description })] })] }));
};
const FeaturedIconTop = ({ status = 'incomplete', connector = true, size = 'sm', icon: Icon, title, description, }) => {
    return (jsxRuntime.jsxs("div", { className: cx.cx('flex w-full flex-col items-center justify-center gap-4', size === 'sm' && 'gap-3'), children: [jsxRuntime.jsxs("div", { className: cx.cx('relative flex w-full flex-col items-center self-stretch'), children: [jsxRuntime.jsx(featuredIcon.FeaturedIcon, { icon: Icon, color: "gray", theme: "modern", size: size === 'sm' ? 'md' : 'lg', className: cx.cx('z-10', status === 'incomplete' && 'opacity-60') }), connector && (jsxRuntime.jsx("span", { className: cx.cx('absolute top-1/2 left-[53%] z-0 w-full flex-1 -translate-y-1/2 rounded-xs border-t-2 border-secondary', status === 'complete' && 'border-fg-secondary') }))] }), jsxRuntime.jsxs("div", { className: cx.cx('flex w-full flex-col items-start gap-0.5 self-stretch', status === 'incomplete' && 'opacity-60', size === 'sm' && 'gap-0'), children: [jsxRuntime.jsx("p", { className: cx.cx('w-full text-center text-secondary', size === 'sm' ? 'text-sm font-semibold' : 'text-md font-semibold'), children: title }), jsxRuntime.jsx("p", { className: cx.cx('w-full text-center text-tertiary', size === 'sm' ? 'text-sm' : 'text-md'), children: description })] })] }));
};
const FeaturedIconLeft = ({ status = 'incomplete', connector = true, size = 'sm', icon: Icon, title, description, }) => {
    return (jsxRuntime.jsxs("div", { className: cx.cx('group flex flex-row gap-4', status !== 'current' && 'opacity-60', size === 'sm' && 'gap-3'), children: [jsxRuntime.jsxs("div", { className: "flex flex-col items-center self-stretch", children: [jsxRuntime.jsx(featuredIcon.FeaturedIcon, { size: "lg", color: "gray", theme: "modern", children: Icon && jsxRuntime.jsx(Icon, {}) }), connector && (jsxRuntime.jsx("span", { className: cx.cx(status === 'current' && 'opacity-60', 'my-1 flex-1 rounded-xs border-l-2 border-secondary') }))] }), jsxRuntime.jsxs("div", { className: cx.cx('flex flex-col items-start', size === 'sm' ? 'pt-1 not-group-last:pb-6' : 'not-group-last:pb-8'), children: [jsxRuntime.jsx("p", { className: cx.cx('text-secondary', size === 'sm' ? 'text-sm font-semibold' : 'text-md font-semibold'), children: title }), jsxRuntime.jsx("p", { className: cx.cx('text-tertiary', size === 'sm' ? 'text-sm' : 'text-md'), children: description })] })] }));
};
const TextLine = ({ status = 'incomplete', size = 'sm', title, description, }) => {
    return (jsxRuntime.jsxs("div", { className: cx.cx('relative flex w-full flex-col items-center justify-center pt-2', size === 'sm' && 'pt-3', size === 'md' && 'pt-4'), children: [jsxRuntime.jsx("div", { className: cx.cx('absolute inset-x-0 top-0 h-1 bg-fg-brand-primary_alt', status === 'incomplete' && 'bg-quaternary') }), jsxRuntime.jsxs("div", { className: cx.cx('flex w-full flex-col items-start gap-0.5 self-stretch', size === 'sm' && 'gap-0'), children: [jsxRuntime.jsx("p", { className: cx.cx('text-secondary', status === 'current' && 'text-brand-secondary', size === 'sm' ? 'text-sm font-semibold' : 'text-md font-semibold'), children: title }), jsxRuntime.jsx("p", { className: cx.cx('text-tertiary', status === 'current' && 'text-brand-tertiary', size === 'sm' ? 'text-sm' : 'text-md'), children: description })] })] }));
};

const progressIcons = {
    horizontal: {
        icon: IconTop,
        number: IconTopNumber,
        'featured-icon': FeaturedIconTop,
    },
    vertical: {
        icon: IconLeft,
        number: IconLeftNumber,
        'featured-icon': FeaturedIconLeft,
    },
};
const IconsWithText = (props) => {
    const { type = 'icon', orientation = 'vertical', size = 'sm', connector = true, items, className, ...rest } = props;
    const length = items.length;
    // Single step component based on the type.
    const StepBase = progressIcons[orientation][type];
    return (jsxRuntime.jsx("div", { style: {
            gridTemplateColumns: `repeat(${length}, minmax(0, 1fr))`,
        }, className: cx.cx('grid items-start justify-start', orientation === 'horizontal' && 'w-full gap-4', orientation === 'vertical' && 'grid-cols-1!', className), ...rest, children: items.map((item, index) => (jsxRuntime.jsx(StepBase, { ...item, size: size, connector: !connector ? false : item.connector || index !== length - 1, type: type, step: index + 1 }, index))) }));
};
const MinimalIcons = ({ items, text = false, size = 'sm', className, ...rest }) => {
    const gaps = {
        sm: 'gap-3',
        md: 'gap-4',
        lg: 'gap-5',
    };
    const completed = items.filter((item) => item.status === 'complete').length;
    return (jsxRuntime.jsxs("div", { className: cx.cx('flex w-full flex-row items-center justify-center', gaps[size], className), ...rest, children: [text && (jsxRuntime.jsxs("p", { className: "text-sm font-medium text-secondary", children: ["Step ", completed, " of ", items.length] })), items.map((item, index) => (jsxRuntime.jsx(IconOnly, { ...item, size: size }, index)))] }));
};
const MinimalIconsConnected = ({ items, size = 'sm', orientation = 'horizontal', className, ...rest }) => {
    return (jsxRuntime.jsx("div", { className: cx.cx('flex w-full items-center justify-center', className), ...rest, children: items.map((item, index) => (jsxRuntime.jsxs("div", { className: "flex items-center justify-center", children: [jsxRuntime.jsx(IconOnly, { ...item, size: size }), jsxRuntime.jsx("span", { className: cx.cx('w-20 flex-1 border-t-2', statuses[item.status].connector, orientation === 'vertical' && 'w-12', index === items.length - 1 && 'hidden') })] }, index))) }));
};
const TextWithLine = ({ items, orientation = 'horizontal', size = 'sm', className, ...rest }) => {
    return (jsxRuntime.jsx("div", { style: {
            gridTemplateColumns: `repeat(${items.length}, minmax(0, 1fr))`,
        }, className: cx.cx('grid w-full items-start justify-start', orientation === 'horizontal' ? 'gap-4' : 'gap-5', orientation === 'vertical' && 'grid-cols-1!', className), ...rest, children: items.map((item, index) => (jsxRuntime.jsx(TextLine, { ...item, size: size }, index))) }));
};
const Progress = {
    IconsWithText,
    MinimalIcons,
    MinimalIconsConnected,
    TextWithLine,
};

const NumberIcon = ({ step, size = 'sm', item }) => {
    const statuses = {
        incomplete: {
            icon: 'bg-primary ring-1 ring-inset ring-secondary text-disabled',
        },
        current: {
            icon: 'bg-primary ring-1 ring-inset ring-secondary text-secondary',
        },
        complete: {
            icon: 'bg-success-solid text-fg-white',
        },
    };
    return (jsxRuntime.jsx("span", { className: cx.cx('z-10 flex items-center justify-center rounded-full', statuses[item.status].icon, item.status === 'incomplete' && 'opacity-60', size === 'sm' ? 'size-6' : 'size-8'), children: item.status === 'complete' ? (jsxRuntime.jsx("svg", { width: "12", height: "12", viewBox: "0 0 12 12", fill: "none", className: cx.cx(size === 'sm' ? 'size-3' : 'size-4'), children: jsxRuntime.jsx("path", { d: "M10 3L4.5 8.5L2 6", stroke: "white", strokeWidth: "2", strokeLinecap: "round", strokeLinejoin: "round" }) })) : (jsxRuntime.jsx("span", { className: cx.cx('font-semibold', size === 'sm' ? 'text-xs' : 'text-sm'), children: step })) }));
};

const MinimalNumbers = ({ items, text = false, size = 'sm', className, ...rest }) => {
    const gaps = {
        sm: 'gap-3',
        md: 'gap-4',
    };
    const completed = items.filter((item) => item.status === 'complete').length;
    return (jsxRuntime.jsxs("div", { className: cx.cx('flex w-full flex-row items-center justify-center', gaps[size], className), ...rest, children: [text && (jsxRuntime.jsxs("p", { className: "text-sm font-medium text-secondary", children: ["Step ", completed, " of ", items.length] })), items.map((item, index) => (jsxRuntime.jsx(NumberIcon, { step: index + 1, item: item, size: size }, index)))] }));
};

const MinimalNumbersConnected = ({ items, size = 'sm', className, ...rest }) => {
    return (jsxRuntime.jsx("div", { className: cx.cx('flex w-full items-center justify-center', className), ...rest, children: items.map((item, index) => (jsxRuntime.jsxs("div", { className: "flex items-center justify-center relative", children: [jsxRuntime.jsx(NumberIcon, { step: index + 1, item: item, size: size }), index !== items.length - 1 && (jsxRuntime.jsxs(jsxRuntime.Fragment, { children: [jsxRuntime.jsx("svg", { className: "absolute top-1/2 left-[26%] z-0 h-[2.5px] w-19 flex-1 -translate-y-1/2", children: jsxRuntime.jsx("line", { x1: "1.2", y1: "1.2", x2: "100%", y2: "1.2", className: "stroke-border-primary", stroke: "black", strokeWidth: "2.4", strokeDasharray: "0,6", strokeLinecap: "round" }) }), jsxRuntime.jsx("span", { className: "w-20" })] }))] }, index))) }));
};

const ProgressMinimalIconWrapper = ({ children, }) => {
    const { text, setText, items, className, size } = children.props;
    const gaps = {
        sm: 'gap-3',
        md: 'gap-4',
    };
    if (text && setText) {
        const current = items.findIndex((item) => item.status === 'current') + 1;
        const total = items.length;
        return (jsxRuntime.jsxs("div", { className: cx.cx('flex w-full flex-row items-center justify-center', className, gaps[size || 'sm']), children: [jsxRuntime.jsx("div", { className: "w-fit", children: setText(current, total) }), h.cloneElement(children, {
                    ...children.props,
                    text: false,
                    className: 'w-fit',
                })] }));
    }
    return children;
};

const ProgressSteps = (props) => {
    switch (props.variant) {
        case 'icons-with-text':
            return jsxRuntime.jsx(Progress.IconsWithText, { ...props });
        case 'minimal-icons':
            return props.type === 'number' ? (jsxRuntime.jsx(ProgressMinimalIconWrapper, { children: jsxRuntime.jsx(MinimalNumbers, { ...props }) })) : (jsxRuntime.jsx(ProgressMinimalIconWrapper, { children: jsxRuntime.jsx(Progress.MinimalIcons, { ...props }) }));
        case 'minimal-icons-connected':
            return props.type === 'number' ? (jsxRuntime.jsx(MinimalNumbersConnected, { ...props })) : (jsxRuntime.jsx(Progress.MinimalIconsConnected, { ...props, orientation: "horizontal" }));
        case 'text-with-line':
            return jsxRuntime.jsx(Progress.TextWithLine, { ...props });
    }
};

exports.ProgressSteps = ProgressSteps;
