// @hostaway/design-system/dist/components/_shared/BaseDropdown-CG2Me0TI.esm.js
import { jsx, jsxs } from 'react/jsx-runtime';
import { DotsHorizontal, DotsVertical } from '@untitledui/icons';
import './MyCustomIcon-BRHTusqN.esm.js';
import { MenuTrigger, MenuSection, Header, Popover, Menu, MenuItem, Separator, Button } from 'react-aria-components';
import { c as cx } from './cx-DnIR8RkH.esm.js';
import { T as Tooltip } from './Tooltip-BtHtZ-Fm.esm.js';

const DropdownItem = ({ label, children, addon, icon: Icon, unstyled, tooltip, renderItem, ...props }) => {
    if (unstyled) {
        return (jsx(MenuItem, { id: label, textValue: label, ...props }));
    }
    return (jsx(MenuItem, { ...props, className: (state) => cx('group block cursor-pointer px-1.5 py-px outline-hidden', state.isDisabled && 'cursor-not-allowed', typeof props.className === 'function'
            ? props.className(state)
            : props.className), children: (state) => (jsx("div", { className: cx('relative flex items-center rounded-md px-2.5 py-2 outline-focus-ring transition duration-100 ease-linear', !state.isDisabled && 'group-hover:bg-primary_hover', state.isFocused && 'bg-primary_hover', state.isFocusVisible && 'outline-2 -outline-offset-2'), children: renderItem ? (renderItem(state)) : (jsxs(Tooltip, { title: tooltip, "data-id": `dropdown-${label}-option-tooltip`, triggerProps: {
                    className: cx('w-full text-left flex items-center cursor-pointer', state.isDisabled && 'cursor-not-allowed'),
                }, children: [Icon && (jsx(Icon, { "aria-hidden": "true", className: cx('mr-2 size-4 shrink-0 stroke-[2.25px]', state.isDisabled
                            ? 'text-fg-disabled'
                            : 'text-fg-quaternary') })), jsx("span", { className: cx('grow truncate text-sm font-semibold', state.isDisabled ? 'text-disabled' : 'text-secondary', state.isFocused && 'text-secondary_hover'), children: label ||
                            (typeof children === 'function' ? children(state) : children) }), addon && (jsx("span", { className: cx('ml-3 shrink-0 rounded px-1 py-px text-xs font-medium ring-1 ring-secondary ring-inset', state.isDisabled ? 'text-disabled' : 'text-quaternary'), children: addon }))] })) })) }));
};
const DropdownMenu = (props) => {
    return (jsx(Menu, { disallowEmptySelection: true, selectionMode: "single", ...props, className: (state) => cx('h-min overflow-y-auto py-1 outline-hidden select-none', typeof props.className === 'function'
            ? props.className(state)
            : props.className) }));
};
const DropdownPopover = (props) => {
    return (jsx(Popover, { placement: "bottom right", ...props, className: (state) => cx('w-62 overflow-auto rounded-lg bg-primary shadow-lg ring-1 ring-secondary_alt will-change-transform', state.isEntering &&
            'duration-150 ease-out animate-in fade-in placement-right:origin-left placement-right:slide-in-from-left-0.5 placement-top:origin-bottom placement-top:slide-in-from-bottom-0.5 placement-bottom:origin-top placement-bottom:slide-in-from-top-0.5', state.isExiting &&
            'duration-100 ease-in animate-out fade-out placement-right:origin-left placement-right:slide-out-to-left-0.5 placement-top:origin-bottom placement-top:slide-out-to-bottom-0.5 placement-bottom:origin-top placement-bottom:slide-out-to-top-0.5', typeof props.className === 'function'
            ? props.className(state)
            : props.className), children: props.children }));
};
const DropdownSeparator = (props) => {
    return (jsx(Separator, { ...props, className: cx('my-1 h-px w-full bg-border-secondary', props.className) }));
};
const DropdownDotsButton = (props) => {
    const Icon = props.horizontal ? DotsHorizontal : DotsVertical;
    return (jsx(Button, { ...props, "aria-label": "Open menu", className: (state) => cx('cursor-pointer rounded-md text-fg-quaternary outline-focus-ring transition duration-100 ease-linear', (state.isPressed || state.isHovered) && 'text-fg-quaternary_hover', (state.isPressed || state.isFocusVisible) &&
            'outline-2 outline-offset-2', typeof props.className === 'function'
            ? props.className(state)
            : props.className), children: jsx(Icon, { className: "size-5 transition-inherit-all" }) }));
};
const Dropdown = {
    Root: MenuTrigger,
    Popover: DropdownPopover,
    Menu: DropdownMenu,
    Section: MenuSection,
    SectionHeader: Header,
    Item: DropdownItem,
    Separator: DropdownSeparator,
    DotsButton: DropdownDotsButton,
};

export { Dropdown as D };
