// @hostaway/design-system/dist/components/_shared/BaseButton-CTtHjEZn.esm.js
import { jsxs, jsx } from 'react/jsx-runtime';
import { isValidElement } from 'react';
import { Link, Button as Button$1 } from 'react-aria-components';
import { c as cx, s as sortCx } from './cx-DnIR8RkH.esm.js';
import { i as isReactComponent } from './is-react-component-CYKUd65Q.esm.js';

const styles = sortCx({
    common: {
        root: [
            'group relative inline-flex h-max cursor-pointer rounded-md items-center justify-center whitespace-nowrap outline-gray-500 transition duration-100 ease-linear before:absolute focus-visible:outline-2 focus-visible:outline-offset-2',
            // When button is used within `InputGroup` as a leading/trailing addon
            'in-data-input-wrapper:in-data-leading:shadow-xs in-data-input-wrapper:in-data-trailing:shadow-xs in-data-input-wrapper:in-data-leading:focus:!z-50 in-data-input-wrapper:in-data-trailing:focus:!z-50 in-data-input-wrapper:in-data-leading:-mr-px in-data-input-wrapper:in-data-leading:rounded-r-none in-data-input-wrapper:in-data-leading:before:rounded-r-none in-data-input-wrapper:in-data-trailing:-ml-px in-data-input-wrapper:in-data-trailing:rounded-l-none in-data-input-wrapper:in-data-trailing:before:rounded-l-none',
            // Disabled styles
            'disabled:cursor-not-allowed disabled:text-fg-disabled',
            // Icon styles
            'disabled:*:data-icon:text-fg-disabled_subtle',
            // Same as `icon` but for SSR icons that cannot be passed to the client as functions.
            '*:data-icon:pointer-events-none *:data-icon:size-5 *:data-icon:shrink-0 *:data-icon:transition-inherit-all',
        ].join(' '),
        icon: 'pointer-events-none size-5 shrink-0 transition-inherit-all',
    },
    sizes: {
        sm: {
            root: [
                'gap-1 rounded-md px-3 py-2 text-sm font-semibold before:rounded-[7px] data-icon-only:p-2',
                'in-data-input-wrapper:in-data-leading:px-3.5 in-data-input-wrapper:in-data-trailing:px-3.5 in-data-input-wrapper:in-data-leading:py-2.5 in-data-input-wrapper:in-data-trailing:py-2.5 in-data-input-wrapper:in-data-leading:data-icon-only:p-2.5 in-data-input-wrapper:in-data-trailing:data-icon-only:p-2.5',
            ].join(' '),
            linkRoot: 'gap-1',
        },
        xl: {
            root: 'gap-1.5 rounded-md px-4.5 py-3 text-md font-semibold before:rounded-[7px] data-icon-only:p-3.5',
            linkRoot: 'gap-1.5',
        },
    },
    colors: {
        primary: {
            root: [
                'bg-brand-solid text-white shadow-xs-skeumorphic ring-1 ring-transparent ring-inset hover:bg-brand-solid_hover data-loading:bg-brand-solid_hover',
                // Inner border gradient
                'before:absolute before:inset-px before:border before:border-white/12 before:mask-b-from-0%',
                // Disabled styles
                'disabled:bg-disabled disabled:shadow-xs disabled:ring-disabled_subtle',
                // Icon styles
                '*:data-icon:text-button-primary-icon hover:*:data-icon:text-button-primary-icon_hover',
            ].join(' '),
        },
        secondary: {
            root: [
                'bg-primary text-secondary shadow-xs-skeumorphic ring-1 ring-primary ring-inset hover:bg-primary_hover hover:text-secondary_hover data-loading:bg-primary_hover',
                // Disabled styles
                'disabled:shadow-xs disabled:ring-disabled_subtle',
                // Icon styles
                '*:data-icon:text-fg-quaternary hover:*:data-icon:text-fg-quaternary_hover',
            ].join(' '),
        },
        tertiary: {
            root: [
                'text-tertiary hover:bg-primary_hover hover:text-tertiary_hover data-loading:bg-primary_hover',
                // Icon styles
                '*:data-icon:text-fg-quaternary hover:*:data-icon:text-fg-quaternary_hover',
            ].join(' '),
        },
        'link-gray': {
            root: [
                'justify-normal rounded-xs p-0! text-tertiary hover:text-tertiary_hover',
                // Inner text underline
                '*:data-text:underline *:data-text:decoration-transparent *:data-text:underline-offset-2 hover:*:data-text:decoration-current',
                // Icon styles
                '*:data-icon:text-fg-quaternary hover:*:data-icon:text-fg-quaternary_hover',
            ].join(' '),
        },
        'link-color': {
            root: [
                'justify-normal rounded-xs p-0! text-tertiary hover:text-tertiary',
                // Inner text underline
                '*:data-text:underline *:data-text:decoration-transparent *:data-text:underline-offset-2 hover:*:data-text:decoration-current',
                // Icon styles
                '*:data-icon:text-fg-brand-secondary_alt hover:*:data-icon:text-fg-brand-secondary_hover',
            ].join(' '),
        },
        'primary-destructive': {
            root: [
                'bg-error-solid text-white shadow-xs-skeumorphic outline-error-500 hover:bg-error-solid_hover',
                // Focus styles
                'focus-visible:outline-error-500',
                // Inner border gradient
                'before:absolute before:inset-px before:border before:border-white/12 before:mask-b-from-0%',
                // Disabled styles
                'disabled:bg-disabled disabled:shadow-xs disabled:ring-disabled_subtle',
                // Icon styles
                '*:data-icon:text-button-destructive-primary-icon hover:*:data-icon:text-button-destructive-primary-icon_hover',
            ].join(' '),
        },
        'secondary-destructive': {
            root: [
                'bg-primary text-error-primary outline-error-500 shadow-xs-skeumorphic ring-1 ring-error_subtle ring-inset hover:bg-error-primary hover:text-error-primary_hover data-loading:bg-error-primary',
                // Focus styles
                'focus-visible:outline-error-500',
                // Disabled styles
                'disabled:bg-primary disabled:shadow-xs disabled:ring-disabled_subtle',
                // Icon styles
                '*:data-icon:text-fg-error-secondary hover:*:data-icon:text-fg-error-primary',
            ].join(' '),
        },
        'tertiary-destructive': {
            root: [
                'text-error-primary outline-error-500 hover:bg-error-primary hover:text-error-primary_hover data-loading:bg-error-primary',
                // Focus styles
                'focus-visible:outline-error-500',
                // Icon styles
                '*:data-icon:text-fg-error-secondary hover:*:data-icon:text-fg-error-primary',
            ].join(' '),
        },
        'link-destructive': {
            root: [
                'justify-normal rounded-xs p-0! outline-error-500 text-error-primary hover:text-error-primary_hover',
                // Focus styles
                'focus-visible:outline-error-500',
                // Inner text underline
                '*:data-text:underline *:data-text:decoration-transparent *:data-text:underline-offset-2 hover:*:data-text:decoration-current',
                // Icon styles
                '*:data-icon:text-fg-error-secondary hover:*:data-icon:text-fg-error-primary',
            ].join(' '),
        },
        'primary-warning': {
            root: [
                'bg-warning-solid text-white shadow-xs-skeumorphic ring-1 ring-transparent outline-warning ring-inset hover:bg-warning-600 data-loading:bg-warning-600',
                // Inner border gradient
                'before:absolute before:inset-px before:border before:border-white/12 before:mask-b-from-0%',
                // Disabled styles
                'disabled:bg-disabled disabled:shadow-xs disabled:ring-disabled_subtle',
                // Icon styles
                '*:data-icon:text-warning-300 hover:*:data-icon:text-warning-200',
            ].join(' '),
        },
    },
});
const Button = ({ size = 'sm', color = 'primary', children, className, noTextPadding, iconLeading: IconLeading, iconTrailing: IconTrailing, isDisabled: disabled, isLoading: loading, showTextWhileLoading, ...otherProps }) => {
    const href = 'href' in otherProps ? otherProps.href : undefined;
    const Component = href ? Link : Button$1;
    const isIcon = (IconLeading || IconTrailing) && !children;
    const isLinkType = ['link-gray', 'link-color', 'link-destructive'].includes(color);
    noTextPadding = isLinkType || noTextPadding;
    let props;
    if (href) {
        props = {
            ...otherProps,
            href: disabled ? undefined : href,
            // Since anchor elements do not support the `disabled` attribute and state,
            // we need to specify `data-rac` and `data-disabled` in order to be able
            // to use the `disabled:` selector in classes.
            ...(disabled ? { 'data-rac': true, 'data-disabled': true } : {}),
        };
    }
    else {
        props = {
            ...otherProps,
            type: otherProps.type || 'button',
            isPending: loading,
            isDisabled: disabled,
        };
    }
    return (jsxs(Component, { "data-loading": loading ? true : undefined, "data-icon-only": isIcon ? true : undefined, ...props, className: cx(styles.common.root, styles.sizes[size].root, styles.colors[color].root, isLinkType && styles.sizes[size].linkRoot, (loading || (href && (disabled || loading))) && 'pointer-events-none', 
        // If in `loading` state, hide everything except the loading icon (and text if `showTextWhileLoading` is true).
        loading &&
            (showTextWhileLoading
                ? '[&>*:not([data-icon=loading]):not([data-text])]:hidden'
                : '[&>*:not([data-icon=loading])]:invisible'), className), children: [isValidElement(IconLeading) && IconLeading, isReactComponent(IconLeading) && (jsx(IconLeading, { "data-icon": "leading", className: styles.common.icon })), loading && (jsxs("svg", { fill: "none", "data-icon": "loading", viewBox: "0 0 20 20", className: cx(styles.common.icon, !showTextWhileLoading &&
                    'absolute top-1/2 left-1/2 -translate-x-1/2 -translate-y-1/2'), children: [jsx("circle", { className: "stroke-current opacity-30", cx: "10", cy: "10", r: "8", fill: "none", strokeWidth: "2" }), jsx("circle", { className: "origin-center animate-spin stroke-current", cx: "10", cy: "10", r: "8", fill: "none", strokeWidth: "2", strokeDasharray: "12.5 50", strokeLinecap: "round" })] })), children && (jsx("span", { "data-text": true, className: cx('transition-inherit-all', !noTextPadding && 'px-0.5'), children: children })), isValidElement(IconTrailing) && IconTrailing, isReactComponent(IconTrailing) && (jsx(IconTrailing, { "data-icon": "trailing", className: styles.common.icon }))] }));
};

export { Button as B };
