// @hostaway/design-system/dist/components/_shared/BaseSelect-CHht82wi.esm.js
import { jsx, jsxs, Fragment } from 'react/jsx-runtime';
import { useEffect, forwardRef, useContext, useRef, useState, useCallback, isValidElement } from 'react';
import { SearchLg, Check, ChevronDown } from '@untitledui/icons';
import './MyCustomIcon-BRHTusqN.esm.js';
import { P as PopoverSearchInput } from './PopoverSearchInput-D81SVeJp.esm.js';
import { Popover as Popover$1, ComboBoxStateContext, Group, Input, ComboBox as ComboBox$1, ListBox, ListBoxItem, Text, Select as Select$1, Button, SelectValue as SelectValue$1 } from 'react-aria-components';
import { A as Avatar } from './avatar-DeJPrYaB.esm.js';
import { c as cx } from './cx-DnIR8RkH.esm.js';
import { i as isReactComponent } from './is-react-component-CYKUd65Q.esm.js';
import { H as HintText } from './hint-text-Ck8Vj_nN.esm.js';
import { L as Label } from './label-pgzo6IFb.esm.js';

const Popover = (props) => {
    return (jsx(Popover$1, { placement: "bottom", containerPadding: 0, offset: 4, ...props, className: (state) => cx('max-h-64! w-(--trigger-width) overflow-x-hidden overflow-y-auto rounded-lg bg-primary py-1 shadow-lg ring-1 ring-secondary_alt outline-hidden will-change-transform', state.isEntering &&
            'duration-150 ease-out animate-in fade-in placement-right:origin-left placement-right:slide-in-from-left-0.5 placement-top:origin-bottom placement-top:slide-in-from-bottom-0.5 placement-bottom:origin-top placement-bottom:slide-in-from-top-0.5', state.isExiting &&
            'duration-100 ease-in animate-out fade-out placement-right:origin-left placement-right:slide-out-to-left-0.5 placement-top:origin-bottom placement-top:slide-out-to-bottom-0.5 placement-bottom:origin-top placement-bottom:slide-out-to-top-0.5', typeof props.className === 'function'
            ? props.className(state)
            : props.className) }));
};

/**
 * Checks if the ResizeObserver API is supported.
 * @returns True if the ResizeObserver API is supported, false otherwise.
 */
function hasResizeObserver() {
    return typeof window.ResizeObserver !== "undefined";
}
/**
 * A hook that observes the size of an element and calls a callback function when the size changes.
 * @param options - The options for the hook.
 */
function useResizeObserver(options) {
    const { ref, box, onResize } = options;
    useEffect(() => {
        const element = ref === null || ref === void 0 ? void 0 : ref.current;
        if (!element) {
            return;
        }
        if (!hasResizeObserver()) {
            window.addEventListener("resize", onResize, false);
            return () => {
                window.removeEventListener("resize", onResize, false);
            };
        }
        else {
            const resizeObserverInstance = new window.ResizeObserver((entries) => {
                if (!entries.length) {
                    return;
                }
                onResize();
            });
            resizeObserverInstance.observe(element, { box });
            return () => {
                if (element) {
                    resizeObserverInstance.unobserve(element);
                }
            };
        }
    }, [onResize, ref, box]);
}

const ComboBoxValue = forwardRef(({ shortcut, placeholder, shortcutClassName, ...otherProps }, ref) => {
    var _a, _b;
    const state = useContext(ComboBoxStateContext);
    const value = ((_a = state === null || state === void 0 ? void 0 : state.selectedItem) === null || _a === void 0 ? void 0 : _a.value) || null;
    const inputValue = (state === null || state === void 0 ? void 0 : state.inputValue) || null;
    const first = ((_b = inputValue === null || inputValue === void 0 ? void 0 : inputValue.split(value === null || value === void 0 ? void 0 : value.supportingText)) === null || _b === void 0 ? void 0 : _b[0]) || '';
    const last = inputValue === null || inputValue === void 0 ? void 0 : inputValue.split(first)[1];
    return (jsx(Group, { ...otherProps, ref: ref, className: ({ isFocusWithin, isDisabled }) => cx('relative flex w-full items-center gap-2 rounded-lg bg-primary shadow-xs ring-1 ring-primary outline-hidden transition-shadow duration-100 ease-linear ring-inset py-2 px-3', isDisabled && 'cursor-not-allowed bg-disabled_subtle', isFocusWithin && 'ring-2 ring-brand'), children: ({ isDisabled }) => (jsxs(Fragment, { children: [jsx(SearchLg, { className: "pointer-events-none size-4 text-fg-quaternary" }), jsxs("div", { className: "relative flex w-full items-center gap-2", children: [inputValue && (jsxs("span", { className: "absolute top-1/2 z-0 inline-flex w-full -translate-y-1/2 gap-2 truncate", "aria-hidden": "true", children: [jsx("p", { className: cx('text-sm font-medium text-primary', isDisabled && 'text-disabled'), children: first }), last && (jsx("p", { className: cx('-ml-0.75 text-sm text-tertiary', isDisabled && 'text-disabled'), children: last }))] })), jsx(Input, { placeholder: placeholder, className: "z-10 w-full appearance-none bg-transparent text-sm text-transparent caret-alpha-black/90 placeholder:text-placeholder focus:outline-hidden disabled:cursor-not-allowed disabled:text-disabled disabled:placeholder:text-disabled" })] }), shortcut && (jsx("div", { className: cx('absolute inset-y-0.5 right-0.5 z-10 flex items-center rounded-r-[inherit] bg-linear-to-r from-transparent to-bg-primary to-40% pl-8 pr-2.5', isDisabled && 'to-bg-disabled_subtle', shortcutClassName), children: jsx("span", { className: cx('pointer-events-none rounded px-1 py-px text-xs font-medium text-quaternary ring-1 ring-secondary select-none ring-inset', isDisabled && 'bg-transparent text-disabled'), "aria-hidden": "true", children: "\u2318K" }) }))] })) }));
});
const ComboBox = forwardRef(({ placeholder = 'Search', shortcut = true, children, items, shortcutClassName, ...otherProps }, ref) => {
    const placeholderRef = useRef(null);
    const [popoverWidth, setPopoverWidth] = useState('');
    // Resize observer for popover width
    const onResize = useCallback(() => {
        var _a;
        if (!placeholderRef.current)
            return;
        const divRect = (_a = placeholderRef.current) === null || _a === void 0 ? void 0 : _a.getBoundingClientRect();
        setPopoverWidth(divRect.width + 'px');
    }, [placeholderRef, setPopoverWidth]);
    useResizeObserver({
        ref: placeholderRef,
        box: 'border-box',
        onResize,
    });
    return (jsx(ComboBox$1, { ref: ref, menuTrigger: "focus", ...otherProps, children: (state) => (jsxs("div", { className: "flex flex-col gap-1.5", children: [otherProps.label && (jsx(Label, { isRequired: state.isRequired, tooltip: otherProps.tooltip, children: otherProps.label })), jsx(ComboBoxValue, { ref: placeholderRef, placeholder: placeholder, shortcut: shortcut, shortcutClassName: shortcutClassName, 
                    // This is a workaround to correctly calculating the trigger width
                    // while using ResizeObserver wasn't 100% reliable.
                    onFocus: onResize, onPointerEnter: onResize }), jsx(Popover, { triggerRef: placeholderRef, style: { width: popoverWidth }, className: otherProps.popoverClassName, children: jsx(ListBox, { items: items, className: "size-full outline-hidden", children: children }) }), otherProps.hint && (jsx(HintText, { isInvalid: state.isInvalid, children: otherProps.hint }))] })) }));
});

const SelectItem = ({ label, id, value, avatarUrl, supportingText, isDisabled, icon: Icon, className, children, action, selectedClassName, renderItem = undefined, ...props }) => {
    const labelOrChildren = (typeof label === 'string' ? label : '') ||
        (typeof children === 'string' ? children : '');
    const textValue = supportingText
        ? labelOrChildren + ' ' + supportingText
        : labelOrChildren;
    return (jsx(ListBoxItem, { id: id, value: value !== null && value !== void 0 ? value : {
            id,
            label: labelOrChildren,
            avatarUrl,
            supportingText,
            isDisabled,
            icon: Icon,
        }, textValue: textValue, isDisabled: isDisabled, ...props, className: (state) => cx('w-full px-1.5 py-px outline-hidden', typeof className === 'function' ? className(state) : className), children: (state) => (jsxs("div", { className: cx('flex cursor-pointer items-center gap-2 rounded-md outline-hidden select-none p-2 pr-2.5', state.isSelected && (selectedClassName || 'bg-active'), state.isDisabled && 'cursor-not-allowed', state.isFocusVisible &&
                'bg-primary_hover ring-2 ring-focus-ring ring-inset', 
            // Icon styles
            '*:data-icon:size-4 *:data-icon:shrink-0 *:data-icon:text-fg-quaternary', state.isDisabled && '*:data-icon:text-fg-disabled'), children: [renderItem && renderItem(state), !renderItem && (jsxs(Fragment, { children: [avatarUrl ? (jsx(Avatar, { "aria-hidden": "true", size: "xs", src: avatarUrl, alt: typeof label === 'string' ? label : '' })) : isReactComponent(Icon) ? (jsx(Icon, { "data-icon": true, "aria-hidden": "true" })) : isValidElement(Icon) ? (Icon) : null, jsxs("div", { className: "flex w-full min-w-0 flex-1 flex-wrap gap-x-2", children: [jsx(Text, { slot: "label", className: cx('truncate text-sm font-medium whitespace-nowrap text-primary', state.isDisabled && 'text-disabled'), children: label ||
                                        (typeof children === 'function'
                                            ? children(state)
                                            : children) }), supportingText && (jsx(Text, { slot: "description", className: cx('text-sm whitespace-nowrap text-tertiary', state.isDisabled && 'text-disabled'), children: supportingText }))] }), action !== null && action !== void 0 ? action : (state.isSelected && (jsx(Check, { "aria-hidden": "true", className: cx('ml-auto size-4 stroke-[2.5px] text-fg-brand-primary', state.isDisabled && 'text-fg-disabled') })))] }))] })) }));
};

const SelectValue = ({ isOpen, isFocusVisible, isDisabled, isInvalid, placeholder, placeholderIcon, selectedItemFallback, noBorder, ref, }) => {
    return (jsx(Button, { ref: ref, className: cx('relative flex w-full cursor-pointer items-center rounded-lg bg-primary outline-hidden transition duration-100 ease-linear', !noBorder && 'shadow-xs ring-1 ring-primary ring-inset', !noBorder && (isFocusVisible || isOpen) && 'ring-2 ring-brand', !noBorder && isInvalid && 'ring-error_subtle', isDisabled && 'cursor-not-allowed bg-disabled_subtle text-disabled', !placeholder && 'min-h-9'), children: jsx(SelectValue$1, { className: cx('flex h-max w-full items-center justify-start gap-2 truncate text-left align-middle', !noBorder && 'py-2 px-3', 
            // Icon styles
            '*:data-icon:size-4 *:data-icon:shrink-0 *:data-icon:text-fg-quaternary in-disabled:*:data-icon:text-fg-disabled'), children: (state) => {
                var _a;
                // Use fallback when the selected item is filtered out of the collection
                const selectedItem = (_a = state.selectedItem) !== null && _a !== void 0 ? _a : selectedItemFallback;
                const Icon = (selectedItem === null || selectedItem === void 0 ? void 0 : selectedItem.icon) || placeholderIcon;
                return (jsxs(Fragment, { children: [(selectedItem === null || selectedItem === void 0 ? void 0 : selectedItem.avatarUrl) ? (jsx(Avatar, { size: "xs", src: selectedItem.avatarUrl, alt: typeof selectedItem.label === 'string'
                                ? selectedItem.label
                                : '' })) : isReactComponent(Icon) ? (jsx(Icon, { "data-icon": true, "aria-hidden": "true" })) : isValidElement(Icon) ? (Icon) : null, selectedItem ? (jsxs("section", { className: "flex w-full gap-2 truncate", children: [selectedItem.renderLabel && selectedItem.renderLabel, !selectedItem.renderLabel && (jsxs(Fragment, { children: [jsx("p", { className: "truncate text-sm font-medium text-primary", children: selectedItem === null || selectedItem === void 0 ? void 0 : selectedItem.label }), (selectedItem === null || selectedItem === void 0 ? void 0 : selectedItem.supportingText) && (jsx("p", { className: "text-sm text-tertiary", children: selectedItem === null || selectedItem === void 0 ? void 0 : selectedItem.supportingText }))] }))] })) : (jsx("p", { className: cx('text-sm text-placeholder', isDisabled && 'text-disabled'), children: placeholder })), jsx(ChevronDown, { "aria-hidden": "true", className: "ml-auto shrink-0 size-4 stroke-[2.5px] text-fg-quaternary" })] }));
            } }) }));
};
function matchesSelectQuery(item, query) {
    var _a;
    if (!query)
        return true;
    const q = query.toLowerCase();
    const fields = [
        typeof item.label === 'string' ? item.label : '',
        String(item.id),
        ...((_a = item.searchAliases) !== null && _a !== void 0 ? _a : []).map(String),
    ];
    return fields.some((field) => field.toLowerCase().includes(q));
}
const Select = ({ placeholder = 'Select', placeholderIcon, children, items, preparedOptions, label: _label, hint: _hint, tooltip: _tooltip, className, searchable, texts, value, noBorder, ...rest }) => {
    var _a, _b, _c;
    const searchPlaceholder = (_a = texts === null || texts === void 0 ? void 0 : texts.searchPlaceholder) !== null && _a !== void 0 ? _a : 'Search';
    const searchNotFoundLabel = (_b = texts === null || texts === void 0 ? void 0 : texts.searchNotFound) !== null && _b !== void 0 ? _b : 'No results';
    const dataId = rest['data-id'];
    const [query, setQuery] = useState('');
    // Get selected value from props (for finding fallback item)
    const selectedValue = value;
    // Find selected item from preparedOptions first, then items
    // This ensures we can display the selected value even when filtered out
    const selectedItemFallback = (_c = preparedOptions === null || preparedOptions === void 0 ? void 0 : preparedOptions.find((item) => item.id === selectedValue)) !== null && _c !== void 0 ? _c : items === null || items === void 0 ? void 0 : items.find((item) => item.id === selectedValue);
    const filteredItems = !searchable || !query.trim()
        ? items
        : (items !== null && items !== void 0 ? items : []).filter((item) => matchesSelectQuery(item, query.trim()));
    return (jsx(Select$1, { ...rest, value: value, className: (state) => cx('flex flex-col gap-1.5', typeof className === 'function' ? className(state) : className), onOpenChange: (isOpen) => {
            if (!isOpen && query && !(filteredItems === null || filteredItems === void 0 ? void 0 : filteredItems.length)) {
                setQuery('');
            }
        }, children: (state) => (jsxs(Fragment, { children: [jsx(SelectValue, { ...state, placeholder: placeholder, placeholderIcon: placeholderIcon, isInvalid: rest.isInvalid, selectedItemFallback: selectedItemFallback, noBorder: noBorder }), jsxs(Popover, { className: cx(
                    // When searchable, override the popover's own scroll/padding
                    // so the search input is a fixed header and only the listbox
                    // scrolls. Mirrors the ColumnsSelector dropdown pattern.
                    searchable && 'flex flex-col overflow-hidden! py-0!', rest.popoverClassName), children: [searchable && (jsx(PopoverSearchInput, { "data-id": `${dataId !== null && dataId !== void 0 ? dataId : 'select'}-search`, value: query, onChange: setQuery, placeholder: searchPlaceholder })), jsx(ListBox, { items: filteredItems, className: cx('outline-hidden', searchable ? 'flex-1 overflow-y-auto' : 'size-full'), renderEmptyState: searchable
                                ? () => (jsx("div", { className: "px-3 py-4 text-sm text-tertiary", role: "status", children: searchNotFoundLabel }))
                                : undefined, children: children })] })] })) }));
};
const _Select = Select;
_Select.ComboBox = ComboBox;
_Select.Item = SelectItem;

export { ComboBox as C, Popover as P, SelectItem as S, _Select as _, useResizeObserver as u };
