// @hostaway/design-system/dist/components/_shared/ButtonUtility-CI_sdBoE.cjs.js
'use strict';

var jsxRuntime = require('react/jsx-runtime');
var h = require('react');
var reactAriaComponents = require('react-aria-components');
var tooltip = require('./tooltip-DDB8niG0.cjs.js');
var cx = require('./cx-CmkbyMOt.cjs.js');
var isReactComponent = require('./is-react-component-DkG20B86.cjs.js');

const styles = {
    secondary: 'bg-primary text-fg-quaternary shadow-xs-skeumorphic ring-1 ring-primary ring-inset hover:bg-primary_hover hover:text-fg-quaternary_hover disabled:bg-primary disabled:shadow-xs disabled:ring-disabled_subtle',
    tertiary: 'text-fg-quaternary hover:bg-primary_hover hover:text-fg-quaternary_hover disabled:bg-transparent',
};
const BaseButtonUtility = ({ tooltip: tooltip$1, className, isDisabled, icon: Icon, size = 'sm', color = 'secondary', tooltipPlacement = 'top', ...otherProps }) => {
    const href = 'href' in otherProps ? otherProps.href : undefined;
    const Component = href ? reactAriaComponents.Link : reactAriaComponents.Button;
    // We don't pass the native `disabled` attribute (or react-aria's `isDisabled` for buttons)
    // because a disabled <button> suppresses all pointer/focus events, which prevents the
    // tooltip from receiving the hover/focus signals it needs to show.
    // Instead we set `aria-disabled` + `data-rac`/`data-disabled` so the element stays
    // interactive at the event level while remaining semantically disabled, and rely on the
    // `disabled:` Tailwind variant (via tailwindcss-react-aria-components) for styling.
    const disabledProps = isDisabled
        ? {
            'aria-disabled': true,
            'data-rac': true,
            'data-disabled': true,
        }
        : {};
    // eslint-disable-next-line no-useless-assignment
    let props = {};
    if (href) {
        props = {
            ...otherProps,
            href: isDisabled ? undefined : href,
            ...disabledProps,
        };
    }
    else {
        const { onClick } = otherProps;
        props = {
            ...otherProps,
            type: otherProps.type || 'button',
            ...disabledProps,
            onClick: isDisabled
                ? (e) => {
                    e.preventDefault();
                    e.stopPropagation();
                }
                : onClick,
        };
    }
    const content = (jsxRuntime.jsxs(Component, { "aria-label": tooltip$1, ...props, className: cx.cx('group relative inline-flex h-max cursor-pointer items-center justify-center rounded-md p-1.5 outline-focus-ring transition duration-100 ease-linear focus-visible:outline-2 focus-visible:outline-offset-2 disabled:cursor-not-allowed disabled:text-fg-disabled_subtle disabled:*:data-icon:text-fg-disabled_subtle', styles[color], 
        // Icon styles
        '*:data-icon:pointer-events-none *:data-icon:shrink-0 *:data-icon:text-current *:data-icon:transition-inherit-all', size === 'xs' ? '*:data-icon:size-4' : '*:data-icon:size-5', isDisabled && 'cursor-not-allowed', className), children: [isReactComponent.isReactComponent(Icon) && jsxRuntime.jsx(Icon, { "data-icon": true }), h.isValidElement(Icon) && Icon] }));
    if (tooltip$1) {
        return (jsxRuntime.jsx(tooltip.Tooltip, { title: tooltip$1, placement: tooltipPlacement, offset: size === 'xs' ? 4 : 6, children: jsxRuntime.jsx(tooltip.TooltipTrigger, { className: cx.cx(isDisabled && 'cursor-not-allowed'), children: content }) }));
    }
    return content;
};

function ButtonUtility({ ...props }) {
    return jsxRuntime.jsx(BaseButtonUtility, { ...props });
}

exports.ButtonUtility = ButtonUtility;
