// @hostaway/design-system/dist/components/_shared/close-button-CGPYesyF.esm.js
import { jsx } from 'react/jsx-runtime';
import { X } from '@untitledui/icons';
import { Button } from 'react-aria-components';
import { c as cx } from './cx-DnIR8RkH.esm.js';

const sizes = {
    xs: { root: "size-7", icon: "size-4" },
    sm: { root: "size-9", icon: "size-5" },
    md: { root: "size-10", icon: "size-5" },
    lg: { root: "size-11", icon: "size-6" },
};
const themes = {
    light: "text-fg-quaternary hover:bg-primary_hover hover:text-fg-quaternary_hover focus-visible:outline-2 focus-visible:outline-offset-2 outline-focus-ring",
    dark: "text-fg-white/70 hover:text-fg-white hover:bg-white/20 focus-visible:outline-2 focus-visible:outline-offset-2 outline-focus-ring",
};
const CloseButton = ({ label, className, size = "sm", theme = "light", ...otherProps }) => {
    return (jsx(Button, { ...otherProps, "aria-label": label || "Close", className: (state) => cx("flex cursor-pointer items-center justify-center rounded-lg p-2 transition duration-100 ease-linear focus:ring-2 focus:ring-offset-1 focus-visible:outline-gray-500", sizes[size].root, themes[theme], typeof className === "function" ? className(state) : className), children: jsx(X, { "aria-hidden": "true", className: cx("shrink-0 transition-inherit-all", sizes[size].icon) }) }));
};

export { CloseButton as C };
