// @hostaway/design-system/dist/components/_shared/FieldWrapper-UtX7P_J-.esm.js
import { jsx, jsxs } from 'react/jsx-runtime';
import { Text, Label } from 'react-aria-components';
import { T as Tooltip, a as TooltipTrigger } from './tooltip-l8auK4sw.esm.js';
import { useRef, useState, useCallback, useEffect } from 'react';
import { c as cx } from './cx-DnIR8RkH.esm.js';
import { H as HintText } from './hint-text-Ck8Vj_nN.esm.js';
import { HelpCircle } from '@untitledui/icons';
import { C as ConditionalTooltip } from './ConditionalTooltip-DT2TnX8y.esm.js';

/**
 * Detects whether the referenced element's content overflows its container.
 *
 * - `axis: 'x'` — single-line / horizontal truncation (`scrollWidth > clientWidth`).
 *   Pair with `whitespace-nowrap overflow-hidden text-ellipsis`.
 * - `axis: 'y'` — multi-line / vertical truncation (`scrollHeight > clientHeight`).
 *   Pair with `line-clamp-N` or a fixed-height container with `overflow-hidden`.
 *
 * Re-checks on element resize via `ResizeObserver`.
 */
function useTruncation(axis) {
    const ref = useRef(null);
    const [isTruncated, setIsTruncated] = useState(false);
    const check = useCallback(() => {
        const el = ref.current;
        if (!el)
            return;
        setIsTruncated(axis === 'y'
            ? el.scrollHeight > el.clientHeight
            : el.scrollWidth > el.clientWidth);
    }, [axis]);
    useEffect(() => {
        const el = ref.current;
        if (!el)
            return;
        check();
        const ro = new ResizeObserver(check);
        ro.observe(el);
        return () => ro.disconnect();
    }, [check]);
    return { ref, isTruncated };
}

function Hint({ isWarning, children, ...props }) {
    return (jsx(HintText, { ...props, className: cx(isWarning && 'text-warning-primary'), children: children }));
}

// ─── Style Maps ──────────────────────────────────────────────────────
const CONTROL_SIZE_STYLES = {
    sm: 'w-[180px]',
    md: 'w-[352px]',
    full: 'flex-1',
    auto: 'w-fit',
};
const LABEL_BASE_WIDTH = 'w-[160px]';
// ─── Helpers ─────────────────────────────────────────────────────────
function getControlClassName(size) {
    return cx('flex min-w-0 flex-col gap-0.5', CONTROL_SIZE_STYLES[size]);
}
// ─── Component ───────────────────────────────────────────────────────
function FieldWrapper({ id = '', labelPosition = 'left', label, hint, children, isRequired, isInvalid, isWarning, tooltip, fieldSize = 'md', description, alignLabel = 'center', }) {
    const fieldId = id;
    const descriptionId = `${fieldId}-description`;
    const isLeft = labelPosition === 'left';
    const hasAdornments = isRequired || tooltip;
    const { ref: labelRef, isTruncated: isLabelTruncated } = useTruncation('y');
    const { ref: descRef, isTruncated: isDescTruncated } = useTruncation('y');
    // ── Shared pieces ───────────────────────────────────────────────
    const descriptionContent = description && (jsx(ConditionalTooltip, { text: description, isTruncated: isDescTruncated, children: jsx(Text, { id: descriptionId, slot: "description", ref: descRef, className: "line-clamp-2 text-xs leading-snug text-gray-500", children: description }) }));
    const hintContent = hint && (jsx(Hint, { isInvalid: isInvalid, isWarning: isWarning, children: hint }));
    // ── Label + Description ─────────────────────────────────────────
    const labelContent = label && (jsxs(Label, { htmlFor: fieldId, className: "flex items-start gap-1 text-sm font-normal leading-snug text-gray-900", "aria-label": label, children: [jsx(ConditionalTooltip, { text: label, isTruncated: isLabelTruncated, children: jsx("span", { ref: labelRef, className: "min-w-0 line-clamp-2", children: label }) }), hasAdornments && (jsxs("span", { className: "inline-flex shrink-0 items-center gap-1", children: [isRequired && (jsx("span", { className: cx('font-bold', isInvalid ? 'text-red-500' : 'text-brand-secondary'), "aria-hidden": "true", children: "*" })), tooltip && (jsx(Tooltip, { title: tooltip, placement: "top", children: jsx(TooltipTrigger, { isDisabled: false, className: "shrink-0 cursor-pointer text-fg-quaternary\n                           transition duration-200\n                           hover:text-fg-quaternary_hover\n                           focus:text-fg-quaternary_hover", children: jsx(HelpCircle, { className: "size-4 mt-[2px]" }) }) }))] }))] }));
    // ── TOP layout ──────────────────────────────────────────────────
    if (!isLeft) {
        return (jsx("div", { "data-invalid": isInvalid || undefined, children: jsxs("div", { className: getControlClassName(fieldSize), children: [(labelContent || descriptionContent) && (jsxs("div", { className: "flex flex-col", children: [labelContent, descriptionContent] })), children, hintContent] }) }));
    }
    // ── LEFT layout ─────────────────────────────────────────────────
    return (jsxs("div", { className: "flex items-start gap-2", "data-invalid": isInvalid || undefined, children: [(label || descriptionContent) && (jsx("div", { className: cx('flex min-h-[36px]', alignLabel === 'center' && 'items-center', alignLabel === 'top' && 'items-start'), children: jsxs("div", { className: cx('flex shrink-0 flex-col', LABEL_BASE_WIDTH), children: [labelContent, descriptionContent] }) })), jsxs("div", { className: getControlClassName(fieldSize), children: [children, hintContent] })] }));
}

export { FieldWrapper as F, Hint as H, useTruncation as u };
