// @hostaway/design-system/dist/components/_shared/FieldWrapper-D6dXzMyx.cjs.js
'use strict';

var jsxRuntime = require('react/jsx-runtime');
var reactAriaComponents = require('react-aria-components');
var tooltip = require('./tooltip-DDB8niG0.cjs.js');
var h = require('react');
var cx = require('./cx-CmkbyMOt.cjs.js');
var hintText = require('./hint-text-LRIVuQnk.cjs.js');
var icons = require('@untitledui/icons');
var ConditionalTooltip = require('./ConditionalTooltip-FLw9yEoi.cjs.js');

/**
 * Detects whether the referenced element's content overflows its container.
 *
 * - `axis: 'x'` — single-line / horizontal truncation (`scrollWidth > clientWidth`).
 *   Pair with `whitespace-nowrap overflow-hidden text-ellipsis`.
 * - `axis: 'y'` — multi-line / vertical truncation (`scrollHeight > clientHeight`).
 *   Pair with `line-clamp-N` or a fixed-height container with `overflow-hidden`.
 *
 * Re-checks on element resize via `ResizeObserver`.
 */
function useTruncation(axis) {
    const ref = h.useRef(null);
    const [isTruncated, setIsTruncated] = h.useState(false);
    const check = h.useCallback(() => {
        const el = ref.current;
        if (!el)
            return;
        setIsTruncated(axis === 'y'
            ? el.scrollHeight > el.clientHeight
            : el.scrollWidth > el.clientWidth);
    }, [axis]);
    h.useEffect(() => {
        const el = ref.current;
        if (!el)
            return;
        check();
        const ro = new ResizeObserver(check);
        ro.observe(el);
        return () => ro.disconnect();
    }, [check]);
    return { ref, isTruncated };
}

function Hint({ isWarning, children, ...props }) {
    return (jsxRuntime.jsx(hintText.HintText, { ...props, className: cx.cx(isWarning && 'text-warning-primary'), children: children }));
}

// ─── Style Maps ──────────────────────────────────────────────────────
const CONTROL_SIZE_STYLES = {
    sm: 'w-[180px]',
    md: 'w-[352px]',
    full: 'flex-1',
    auto: 'w-fit',
};
const LABEL_BASE_WIDTH = 'w-[160px]';
// ─── Helpers ─────────────────────────────────────────────────────────
function getControlClassName(size) {
    return cx.cx('flex min-w-0 flex-col gap-0.5', CONTROL_SIZE_STYLES[size]);
}
// ─── Component ───────────────────────────────────────────────────────
function FieldWrapper({ id = '', labelPosition = 'left', label, hint, children, isRequired, isInvalid, isWarning, tooltip: tooltip$1, fieldSize = 'md', description, alignLabel = 'center', }) {
    const fieldId = id;
    const descriptionId = `${fieldId}-description`;
    const isLeft = labelPosition === 'left';
    const hasAdornments = isRequired || tooltip$1;
    const { ref: labelRef, isTruncated: isLabelTruncated } = useTruncation('y');
    const { ref: descRef, isTruncated: isDescTruncated } = useTruncation('y');
    // ── Shared pieces ───────────────────────────────────────────────
    const descriptionContent = description && (jsxRuntime.jsx(ConditionalTooltip.ConditionalTooltip, { text: description, isTruncated: isDescTruncated, children: jsxRuntime.jsx(reactAriaComponents.Text, { id: descriptionId, slot: "description", ref: descRef, className: "line-clamp-2 text-xs leading-snug text-gray-500", children: description }) }));
    const hintContent = hint && (jsxRuntime.jsx(Hint, { isInvalid: isInvalid, isWarning: isWarning, children: hint }));
    // ── Label + Description ─────────────────────────────────────────
    const labelContent = label && (jsxRuntime.jsxs(reactAriaComponents.Label, { htmlFor: fieldId, className: "flex items-start gap-1 text-sm font-normal leading-snug text-gray-900", "aria-label": label, children: [jsxRuntime.jsx(ConditionalTooltip.ConditionalTooltip, { text: label, isTruncated: isLabelTruncated, children: jsxRuntime.jsx("span", { ref: labelRef, className: "min-w-0 line-clamp-2", children: label }) }), hasAdornments && (jsxRuntime.jsxs("span", { className: "inline-flex shrink-0 items-center gap-1", children: [isRequired && (jsxRuntime.jsx("span", { className: cx.cx('font-bold', isInvalid ? 'text-red-500' : 'text-brand-secondary'), "aria-hidden": "true", children: "*" })), tooltip$1 && (jsxRuntime.jsx(tooltip.Tooltip, { title: tooltip$1, placement: "top", children: jsxRuntime.jsx(tooltip.TooltipTrigger, { isDisabled: false, className: "shrink-0 cursor-pointer text-fg-quaternary\n                           transition duration-200\n                           hover:text-fg-quaternary_hover\n                           focus:text-fg-quaternary_hover", children: jsxRuntime.jsx(icons.HelpCircle, { className: "size-4 mt-[2px]" }) }) }))] }))] }));
    // ── TOP layout ──────────────────────────────────────────────────
    if (!isLeft) {
        return (jsxRuntime.jsx("div", { "data-invalid": isInvalid || undefined, children: jsxRuntime.jsxs("div", { className: getControlClassName(fieldSize), children: [(labelContent || descriptionContent) && (jsxRuntime.jsxs("div", { className: "flex flex-col", children: [labelContent, descriptionContent] })), children, hintContent] }) }));
    }
    // ── LEFT layout ─────────────────────────────────────────────────
    return (jsxRuntime.jsxs("div", { className: "flex items-start gap-2", "data-invalid": isInvalid || undefined, children: [(label || descriptionContent) && (jsxRuntime.jsx("div", { className: cx.cx('flex min-h-[36px]', alignLabel === 'center' && 'items-center', alignLabel === 'top' && 'items-start'), children: jsxRuntime.jsxs("div", { className: cx.cx('flex shrink-0 flex-col', LABEL_BASE_WIDTH), children: [labelContent, descriptionContent] }) })), jsxRuntime.jsxs("div", { className: getControlClassName(fieldSize), children: [children, hintContent] })] }));
}

exports.FieldWrapper = FieldWrapper;
exports.Hint = Hint;
exports.useTruncation = useTruncation;
