// @hostaway/design-system/dist/components/_shared/input-BOpcbRQD.esm.js
import { jsx, jsxs, Fragment } from 'react/jsx-runtime';
import { createContext, useContext } from 'react';
import { HelpCircle, InfoCircle } from '@untitledui/icons';
import './MyCustomIcon-BRHTusqN.esm.js';
import { TextField as TextField$1, Group, Input as Input$1 } from 'react-aria-components';
import { H as HintText } from './hint-text-Ck8Vj_nN.esm.js';
import { L as Label } from './label-pgzo6IFb.esm.js';
import { T as Tooltip, a as TooltipTrigger } from './tooltip-l8auK4sw.esm.js';
import { c as cx, s as sortCx } from './cx-DnIR8RkH.esm.js';

const InputBase = ({ ref, tooltip, shortcut, groupRef, size = 'md', isInvalid, isDisabled, icon: Icon, placeholder, wrapperClassName, tooltipClassName, inputClassName, iconClassName, 
// Omit this prop to avoid invalid HTML attribute warning
isRequired: _isRequired, ...inputProps }) => {
    // Check if the input has a leading icon or tooltip
    const hasTrailingIcon = tooltip || isInvalid;
    const hasLeadingIcon = Icon;
    // If the input is inside a `TextFieldWrapper`, use its context to simplify applying styles
    const context = useContext(TextFieldWrapper);
    const inputSize = (context === null || context === void 0 ? void 0 : context.size) || size;
    const sizes = sortCx({
        sm: {
            root: cx('px-3 py-2', hasTrailingIcon && 'pr-9', hasLeadingIcon && 'pl-10'),
            iconLeading: 'left-3',
            iconTrailing: 'right-3',
            shortcut: 'pr-2.5',
        },
        md: {
            root: cx('px-3.5 py-2.5', hasTrailingIcon && 'pr-9.5', hasLeadingIcon && 'pl-10.5'),
            iconLeading: 'left-3.5',
            iconTrailing: 'right-3.5',
            shortcut: 'pr-3',
        },
    });
    return (jsxs(Group, { isDisabled, isInvalid, ref: groupRef, className: ({ isFocusWithin, isDisabled, isInvalid }) => cx('relative flex w-full flex-row place-content-center place-items-center rounded-lg bg-primary shadow-xs ring-1 ring-primary transition-shadow duration-100 ease-linear ring-inset', isFocusWithin && !isDisabled && 'ring-2 ring-brand', 
        // Disabled state styles
        isDisabled && 'cursor-not-allowed bg-disabled_subtle ring-disabled', 'group-disabled:cursor-not-allowed group-disabled:bg-disabled_subtle group-disabled:ring-disabled', 
        // Invalid state styles
        isInvalid && 'ring-error_subtle', 'group-invalid:ring-error_subtle', 
        // Invalid state with focus-within styles
        isInvalid && isFocusWithin && 'ring-2 ring-error', isFocusWithin && 'group-invalid:ring-2 group-invalid:ring-error', context === null || context === void 0 ? void 0 : context.wrapperClassName, wrapperClassName), children: [Icon && (jsx(Icon, { className: cx('pointer-events-none absolute size-5 text-fg-quaternary', isDisabled && 'text-fg-disabled', sizes[inputSize].iconLeading, context === null || context === void 0 ? void 0 : context.iconClassName, iconClassName) })), jsx(Input$1, { ...inputProps, ref: ref, placeholder: placeholder, className: cx('m-0 w-full bg-transparent text-md text-primary ring-0 outline-hidden placeholder:text-placeholder autofill:rounded-lg autofill:text-primary', isDisabled && 'cursor-not-allowed text-disabled', sizes[inputSize].root, context === null || context === void 0 ? void 0 : context.inputClassName, inputClassName) }), tooltip && !isInvalid && (jsx(Tooltip, { title: tooltip, placement: "top", children: jsx(TooltipTrigger, { className: cx('absolute cursor-pointer text-fg-quaternary transition duration-200 hover:text-fg-quaternary_hover focus:text-fg-quaternary_hover', sizes[inputSize].iconTrailing, context === null || context === void 0 ? void 0 : context.tooltipClassName, tooltipClassName), children: jsx(HelpCircle, { className: "size-4" }) }) })), isInvalid && (jsx(InfoCircle, { className: cx('pointer-events-none absolute size-4 text-fg-error-secondary', sizes[inputSize].iconTrailing, context === null || context === void 0 ? void 0 : context.tooltipClassName, tooltipClassName) })), shortcut && (jsx("div", { className: cx('pointer-events-none absolute inset-y-0.5 right-0.5 z-10 flex items-center rounded-r-[inherit] bg-linear-to-r from-transparent to-bg-primary to-40% pl-8', sizes[inputSize].shortcut), children: jsx("span", { className: cx('pointer-events-none rounded px-1 py-px text-xs font-medium text-quaternary ring-1 ring-secondary select-none ring-inset', isDisabled && 'bg-transparent text-disabled'), "aria-hidden": "true", children: typeof shortcut === 'string' ? shortcut : '⌘K' }) }))] }));
};
InputBase.displayName = 'InputBase';
const TextFieldWrapper = createContext({});
const TextField = ({ className, ...props }) => {
    return (jsx(TextFieldWrapper.Provider, { value: props, children: jsx(TextField$1, { ...props, "data-input-wrapper": true, className: (state) => cx('group flex h-max w-full flex-col items-start justify-start gap-1.5', typeof className === 'function' ? className(state) : className) }) }));
};
TextField.displayName = 'TextField';
const Input = ({ size = 'md', placeholder, icon: Icon, label, hint, shortcut, isRequired, className, ref, groupRef, tooltip, iconClassName, inputClassName, wrapperClassName, tooltipClassName, ...props }) => {
    return (jsx(TextField, { "aria-label": !label ? placeholder : undefined, ...props, className: className, children: ({ isInvalid }) => (jsxs(Fragment, { children: [label && jsx(Label, { isRequired: isRequired, children: label }), jsx(InputBase, { ref,
                    groupRef,
                    size,
                    placeholder,
                    icon: Icon,
                    shortcut,
                    iconClassName,
                    inputClassName,
                    wrapperClassName,
                    tooltipClassName,
                    tooltip }), hint && jsx(HintText, { isInvalid: isInvalid, children: hint })] })) }));
};
Input.displayName = 'Input';

export { Input as I, TextField as T, InputBase as a };
