// @hostaway/design-system/dist/components/_shared/input-CVSUknGT.cjs.js
'use strict';

var jsxRuntime = require('react/jsx-runtime');
var h = require('react');
var icons = require('@untitledui/icons');
require('./MyCustomIcon-DYQF4lJT.cjs.js');
var reactAriaComponents = require('react-aria-components');
var hintText = require('./hint-text-LRIVuQnk.cjs.js');
var label = require('./label-D0WtNYRO.cjs.js');
var tooltip = require('./tooltip-DDB8niG0.cjs.js');
var cx = require('./cx-CmkbyMOt.cjs.js');

const InputBase = ({ ref, tooltip: tooltip$1, shortcut, groupRef, size = 'md', isInvalid, isDisabled, icon: Icon, placeholder, wrapperClassName, tooltipClassName, inputClassName, iconClassName, 
// Omit this prop to avoid invalid HTML attribute warning
isRequired: _isRequired, ...inputProps }) => {
    // Check if the input has a leading icon or tooltip
    const hasTrailingIcon = tooltip$1 || isInvalid;
    const hasLeadingIcon = Icon;
    // If the input is inside a `TextFieldWrapper`, use its context to simplify applying styles
    const context = h.useContext(TextFieldWrapper);
    const inputSize = (context === null || context === void 0 ? void 0 : context.size) || size;
    const sizes = cx.sortCx({
        sm: {
            root: cx.cx('px-3 py-2', hasTrailingIcon && 'pr-9', hasLeadingIcon && 'pl-10'),
            iconLeading: 'left-3',
            iconTrailing: 'right-3',
            shortcut: 'pr-2.5',
        },
        md: {
            root: cx.cx('px-3.5 py-2.5', hasTrailingIcon && 'pr-9.5', hasLeadingIcon && 'pl-10.5'),
            iconLeading: 'left-3.5',
            iconTrailing: 'right-3.5',
            shortcut: 'pr-3',
        },
    });
    return (jsxRuntime.jsxs(reactAriaComponents.Group, { isDisabled, isInvalid, ref: groupRef, className: ({ isFocusWithin, isDisabled, isInvalid }) => cx.cx('relative flex w-full flex-row place-content-center place-items-center rounded-lg bg-primary shadow-xs ring-1 ring-primary transition-shadow duration-100 ease-linear ring-inset', isFocusWithin && !isDisabled && 'ring-2 ring-brand', 
        // Disabled state styles
        isDisabled && 'cursor-not-allowed bg-disabled_subtle ring-disabled', 'group-disabled:cursor-not-allowed group-disabled:bg-disabled_subtle group-disabled:ring-disabled', 
        // Invalid state styles
        isInvalid && 'ring-error_subtle', 'group-invalid:ring-error_subtle', 
        // Invalid state with focus-within styles
        isInvalid && isFocusWithin && 'ring-2 ring-error', isFocusWithin && 'group-invalid:ring-2 group-invalid:ring-error', context === null || context === void 0 ? void 0 : context.wrapperClassName, wrapperClassName), children: [Icon && (jsxRuntime.jsx(Icon, { className: cx.cx('pointer-events-none absolute size-5 text-fg-quaternary', isDisabled && 'text-fg-disabled', sizes[inputSize].iconLeading, context === null || context === void 0 ? void 0 : context.iconClassName, iconClassName) })), jsxRuntime.jsx(reactAriaComponents.Input, { ...inputProps, ref: ref, placeholder: placeholder, className: cx.cx('m-0 w-full bg-transparent text-md text-primary ring-0 outline-hidden placeholder:text-placeholder autofill:rounded-lg autofill:text-primary', isDisabled && 'cursor-not-allowed text-disabled', sizes[inputSize].root, context === null || context === void 0 ? void 0 : context.inputClassName, inputClassName) }), tooltip$1 && !isInvalid && (jsxRuntime.jsx(tooltip.Tooltip, { title: tooltip$1, placement: "top", children: jsxRuntime.jsx(tooltip.TooltipTrigger, { className: cx.cx('absolute cursor-pointer text-fg-quaternary transition duration-200 hover:text-fg-quaternary_hover focus:text-fg-quaternary_hover', sizes[inputSize].iconTrailing, context === null || context === void 0 ? void 0 : context.tooltipClassName, tooltipClassName), children: jsxRuntime.jsx(icons.HelpCircle, { className: "size-4" }) }) })), isInvalid && (jsxRuntime.jsx(icons.InfoCircle, { className: cx.cx('pointer-events-none absolute size-4 text-fg-error-secondary', sizes[inputSize].iconTrailing, context === null || context === void 0 ? void 0 : context.tooltipClassName, tooltipClassName) })), shortcut && (jsxRuntime.jsx("div", { className: cx.cx('pointer-events-none absolute inset-y-0.5 right-0.5 z-10 flex items-center rounded-r-[inherit] bg-linear-to-r from-transparent to-bg-primary to-40% pl-8', sizes[inputSize].shortcut), children: jsxRuntime.jsx("span", { className: cx.cx('pointer-events-none rounded px-1 py-px text-xs font-medium text-quaternary ring-1 ring-secondary select-none ring-inset', isDisabled && 'bg-transparent text-disabled'), "aria-hidden": "true", children: typeof shortcut === 'string' ? shortcut : '⌘K' }) }))] }));
};
InputBase.displayName = 'InputBase';
const TextFieldWrapper = h.createContext({});
const TextField = ({ className, ...props }) => {
    return (jsxRuntime.jsx(TextFieldWrapper.Provider, { value: props, children: jsxRuntime.jsx(reactAriaComponents.TextField, { ...props, "data-input-wrapper": true, className: (state) => cx.cx('group flex h-max w-full flex-col items-start justify-start gap-1.5', typeof className === 'function' ? className(state) : className) }) }));
};
TextField.displayName = 'TextField';
const Input = ({ size = 'md', placeholder, icon: Icon, label: label$1, hint, shortcut, isRequired, className, ref, groupRef, tooltip, iconClassName, inputClassName, wrapperClassName, tooltipClassName, ...props }) => {
    return (jsxRuntime.jsx(TextField, { "aria-label": !label$1 ? placeholder : undefined, ...props, className: className, children: ({ isInvalid }) => (jsxRuntime.jsxs(jsxRuntime.Fragment, { children: [label$1 && jsxRuntime.jsx(label.Label, { isRequired: isRequired, children: label$1 }), jsxRuntime.jsx(InputBase, { ref,
                    groupRef,
                    size,
                    placeholder,
                    icon: Icon,
                    shortcut,
                    iconClassName,
                    inputClassName,
                    wrapperClassName,
                    tooltipClassName,
                    tooltip }), hint && jsxRuntime.jsx(hintText.HintText, { isInvalid: isInvalid, children: hint })] })) }));
};
Input.displayName = 'Input';

exports.Input = Input;
exports.InputBase = InputBase;
exports.TextField = TextField;
