// @hostaway/design-system/dist/components/_shared/MenuPrimitives-CGzbU4z4.esm.js
import { jsxs, jsx } from 'react/jsx-runtime';
import { MenuItem as MenuItem$1, Menu, Popover, MenuTrigger as MenuTrigger$1 } from 'react-aria-components';
import { Check } from '@untitledui/icons';
import './MyCustomIcon-BRHTusqN.esm.js';
import { a as CheckboxBase } from './checkbox-oR9JZpTn.esm.js';
import { c as cx } from './cx-DnIR8RkH.esm.js';

function MenuTrigger(props) {
    return jsx(MenuTrigger$1, { ...props });
}
function MenuPopover({ className, ...props }) {
    return (jsx(Popover, { containerPadding: 0, className: cx('max-w-[min(32rem,90vw)] rounded-lg bg-primary shadow-lg ring ring-secondary_alt', className), ...props }));
}
function MenuList({ className, ...props }) {
    return (jsx(Menu, { className: cx('flex flex-col gap-0.5 px-1.5 py-1 outline-hidden select-none overflow-y-auto', className), ...props }));
}
function MenuItem({ className, leading, trailing, withCheckbox, isSelected, withCheck, children, ...props }) {
    return (jsxs(MenuItem$1, { className: cx('cursor-pointer rounded-md px-2.5 py-2 hover:bg-primary_hover text-secondary text-sm outline-hidden', (leading || trailing || withCheckbox || withCheck) &&
            'flex items-center gap-2', isSelected && 'bg-primary_hover', className), "aria-selected": isSelected || undefined, ...props, children: [withCheckbox && (jsx(CheckboxBase, { isSelected: isSelected, size: "sm" })), leading, jsx("span", { className: "flex-1 truncate", children: children }), trailing, withCheck && isSelected && jsx(Check, { className: "size-3.5 shrink-0" })] }));
}

export { MenuTrigger as M, MenuPopover as a, MenuList as b, MenuItem as c };
