// @hostaway/design-system/dist/components/_shared/PageHeader-CtA7v5jh.esm.js
import { jsx, jsxs } from 'react/jsx-runtime';
import { createContext, useContext } from 'react';
import { ArrowLeft } from '@untitledui/icons';
import { c as cx } from './cx-DnIR8RkH.esm.js';
import { B as Button } from './Button-C3YUMQZU.esm.js';

const defaultStyles = {
    gridTemplateAreas: "[grid-template-areas:'back_back'_'header_header'_'divider_divider'_'content_actions']",
    padding: 'px-6 pt-3',
    backNavigationClassName: 'hover:bg-transparent pl-0',
    withDivider: true,
    actionsStyles: '',
};
const variants = {
    primary: {
        ...defaultStyles,
        actionsStyles: 'border-b border-secondary w-full h-full flex justify-end items-center',
    },
    secondary: {
        ...defaultStyles,
        gridTemplateAreas: "[grid-template-areas:'back_back'_'header_header'_'subHeader_actions'_'divider_divider'_'content_content']",
    },
    tertiary: {
        ...defaultStyles,
        gridTemplateAreas: "[grid-template-areas:'back_topActions'_'header_header'_'subHeader_actions']",
        padding: 'px-6 py-3',
        backNavigationClassName: 'hover:bg-transparent p-0',
        withDivider: false,
    },
    quaternary: {
        ...defaultStyles,
        gridTemplateAreas: "[grid-template-areas:'back_back'_'header_actions']",
        padding: 'px-6 py-3',
        withDivider: false,
    },
};
const PageHeaderContext = createContext(variants.primary);
function BackNavigation({ name, onClick, 'data-id': dataId, ...rest }) {
    const { backNavigationClassName } = useContext(PageHeaderContext);
    return (jsx("div", { className: "[grid-area:back]", ...rest, children: jsx(Button, { className: backNavigationClassName, color: "tertiary", iconLeading: ArrowLeft, onClick: onClick, "data-id": dataId, children: name }) }));
}
function Header({ children, className, ...rest }) {
    if (typeof children === 'string') {
        return (jsx("h1", { className: cx('[grid-area:header] text-xl font-semibold text-primary lg:text-display-xs pb-1 overflow-hidden text-ellipsis whitespace-nowrap', className), ...rest, children: children }));
    }
    return (jsx("div", { className: cx('[grid-area:header] pb-1', className), ...rest, children: children }));
}
function SubHeader({ children, ...rest }) {
    if (typeof children === 'string') {
        return (jsx("p", { className: "[grid-area:subHeader] text-md text-tertiary flex items-end overflow-hidden text-ellipsis whitespace-nowrap", ...rest, children: children }));
    }
    return (jsx("div", { className: "[grid-area:subHeader] flex gap-3 items-end flex-wrap", ...rest, children: children }));
}
function PageHeader({ variant = 'primary', children, className = '', withDivider, ...rest }) {
    const variantConfig = variants[variant];
    const { gridTemplateAreas, padding } = variantConfig;
    const showDivider = withDivider !== undefined ? withDivider : variantConfig.withDivider;
    return (jsx(PageHeaderContext.Provider, { value: variantConfig, children: jsxs("header", { className: cx('grid', padding, gridTemplateAreas, className), ...rest, children: [children, showDivider && jsx(Divider, {})] }) }));
}
function Divider() {
    return jsx("div", { className: "[grid-area:divider] border-t border-secondary mt-2" });
}
function Actions({ children, className = '', ...rest }) {
    const { actionsStyles } = useContext(PageHeaderContext);
    return (jsx("div", { className: cx('[grid-area:actions] [justify-self:end] self-end', actionsStyles, className), ...rest, children: jsx("div", { className: "flex gap-3", children: children }) }));
}
function TopActions({ children, ...rest }) {
    return (jsx("div", { className: "[grid-area:topActions] [justify-self:end] self-end", ...rest, children: jsx("div", { className: "flex gap-3", children: children }) }));
}
function Content({ children, className, ...rest }) {
    return (jsx("div", { className: cx('[grid-area:content]', className), ...rest, children: children }));
}
PageHeader.BackNavigation = BackNavigation;
PageHeader.Header = Header;
PageHeader.SubHeader = SubHeader;
PageHeader.Actions = Actions;
PageHeader.TopActions = TopActions;
PageHeader.Content = Content;

export { PageHeader as P };
