// @hostaway/design-system/dist/components/_shared/pagination-base-QUPW0JXw.cjs.js
'use strict';

var jsxRuntime = require('react/jsx-runtime');
var h = require('react');

const PaginationContext = h.createContext(undefined);
const PaginationRoot = ({ total, siblingCount = 1, page, onPageChange, children, style, className, }) => {
    const [pages, setPages] = h.useState([]);
    const createPaginationItems = h.useCallback(() => {
        const items = [];
        // Calculate the maximum number of pagination elements (pages, potential ellipsis, first and last) to show
        const totalPageNumbers = siblingCount * 2 + 5;
        // If the total number of items to show is greater than or equal to the total pages,
        // we can simply list all pages without needing to collapse with ellipsis
        if (totalPageNumbers >= total) {
            for (let i = 1; i <= total; i++) {
                items.push({
                    type: 'page',
                    value: i,
                    isCurrent: i === page,
                });
            }
        }
        else {
            // Calculate left and right sibling boundaries around the current page
            const leftSiblingIndex = Math.max(page - siblingCount, 1);
            const rightSiblingIndex = Math.min(page + siblingCount, total);
            // Determine if we need to show ellipsis on either side
            const showLeftEllipsis = leftSiblingIndex > 2;
            const showRightEllipsis = rightSiblingIndex < total - 1;
            // Case 1: No left ellipsis, but right ellipsis is needed
            if (!showLeftEllipsis && showRightEllipsis) {
                // Calculate how many page numbers to show starting from the beginning
                const leftItemCount = siblingCount * 2 + 3;
                const leftRange = range(1, leftItemCount);
                leftRange.forEach((pageNum) => items.push({
                    type: 'page',
                    value: pageNum,
                    isCurrent: pageNum === page,
                }));
                // Insert ellipsis after the left range and add the last page
                items.push({ type: 'ellipsis', key: leftItemCount + 1 });
                items.push({
                    type: 'page',
                    value: total,
                    isCurrent: total === page,
                });
            }
            // Case 2: Left ellipsis needed, but right ellipsis is not needed
            else if (showLeftEllipsis && !showRightEllipsis) {
                // Determine how many items from the end should be shown
                const rightItemCount = siblingCount * 2 + 3;
                const rightRange = range(total - rightItemCount + 1, total);
                // Always show the first page, then add an ellipsis to indicate skipped pages
                items.push({
                    type: 'page',
                    value: 1,
                    isCurrent: page === 1,
                });
                items.push({ type: 'ellipsis', key: total - rightItemCount });
                rightRange.forEach((pageNum) => items.push({
                    type: 'page',
                    value: pageNum,
                    isCurrent: pageNum === page,
                }));
            }
            // Case 3: Both left and right ellipsis are needed
            else if (showLeftEllipsis && showRightEllipsis) {
                // Always show the first page
                items.push({
                    type: 'page',
                    value: 1,
                    isCurrent: page === 1,
                });
                // Insert left ellipsis after the first page
                items.push({ type: 'ellipsis', key: leftSiblingIndex - 1 });
                // Show a range of pages around the current page
                const middleRange = range(leftSiblingIndex, rightSiblingIndex);
                middleRange.forEach((pageNum) => items.push({
                    type: 'page',
                    value: pageNum,
                    isCurrent: pageNum === page,
                }));
                // Insert right ellipsis and finally the last page
                items.push({ type: 'ellipsis', key: rightSiblingIndex + 1 });
                items.push({
                    type: 'page',
                    value: total,
                    isCurrent: total === page,
                });
            }
        }
        return items;
    }, [total, siblingCount, page]);
    h.useEffect(() => {
        const paginationItems = createPaginationItems();
        setPages(paginationItems);
    }, [createPaginationItems]);
    const onPageChangeHandler = (newPage) => {
        onPageChange === null || onPageChange === void 0 ? void 0 : onPageChange(newPage);
    };
    const paginationContextValue = {
        pages,
        currentPage: page,
        total,
        onPageChange: onPageChangeHandler,
    };
    return (jsxRuntime.jsx(PaginationContext.Provider, { value: paginationContextValue, children: jsxRuntime.jsx("nav", { "aria-label": "Pagination Navigation", style: style, className: className, children: children }) }));
};
/**
 * Creates an array of numbers from start to end.
 * @param start - The start number.
 * @param end - The end number.
 * @returns An array of numbers from start to end.
 */
const range = (start, end) => {
    const length = end - start + 1;
    return Array.from({ length }, (_, index) => index + start);
};
const Trigger = ({ children, style, className, asChild = false, direction, ariaLabel, }) => {
    const context = h.useContext(PaginationContext);
    if (!context) {
        throw new Error('Pagination components must be used within a Pagination.Root');
    }
    const { currentPage, total, onPageChange } = context;
    const isDisabled = direction === 'prev' ? currentPage <= 1 : currentPage >= total;
    const handleClick = () => {
        if (isDisabled)
            return;
        const newPage = direction === 'prev' ? currentPage - 1 : currentPage + 1;
        onPageChange === null || onPageChange === void 0 ? void 0 : onPageChange(newPage);
    };
    const computedClassName = typeof className === 'function' ? className({ isDisabled }) : className;
    const defaultAriaLabel = direction === 'prev' ? 'Previous Page' : 'Next Page';
    // If the children is a render prop, we need to pass the isDisabled and onClick to the render prop.
    if (typeof children === 'function') {
        return jsxRuntime.jsx(jsxRuntime.Fragment, { children: children({ isDisabled, onClick: handleClick }) });
    }
    // If the children is a valid element, we need to clone it and pass the isDisabled and onClick to the cloned element.
    if (asChild && h.isValidElement(children)) {
        return h.cloneElement(children, {
            onClick: handleClick,
            disabled: isDisabled,
            isDisabled,
            'aria-label': ariaLabel || defaultAriaLabel,
            style: {
                ...children.props.style,
                ...style,
            },
            className: [
                computedClassName,
                children.props.className,
            ]
                .filter(Boolean)
                .join(' ') || undefined,
        });
    }
    return (jsxRuntime.jsx("button", { "aria-label": ariaLabel || defaultAriaLabel, onClick: handleClick, disabled: isDisabled, style: style, className: computedClassName, children: children }));
};
const PaginationPrevTrigger = (props) => (jsxRuntime.jsx(Trigger, { ...props, direction: "prev" }));
const PaginationNextTrigger = (props) => (jsxRuntime.jsx(Trigger, { ...props, direction: "next" }));
const PaginationItem = ({ value, isCurrent, children, style, className, ariaLabel, asChild = false, }) => {
    const context = h.useContext(PaginationContext);
    if (!context) {
        throw new Error('Pagination components must be used within a <Pagination.Root />');
    }
    const { onPageChange } = context;
    const isSelected = isCurrent;
    const handleClick = () => {
        onPageChange === null || onPageChange === void 0 ? void 0 : onPageChange(value);
    };
    const computedClassName = typeof className === 'function' ? className({ isSelected }) : className;
    // If the children is a render prop, we need to pass the necessary props to the render prop.
    if (typeof children === 'function') {
        return (jsxRuntime.jsx(jsxRuntime.Fragment, { children: children({
                isSelected,
                onClick: handleClick,
                value,
                'aria-current': isCurrent ? 'page' : undefined,
                'aria-label': ariaLabel || `Page ${value}`,
            }) }));
    }
    // If the children is a valid element, we need to clone it and pass the necessary props to the cloned element.
    if (asChild && h.isValidElement(children)) {
        return h.cloneElement(children, {
            onClick: handleClick,
            'aria-current': isCurrent ? 'page' : undefined,
            'aria-label': ariaLabel || `Page ${value}`,
            style: {
                ...children.props.style,
                ...style,
            },
            className: [
                computedClassName,
                children.props.className,
            ]
                .filter(Boolean)
                .join(' ') || undefined,
        });
    }
    return (jsxRuntime.jsx("button", { onClick: handleClick, style: style, className: computedClassName, "aria-current": isCurrent ? 'page' : undefined, "aria-label": ariaLabel || `Page ${value}`, role: "listitem", children: children }));
};
const PaginationEllipsis = ({ children, style, className, }) => {
    const computedClassName = typeof className === 'function' ? className() : className;
    return (jsxRuntime.jsx("span", { style: style, className: computedClassName, "aria-hidden": "true", children: children }));
};
const PaginationContextComponent = ({ children, }) => {
    const context = h.useContext(PaginationContext);
    if (!context) {
        throw new Error('Pagination components must be used within a Pagination.Root');
    }
    return jsxRuntime.jsx(jsxRuntime.Fragment, { children: children(context) });
};
const Pagination = {
    Root: PaginationRoot,
    PrevTrigger: PaginationPrevTrigger,
    NextTrigger: PaginationNextTrigger,
    Item: PaginationItem,
    Ellipsis: PaginationEllipsis,
    Context: PaginationContextComponent,
};

exports.Pagination = Pagination;
