// @hostaway/design-system/dist/components/_shared/SectionHeading-GVABFozG.esm.js
import { jsx, jsxs } from 'react/jsx-runtime';
import { c as cx } from './cx-DnIR8RkH.esm.js';
import { I as InputSearchField } from './InputSearchField-CTy3JMq7.esm.js';

const SectionHeaderRoot = (props) => (jsx("div", { ...props, className: cx("flex flex-col gap-5 border-b border-secondary pb-5", props.className), children: props.children }));
const SectionHeaderGroup = (props) => (jsx("div", { ...props, className: cx("relative flex flex-col items-start gap-4 md:flex-row", props.className), children: props.children }));
const SectionHeaderActions = (props) => (jsx("div", { ...props, className: cx("flex gap-3", props.className), children: props.children }));
const SectionHeaderHeading = (props) => (jsx("h2", { ...props, className: cx("text-lg font-semibold text-primary", props.className), children: props.children }));
const SectionHeaderSubheading = (props) => (jsx("p", { ...props, className: cx("text-sm text-tertiary", props.className), children: props.children }));
const SectionHeader = {
    Root: SectionHeaderRoot,
    Group: SectionHeaderGroup,
    Actions: SectionHeaderActions,
    Heading: SectionHeaderHeading,
    Subheading: SectionHeaderSubheading,
};

const SectionHeading = {
    Root,
    Group,
    Header,
    Actions,
    Search,
};
function Root(props) {
    return jsx(SectionHeader.Root, { ...props });
}
function Group(props) {
    return jsx(SectionHeader.Group, { ...props });
}
function Header({ title, subtitle, className }) {
    return (jsxs("div", { className: cx('flex flex-col gap-0.5 flex-1', className), children: [jsx(SectionHeader.Heading, { children: title }), subtitle && (jsx(SectionHeader.Subheading, { children: subtitle }))] }));
}
function Actions(props) {
    return (jsx(SectionHeader.Actions, { ...props, "data-id": props['data-id'] }));
}
function Search(props) {
    return (jsx("div", { className: "w-full md:w-80", children: jsx(InputSearchField, { ...props }) }));
}

export { SectionHeading as S };
