// @hostaway/design-system/dist/components/_shared/SectionHeading-DStys-PJ.cjs.js
'use strict';

var jsxRuntime = require('react/jsx-runtime');
var cx = require('./cx-CmkbyMOt.cjs.js');
var InputSearchField = require('./InputSearchField-jhe85A4A.cjs.js');

const SectionHeaderRoot = (props) => (jsxRuntime.jsx("div", { ...props, className: cx.cx("flex flex-col gap-5 border-b border-secondary pb-5", props.className), children: props.children }));
const SectionHeaderGroup = (props) => (jsxRuntime.jsx("div", { ...props, className: cx.cx("relative flex flex-col items-start gap-4 md:flex-row", props.className), children: props.children }));
const SectionHeaderActions = (props) => (jsxRuntime.jsx("div", { ...props, className: cx.cx("flex gap-3", props.className), children: props.children }));
const SectionHeaderHeading = (props) => (jsxRuntime.jsx("h2", { ...props, className: cx.cx("text-lg font-semibold text-primary", props.className), children: props.children }));
const SectionHeaderSubheading = (props) => (jsxRuntime.jsx("p", { ...props, className: cx.cx("text-sm text-tertiary", props.className), children: props.children }));
const SectionHeader = {
    Root: SectionHeaderRoot,
    Group: SectionHeaderGroup,
    Actions: SectionHeaderActions,
    Heading: SectionHeaderHeading,
    Subheading: SectionHeaderSubheading,
};

const SectionHeading = {
    Root,
    Group,
    Header,
    Actions,
    Search,
};
function Root(props) {
    return jsxRuntime.jsx(SectionHeader.Root, { ...props });
}
function Group(props) {
    return jsxRuntime.jsx(SectionHeader.Group, { ...props });
}
function Header({ title, subtitle, className }) {
    return (jsxRuntime.jsxs("div", { className: cx.cx('flex flex-col gap-0.5 flex-1', className), children: [jsxRuntime.jsx(SectionHeader.Heading, { children: title }), subtitle && (jsxRuntime.jsx(SectionHeader.Subheading, { children: subtitle }))] }));
}
function Actions(props) {
    return (jsxRuntime.jsx(SectionHeader.Actions, { ...props, "data-id": props['data-id'] }));
}
function Search(props) {
    return (jsxRuntime.jsx("div", { className: "w-full md:w-80", children: jsxRuntime.jsx(InputSearchField.InputSearchField, { ...props }) }));
}

exports.SectionHeading = SectionHeading;
