// @hostaway/design-system/dist/components/_shared/Tabs-Bsc8I2Mo.cjs.js
'use strict';

var jsxRuntime = require('react/jsx-runtime');
var h = require('react');
var reactAriaComponents = require('react-aria-components');
var badges = require('./badges-xpScvf89.cjs.js');
var cx = require('./cx-CmkbyMOt.cjs.js');

// Styles for different types of tab
const getTabStyles = ({ isFocusVisible, isSelected, isHovered, }) => ({
    'button-brand': cx.cx('outline-focus-ring', isFocusVisible && 'outline-2 -outline-offset-2', (isSelected || isHovered) && 'bg-brand-primary_alt text-brand-secondary'),
    'button-gray': cx.cx('outline-focus-ring', isHovered && 'bg-primary_hover text-secondary', isFocusVisible && 'outline-2 -outline-offset-2', isSelected && 'bg-active text-secondary'),
    'button-border': cx.cx('outline-focus-ring', (isSelected || isHovered) && 'bg-primary_alt text-secondary shadow-sm', isFocusVisible && 'outline-2 -outline-offset-2'),
    'button-minimal': cx.cx('rounded-lg outline-focus-ring', isHovered && 'text-secondary', isFocusVisible && 'outline-2 -outline-offset-2', isSelected &&
        'bg-primary_alt text-secondary shadow-xs ring-1 ring-primary ring-inset'),
    underline: cx.cx('rounded-none border-b-2 border-transparent outline-focus-ring', (isSelected || isHovered) && 'border-brand text-primary', isFocusVisible && 'outline-2 -outline-offset-2'),
    line: cx.cx('rounded-none border-l-2 border-transparent outline-focus-ring', (isSelected || isHovered) && 'border-brand text-primary', isFocusVisible && 'outline-2 -outline-offset-2'),
});
const sizes = {
    sm: {
        'button-brand': 'text-sm font-semibold py-2 px-3',
        'button-gray': 'text-sm font-semibold py-2 px-3',
        'button-border': 'text-sm font-semibold py-2 px-3',
        'button-minimal': 'text-sm font-semibold py-2 px-3',
        underline: 'text-sm font-semibold px-1 pb-4 pt-3',
        line: 'text-sm font-semibold pl-2.5 pr-3 py-0.5',
    },
    md: {
        'button-brand': 'text-md font-semibold py-2.5 px-3',
        'button-gray': 'text-md font-semibold py-2.5 px-3',
        'button-border': 'text-md font-semibold py-2.5 px-3',
        'button-minimal': 'text-md font-semibold py-2.5 px-3',
        underline: 'text-md font-semibold px-1 pb-4 pt-3',
        line: 'text-md font-semibold pr-3.5 pl-3 py-1',
    },
};
// Styles for different types of horizontal tabs
const getHorizontalStyles = ({ size, fullWidth, }) => ({
    'button-brand': 'gap-1',
    'button-gray': 'gap-1',
    'button-border': cx.cx('gap-1 rounded-[10px] bg-secondary_alt p-1 ring-1 ring-secondary ring-inset', size === 'md' && 'rounded-xl p-1.5'),
    'button-minimal': 'gap-0.5 rounded-lg bg-secondary_alt ring-1 ring-inset ring-secondary',
    underline: cx.cx('gap-3', fullWidth && 'w-full gap-4'),
    line: 'gap-2',
});
const getColorStyles = ({ isSelected, isHovered, }) => ({
    'button-brand': isSelected || isHovered ? 'brand' : 'gray',
    'button-gray': 'gray',
    'button-border': 'gray',
    'button-minimal': 'gray',
    underline: isSelected || isHovered ? 'brand' : 'gray',
    line: isSelected || isHovered ? 'brand' : 'gray',
});
const TabListContext = h.createContext({
    size: 'sm',
    type: 'button-brand',
});
function TabList$1({ size = 'sm', type = 'button-brand', orientation: orientationProp, fullWidth, className, children, ...otherProps }) {
    var _a;
    const context = reactAriaComponents.useSlottedContext(reactAriaComponents.TabsContext);
    const orientation = (_a = orientationProp !== null && orientationProp !== void 0 ? orientationProp : context === null || context === void 0 ? void 0 : context.orientation) !== null && _a !== void 0 ? _a : 'horizontal';
    return (jsxRuntime.jsx(TabListContext.Provider, { value: { size, type, orientation, fullWidth }, children: jsxRuntime.jsx(reactAriaComponents.TabList, { ...otherProps, className: (state) => cx.cx('group flex', getHorizontalStyles({
                size,
                fullWidth,
            })[type], orientation === 'vertical' && 'w-max flex-col', 
            // Only horizontal tabs with underline type have bottom border
            orientation === 'horizontal' &&
                type === 'underline' &&
                'relative before:absolute before:inset-x-0 before:bottom-0 before:h-[0.5px] before:bg-border-secondary', typeof className === 'function' ? className(state) : className), children: children !== null && children !== void 0 ? children : ((item) => jsxRuntime.jsx(Tab$1, { ...item, children: item.children })) }) }));
}
function TabPanel$1(props) {
    return (jsxRuntime.jsx(reactAriaComponents.TabPanel, { ...props, className: (state) => cx.cx('outline-focus-ring focus-visible:outline-2 focus-visible:outline-offset-2', typeof props.className === 'function'
            ? props.className(state)
            : props.className) }));
}
function Tab$1(props) {
    const { label, children, badge, ...otherProps } = props;
    const { size = 'sm', type = 'button-brand', fullWidth, } = h.useContext(TabListContext);
    return (jsxRuntime.jsx(reactAriaComponents.Tab, { ...otherProps, className: (prop) => cx.cx('z-10 flex h-max cursor-pointer items-center justify-center gap-2 rounded-md whitespace-nowrap text-quaternary transition duration-100 ease-linear', 'group-orientation-vertical:justify-start', fullWidth && 'w-full flex-1', sizes[size][type], getTabStyles(prop)[type], typeof props.className === 'function'
            ? props.className(prop)
            : props.className), children: (state) => (jsxRuntime.jsxs(jsxRuntime.Fragment, { children: [typeof children === 'function' ? children(state) : children || label, badge && (jsxRuntime.jsx(badges.Badge, { type: "pill-color", color: getColorStyles(state)[type], className: cx.cx('hidden transition-inherit-all md:flex -my-px'), children: badge }))] })) }));
}
function Tabs$1({ className, ...props }) {
    return (jsxRuntime.jsx(reactAriaComponents.Tabs, { keyboardActivation: "manual", ...props, className: (state) => cx.cx('flex w-full flex-col', typeof className === 'function' ? className(state) : className) }));
}
Tabs$1.Panel = TabPanel$1;
Tabs$1.List = TabList$1;
Tabs$1.Item = Tab$1;

function Tabs(props) {
    return jsxRuntime.jsx(Tabs$1, { ...props });
}
Tabs.Panel = TabPanel;
Tabs.List = TabList;
Tabs.Item = Tab;
function Tab(props) {
    return jsxRuntime.jsx(Tab$1, { ...props });
}
function TabList(props) {
    return jsxRuntime.jsx(TabList$1, { ...props });
}
function TabPanel(props) {
    return jsxRuntime.jsx(TabPanel$1, { ...props });
}

exports.Tab = Tab;
exports.TabList = TabList;
exports.TabPanel = TabPanel;
exports.Tabs = Tabs;
