// @hostaway/design-system/dist/components/_shared/tooltip-DDB8niG0.cjs.js
'use strict';

var jsxRuntime = require('react/jsx-runtime');
var reactAriaComponents = require('react-aria-components');
var cx = require('./cx-CmkbyMOt.cjs.js');

const transformOrigins = {
    top: 'origin-bottom',
    'top left': 'origin-bottom-left',
    'top right': 'origin-bottom-right',
    bottom: 'origin-top',
    'bottom left': 'origin-top-left',
    'bottom right': 'origin-top-right',
    left: 'origin-right',
    'left top': 'origin-top-right',
    'left bottom': 'origin-bottom-right',
    start: 'origin-right',
    'start top': 'origin-top-left',
    'start bottom': 'origin-bottom-left',
    right: 'origin-left',
    'right top': 'origin-top-left',
    'right bottom': 'origin-bottom-left',
    end: 'origin-left',
    'end top': 'origin-top-left',
    'end bottom': 'origin-bottom-left',
};
const Tooltip = ({ title, description, children, arrow = false, delay = 300, closeDelay = 0, trigger, isDisabled, isOpen, defaultOpen, offset = 6, crossOffset, placement = 'top', onOpenChange, ...tooltipProps }) => {
    const isTopOrBottomLeft = [
        'top left',
        'top end',
        'bottom left',
        'bottom end',
    ].includes(placement);
    const isTopOrBottomRight = [
        'top right',
        'top start',
        'bottom right',
        'bottom start',
    ].includes(placement);
    // Set negative cross offset for left and right placement to visually balance the tooltip.
    const calculatedCrossOffset = isTopOrBottomLeft
        ? -12
        : isTopOrBottomRight
            ? 12
            : 0;
    return (jsxRuntime.jsxs(reactAriaComponents.TooltipTrigger, { trigger,
        delay,
        closeDelay,
        isDisabled,
        isOpen,
        defaultOpen,
        onOpenChange, children: [children, jsxRuntime.jsx(reactAriaComponents.Tooltip, { ...tooltipProps, offset: offset, placement: placement, crossOffset: crossOffset !== null && crossOffset !== void 0 ? crossOffset : calculatedCrossOffset, className: ({ isEntering, isExiting }) => cx.cx(isEntering && 'ease-out animate-in', isExiting && 'ease-in animate-out'), children: ({ isEntering, isExiting }) => (jsxRuntime.jsxs("div", { className: cx.cx('z-50 flex max-w-xs flex-col items-start gap-1 rounded-lg bg-primary-solid px-3 shadow-lg will-change-transform', description ? 'py-3' : 'py-2', 
                    // Map placements to transform-origin
                    transformOrigins[placement], isEntering &&
                        'ease-out animate-in fade-in zoom-in-95 in-placement-left:slide-in-from-right-0.5 in-placement-right:slide-in-from-left-0.5 in-placement-top:slide-in-from-bottom-0.5 in-placement-bottom:slide-in-from-top-0.5', isExiting &&
                        'ease-in animate-out fade-out zoom-out-95 in-placement-left:slide-out-to-right-0.5 in-placement-right:slide-out-to-left-0.5 in-placement-top:slide-out-to-bottom-0.5 in-placement-bottom:slide-out-to-top-0.5'), children: [jsxRuntime.jsx("span", { className: "text-xs font-semibold text-white", children: title }), description && (jsxRuntime.jsx("span", { className: "text-xs font-medium text-tooltip-supporting-text", children: description })), arrow && (jsxRuntime.jsx(reactAriaComponents.OverlayArrow, { children: jsxRuntime.jsx("svg", { viewBox: "0 0 100 100", className: "size-2.5 fill-bg-primary-solid in-placement-left:-rotate-90 in-placement-right:rotate-90 in-placement-top:rotate-0 in-placement-bottom:rotate-180", children: jsxRuntime.jsx("path", { d: "M0,0 L35.858,35.858 Q50,50 64.142,35.858 L100,0 Z" }) }) }))] })) })] }));
};
const TooltipTrigger = ({ children, className, ...buttonProps }) => {
    return (jsxRuntime.jsx(reactAriaComponents.Button, { ...buttonProps, className: (values) => cx.cx('h-max w-max outline-hidden', typeof className === 'function' ? className(values) : className), children: children }));
};

exports.Tooltip = Tooltip;
exports.TooltipTrigger = TooltipTrigger;
