// @hostaway/design-system/dist/components/Breadcrumbs/index.cjs.js
'use strict';

var jsxRuntime = require('react/jsx-runtime');
var h = require('react');
var reactAriaComponents = require('react-aria-components');
var icons = require('@untitledui/icons');
var cx = require('../_shared/cx-CmkbyMOt.cjs.js');
var isReactComponent = require('../_shared/is-react-component-DkG20B86.cjs.js');
require('tailwind-merge');

const baseStyles = {
    text: {
        root: "",
        icon: "text-fg-quaternary group-hover:text-fg-quaternary_hover",
        label: "text-quaternary group-hover:text-tertiary_hover",
        current: { root: "", icon: "text-fg-brand-primary group-hover:text-fg-brand-primary", label: "text-brand-secondary group-hover:text-brand-secondary" },
    },
    button: {
        root: "p-1 hover:bg-primary_hover",
        icon: "text-fg-quaternary group-hover:text-fg-quaternary_hover",
        label: "px-1 text-quaternary group-hover:text-tertiary_hover",
        current: { root: "bg-primary_hover", icon: "text-fg-quaternary_hover", label: "text-fg-tertiary_hover" },
    },
};
const BreadcrumbBase = ({ href, children, icon: Icon, type = "text", current, className, ...otherProps }) => {
    const Wrapper = href ? reactAriaComponents.Link : "button";
    return (jsxRuntime.jsxs(Wrapper, { ...otherProps, href: href, className: cx.cx("group inline-flex cursor-pointer items-center justify-center gap-1 rounded-md outline-focus-ring transition duration-100 ease-linear focus-visible:outline-2 focus-visible:outline-offset-2 in-current:max-w-full", baseStyles[type].root, current && baseStyles[type].current.root, !href && !otherProps.onClick && "cursor-default", className), children: [isReactComponent.isReactComponent(Icon) && (jsxRuntime.jsx(Icon, { className: cx.cx("size-5 transition-inherit-all", baseStyles[type].icon, current && baseStyles[type].current.icon) })), h.isValidElement(Icon) && Icon, children && (jsxRuntime.jsx("span", { className: cx.cx("text-sm font-semibold whitespace-nowrap transition-inherit-all in-current:truncate", baseStyles[type].label, current && baseStyles[type].current.label), children: children }))] }));
};
const BreadcrumbItem$1 = ({ href, icon, divider, type, isEllipsis, children, onClick, 'data-id': dataId, ...otherProps }) => {
    const context = h.useContext(BreadcrumbsContext);
    type = context.type || "text";
    divider = context.divider || "chevron";
    return (jsxRuntime.jsx(reactAriaComponents.Breadcrumb, { ...otherProps, className: cx.cx("flex items-center current:overflow-hidden", type === "text" || type === "text-line" ? "gap-1.5 md:gap-2" : "gap-0.5 md:gap-1"), children: ({ isCurrent }) => (jsxRuntime.jsxs(jsxRuntime.Fragment, { children: [isEllipsis ? (jsxRuntime.jsx(BreadcrumbBase
                // The label for screen readers.
                , { "aria-label": "See all breadcrumb items", type: type === "text-line" ? "text" : type, onClick: onClick, "data-id": `${dataId}-breadcrumb-item-label`, children: "..." })) : (jsxRuntime.jsx(BreadcrumbBase, { href: href, icon: icon, current: isCurrent, type: type === "text-line" ? "text" : type, onClick: onClick, children: children })), !isCurrent && (jsxRuntime.jsx("div", { className: "shrink-0 text-fg-quaternary", children: divider === "slash" ? jsxRuntime.jsx(icons.SlashDivider, { className: "size-5" }) : jsxRuntime.jsx(icons.ChevronRight, { className: "size-4" }) }))] })) }));
};

const BreadcrumbsContext = h.createContext({
    divider: "chevron",
    type: "text",
});
const styles = {
    text: "gap-1.5 md:gap-2",
    "text-line": "pl-2 gap-1.5 md:gap-2 py-2 after:pointer-events-none after:absolute after:inset-0 after:border-b after:border-t after:border-secondary",
    button: "gap-0.5 md:gap-1",
};
const Breadcrumbs$1 = ({ children, divider = "chevron", type = "text", className, maxVisibleItems = 4, 'data-id': dataId }) => {
    const [isExpanded, setIsExpanded] = h.useState(false);
    const visibleItems = (() => {
        const childrenArray = h.Children.toArray(children);
        if (!maxVisibleItems || childrenArray.length <= maxVisibleItems || isExpanded) {
            return childrenArray;
        }
        const firstItems = childrenArray.slice(0, Math.ceil(maxVisibleItems / 2));
        const lastItems = childrenArray.slice(-Math.floor((maxVisibleItems - 1) / 2));
        const ellipsisItem = jsxRuntime.jsx(BreadcrumbItem$1, { isEllipsis: true, divider: divider, type: type, onClick: () => setIsExpanded(true), "data-id": dataId }, "ellipsis");
        return [...firstItems, ellipsisItem, ...lastItems];
    })();
    return (jsxRuntime.jsx("nav", { "aria-label": "Breadcrumbs", className: cx.cx("min-w-0", className), children: jsxRuntime.jsx(BreadcrumbsContext.Provider, { value: { divider, type }, children: jsxRuntime.jsx(reactAriaComponents.Breadcrumbs, { className: cx.cx("relative flex", styles[type]), children: visibleItems }) }) }));
};
Breadcrumbs$1.Item = BreadcrumbItem$1;

function BreadcrumbItem(props) {
    return jsxRuntime.jsx(BreadcrumbItem$1, { ...props });
}

function Breadcrumbs(props) {
    return jsxRuntime.jsx(Breadcrumbs$1, { ...props });
}
Breadcrumbs.Item = BreadcrumbItem;

exports.BreadcrumbItem = BreadcrumbItem;
exports.Breadcrumbs = Breadcrumbs;
