// @hostaway/design-system/dist/components/Avatar/index.esm.js
export { A as Avatar } from '../_shared/Avatar-y2Sg5Aio.esm.js';
export { A as AvatarAddButton } from '../_shared/AvatarAddButton-BNXEkvZt.esm.js';
import { jsxs, jsx } from 'react/jsx-runtime';
import { useState } from 'react';
import { User01 } from '@untitledui/icons';
import { c as cx } from '../_shared/cx-DnIR8RkH.esm.js';
import 'react-aria-components';
import { a as AvatarOnlineIndicator, V as VerifiedTick } from '../_shared/avatar-DeJPrYaB.esm.js';
import { A as AvatarLabelGroup$1 } from '../_shared/avatar-label-group-D5Odzrz2.esm.js';
import '../_shared/tooltip-l8auK4sw.esm.js';
import 'tailwind-merge';

const styles = {
    sm: {
        root: "size-18 p-0.75",
        rootWithPlaceholder: "p-1",
        content: "",
        icon: "size-9",
        initials: "text-display-sm font-semibold",
        badge: "bottom-0.5 right-0.5",
    },
    md: {
        root: "size-24 p-1",
        rootWithPlaceholder: "p-1.25",
        content: "shadow-xl",
        icon: "size-12",
        initials: "text-display-md font-semibold",
        badge: "bottom-1 right-1",
    },
    lg: {
        root: "size-40 p-1.5",
        rootWithPlaceholder: "p-1.75",
        content: "shadow-2xl",
        icon: "size-20",
        initials: "text-display-xl font-semibold",
        badge: "bottom-2 right-2",
    },
};
const tickSizeMap = {
    sm: "2xl",
    md: "3xl",
    lg: "4xl",
};
const AvatarProfilePhoto$1 = ({ contrastBorder = true, size = "md", src, alt, initials, placeholder, placeholderIcon: PlaceholderIcon, verified, badge, status, className, }) => {
    const [isFailed, setIsFailed] = useState(false);
    const renderMainContent = () => {
        if (src && !isFailed) {
            return (jsx("img", { src: src, alt: alt, onError: () => setIsFailed(true), className: cx("size-full rounded-full object-cover", contrastBorder && "outline-1 -outline-offset-1 outline-avatar-contrast-border", styles[size].content) }));
        }
        if (initials) {
            return (jsx("div", { className: cx("flex size-full items-center justify-center rounded-full bg-tertiary ring-1 ring-secondary_alt", styles[size].content), children: jsx("span", { className: cx("text-quaternary", styles[size].initials), children: initials }) }));
        }
        if (PlaceholderIcon) {
            return (jsx("div", { className: cx("flex size-full items-center justify-center rounded-full bg-tertiary ring-1 ring-secondary_alt", styles[size].content), children: jsx(PlaceholderIcon, { className: cx("text-fg-quaternary", styles[size].icon) }) }));
        }
        return (jsx("div", { className: cx("flex size-full items-center justify-center rounded-full bg-tertiary ring-1 ring-secondary_alt", styles[size].content), children: placeholder || jsx(User01, { className: cx("text-fg-quaternary", styles[size].icon) }) }));
    };
    const renderBadgeContent = () => {
        if (status) {
            return jsx(AvatarOnlineIndicator, { status: status, size: tickSizeMap[size], className: styles[size].badge });
        }
        if (verified) {
            return jsx(VerifiedTick, { size: tickSizeMap[size], className: cx("absolute", styles[size].badge) });
        }
        return badge;
    };
    return (jsxs("div", { className: cx("relative flex shrink-0 items-center justify-center rounded-full bg-primary ring-1 ring-secondary_alt", styles[size].root, (!src || isFailed) && styles[size].rootWithPlaceholder, className), children: [renderMainContent(), renderBadgeContent()] }));
};

function AvatarProfilePhoto({ ...props }) {
    return jsx(AvatarProfilePhoto$1, { ...props });
}

function AvatarLabelGroup(props) {
    return jsx(AvatarLabelGroup$1, { ...props });
}

export { AvatarLabelGroup, AvatarProfilePhoto };
