// @hostaway/design-system/dist/components/Avatar/index.cjs.js
'use strict';

var Avatar = require('../_shared/Avatar-C1JKrxyc.cjs.js');
var AvatarAddButton = require('../_shared/AvatarAddButton-DvP2ARtY.cjs.js');
var jsxRuntime = require('react/jsx-runtime');
var h = require('react');
var icons = require('@untitledui/icons');
var cx = require('../_shared/cx-CmkbyMOt.cjs.js');
require('react-aria-components');
var avatar = require('../_shared/avatar-C5INEgx9.cjs.js');
var avatarLabelGroup = require('../_shared/avatar-label-group-Vv_UjMoJ.cjs.js');
require('../_shared/tooltip-DDB8niG0.cjs.js');
require('tailwind-merge');

const styles = {
    sm: {
        root: "size-18 p-0.75",
        rootWithPlaceholder: "p-1",
        content: "",
        icon: "size-9",
        initials: "text-display-sm font-semibold",
        badge: "bottom-0.5 right-0.5",
    },
    md: {
        root: "size-24 p-1",
        rootWithPlaceholder: "p-1.25",
        content: "shadow-xl",
        icon: "size-12",
        initials: "text-display-md font-semibold",
        badge: "bottom-1 right-1",
    },
    lg: {
        root: "size-40 p-1.5",
        rootWithPlaceholder: "p-1.75",
        content: "shadow-2xl",
        icon: "size-20",
        initials: "text-display-xl font-semibold",
        badge: "bottom-2 right-2",
    },
};
const tickSizeMap = {
    sm: "2xl",
    md: "3xl",
    lg: "4xl",
};
const AvatarProfilePhoto$1 = ({ contrastBorder = true, size = "md", src, alt, initials, placeholder, placeholderIcon: PlaceholderIcon, verified, badge, status, className, }) => {
    const [isFailed, setIsFailed] = h.useState(false);
    const renderMainContent = () => {
        if (src && !isFailed) {
            return (jsxRuntime.jsx("img", { src: src, alt: alt, onError: () => setIsFailed(true), className: cx.cx("size-full rounded-full object-cover", contrastBorder && "outline-1 -outline-offset-1 outline-avatar-contrast-border", styles[size].content) }));
        }
        if (initials) {
            return (jsxRuntime.jsx("div", { className: cx.cx("flex size-full items-center justify-center rounded-full bg-tertiary ring-1 ring-secondary_alt", styles[size].content), children: jsxRuntime.jsx("span", { className: cx.cx("text-quaternary", styles[size].initials), children: initials }) }));
        }
        if (PlaceholderIcon) {
            return (jsxRuntime.jsx("div", { className: cx.cx("flex size-full items-center justify-center rounded-full bg-tertiary ring-1 ring-secondary_alt", styles[size].content), children: jsxRuntime.jsx(PlaceholderIcon, { className: cx.cx("text-fg-quaternary", styles[size].icon) }) }));
        }
        return (jsxRuntime.jsx("div", { className: cx.cx("flex size-full items-center justify-center rounded-full bg-tertiary ring-1 ring-secondary_alt", styles[size].content), children: placeholder || jsxRuntime.jsx(icons.User01, { className: cx.cx("text-fg-quaternary", styles[size].icon) }) }));
    };
    const renderBadgeContent = () => {
        if (status) {
            return jsxRuntime.jsx(avatar.AvatarOnlineIndicator, { status: status, size: tickSizeMap[size], className: styles[size].badge });
        }
        if (verified) {
            return jsxRuntime.jsx(avatar.VerifiedTick, { size: tickSizeMap[size], className: cx.cx("absolute", styles[size].badge) });
        }
        return badge;
    };
    return (jsxRuntime.jsxs("div", { className: cx.cx("relative flex shrink-0 items-center justify-center rounded-full bg-primary ring-1 ring-secondary_alt", styles[size].root, (!src || isFailed) && styles[size].rootWithPlaceholder, className), children: [renderMainContent(), renderBadgeContent()] }));
};

function AvatarProfilePhoto({ ...props }) {
    return jsxRuntime.jsx(AvatarProfilePhoto$1, { ...props });
}

function AvatarLabelGroup(props) {
    return jsxRuntime.jsx(avatarLabelGroup.AvatarLabelGroup, { ...props });
}

exports.Avatar = Avatar.Avatar;
exports.AvatarAddButton = AvatarAddButton.AvatarAddButton;
exports.AvatarLabelGroup = AvatarLabelGroup;
exports.AvatarProfilePhoto = AvatarProfilePhoto;
